{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "machine_shape": "hm",
      "collapsed_sections": [
        "QbelVNKWf4X_"
      ],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU",
    "gpuClass": "standard"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/DPariser/DataScience/blob/main/Preprocessing/041823_DNP1_QC_and_Pre_Processing_FASTQ.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Single-cell RNA-seq data processing\n",
        "** Information found in publication: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7857060/\n",
        "\n",
        "Single-cell sequencing data were aligned and quantified using kallisto/bustools (KB, v0.24.4) (Bray et al., 2016 [link text](https://www.nature.com/articles/nbt.3519)) against the GRCh38 human reference genome downloaded from 10x Genomics official website. Preliminary counts were then used for downstream analysis. Quality control was applied to cells based on three metrics step by step: the total UMI counts, number of detected genes and proportion of mitochondrial gene counts per cell. Specifically, cells with less than 1000 UMI counts and 500 detected genes were filtered, as well as cells with more than 10% mitochondrial gene counts. To remove potential doublets, for PBMC samples, cells with UMI counts above 25,000 and detected genes above 5,000 are filtered out. For other tissues, cells with UMI counts above\n",
        "70,000 and detected genes above 7,500 are filtered out. Additionally, we applied Scrublet (Wolock et al., 2019 [link text](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6625319/pdf/nihms-1515604.pdf)) to identify potential\n",
        "doublets. The doublet score for each single cell and the threshold based on the bimodal distribution was calculated using default\n",
        "parameters. The expected doublet rate was set to be 0.08, and cells predicted to be doublets or with doubletScore larger than  0.25 were filtered. After quality control, a total of 1,598,708 cells were remained. The stepwise quality control metrics used for indi-\n",
        "vidual samples were listed in Table S1. The resulting distribution of UMI counts, gene counts as well as mitochondrial gene percent-\n",
        "age were shown in Figures S1C–S1E. We normalized the UMI counts with the deconvolution strategy implemented in the R package scran. Specifically, cell-specific size factors were computed by computeSumFactors function and further used to scale the counts for\n",
        "each cell. Then the logarithmic normalized counts were used for the downstream analysis.\n"
      ],
      "metadata": {
        "id": "hDqtukMzF_5K"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Setup Environment\n",
        "\n",
        "This Notebook is created fresh with nothing else installed explicitly besides what is shown. So we assume that if you follow the instruction exactly, it should run out of the box."
      ],
      "metadata": {
        "id": "oGtI6PQ3ZBeD"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "vYulgDUCh3il"
      },
      "outputs": [],
      "source": [
        "# This is  used to time the running of the notebook\n",
        "import time\n",
        "start_time = time.time()"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# These packages are pre-installed on Google Colab, but are included here to simplify running this notebook locally\n",
        "%%capture\n",
        "!pip install matplotlib\n",
        "!pip install scikit-learn\n",
        "!pip install numpy\n",
        "!pip install scipy"
      ],
      "metadata": {
        "id": "5FmtakqDoNt-"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Install packages for analysis and plotting\n",
        "from scipy.io import mmread\n",
        "from sklearn.decomposition import TruncatedSVD\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "import matplotlib\n",
        "import os\n",
        "\n",
        "from scipy.sparse import csr_matrix\n",
        "matplotlib.rcParams.update({'font.size': 22})\n",
        "%config InlineBackend.figure_format = 'retina'"
      ],
      "metadata": {
        "id": "qI2gKIkYYmbO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Install kb\n",
        "\n",
        "Tutorials & Notebooks\n",
        "* https://www.kallistobus.tools/tutorials/kb_quality_control/python/kb_intro_1_python/\n",
        "* https://github.com/pachterlab/kallisto-transcriptome-indices\n",
        "* https://docs.google.com/presentation/d/1QUmi1Mm5dJ1UyQIT_5XAG9806XL4qGfb3OUDrlIvIqs/edit#slide=id.gef29e9d7dc_1_82"
      ],
      "metadata": {
        "id": "9KZz4ZNYjW5O"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%time\n",
        "%%capture\n",
        "# `kb` is a wrapper for the kallisto and bustools program, and the kb-python package contains the kallisto and bustools executables.\n",
        "!pip install kb-python==0.24.1"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "iQoeB8XfjZgX",
        "outputId": "46589c9c-8708-40a6-e268-0702c8205895"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "CPU times: user 15.3 ms, sys: 12.8 ms, total: 28.1 ms\n",
            "Wall time: 1.82 s\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "9jVVpWZu6iBI",
        "outputId": "464f6686-e684-4efb-bdcb-82d0d1ed8839"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## ❗**Connect to the Data**\n",
        "\n",
        "The data is stored on a shared location in Google Drive. Since many of the files are very large and thus it is not feasable to download them to a location and use them. One good way of dealing with this situation is to create a shortcut to your own Google Drive and point to the shortcut and use them just like they are your own files on Google Drive. Here is the instruction how to set this up.\n",
        "\n",
        "* Click on the link to the share location of the data.\n",
        "* Nevigate to the \"Data files\" folder.\n",
        "* Click on the \"Dropdown\" arrow right next to the breaksrumb on the top right.\n",
        "* Choose \"Add shortcut to Drive\".\n",
        "\n",
        "Now it should appear in your Google Drive as the \"Data files\" folder.\n",
        "You can now connect to your Google Drive and access the file.\n",
        "From this point on, we assume that you have the Google Drive setup this way.\n",
        "\n",
        "Let's mount the Google Drive:\n",
        "\n"
      ],
      "metadata": {
        "id": "mD0SrEINZiYW"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Google drive root\n",
        "gd_root = \"/content/drive/MyDrive/Pate_Lab/DNP/Bioinformatics/H17_LungMk\"\n",
        "# Data root\n",
        "data_root = f\"{gd_root}/Data_files/HRA001149/HRR339729\"\n",
        "# working directory\n",
        "work_dir = f\"{gd_root}/LungMk\"\n",
        "\n",
        "# create the directory \n",
        "!mkdir -p \"{work_dir}\""
      ],
      "metadata": {
        "id": "nzgOn66sZ78p"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "print(work_dir)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8MtmMNirElyc",
        "outputId": "06e5b924-2596-4c0b-9e99-53c3a8abfd37"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/content/drive/MyDrive/Pate_Lab/DNP/Bioinformatics/H17_LungMk/LungMk\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        " !ls \"{work_dir}\""
      ],
      "metadata": {
        "id": "JMlBv3laEo7u",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "378edfaf-4516-42ee-eba3-4a7a36679517"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            " genes_t2g.txt\t genome.idx   GRCh38genome.idx\t'GRCh38genome.idx 2'\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "if os.path.exists(work_dir):\n",
        "    print(f\"The directory {work_dir} exists.\")\n",
        "else:\n",
        "    print(f\"The directory {work_dir} does not exist.\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "eljz9Pl8WwIO",
        "outputId": "f02784cf-1e2d-40fa-bb2c-c885eb022c1a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "The directory /content/drive/MyDrive/Pate_Lab/DNP/Bioinformatics/H17_LungMk/LungMk exists.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "Ty-3XI_EElbp"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Prepare for quantification for GRCh38\n",
        "\n",
        "GRCh38 stands for Genome Reference Consortium Human Build 38, which is the most recent version of the human reference genome. It is a widely used reference genome for human genetics and genomics research, and provides a standardized framework for comparing and interpreting genomic data across studies. The GRCh38 assembly was released in December 2013 and contains more than 3 billion base pairs, representing the genetic information of a complete set of human chromosomes. It includes both coding and non-coding regions of the genome and is used as a reference for many different applications, such as genome sequencing, variant calling, and gene expression analysis.\n",
        "\n",
        "Here we are downloading two files that enable us to process the FastQ files we **have**:\n",
        "\n",
        "1. The index file: this is a k-mer index built from a reference transcriptome. It is a binary file that contains information on the position of each k-mer within the transcriptome, which is used to map the reads to the transcriptome during the quantification step. The index file is generated by a software tool such as kallisto or salmon using a reference transcriptome in FASTA format. \n",
        "2. The t2g file: this is a file that maps each transcript to a corresponding gene. This mapping is necessary because RNA sequencing reads can align to either a specific transcript or a gene, and the counts for each transcript need to be aggregated to obtain gene-level expression. The t2g file can be used to generate gene-level counts from transcript-level counts."
      ],
      "metadata": {
        "id": "ZX8PVZCjavjK"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Download the index and t2g files"
      ],
      "metadata": {
        "id": "WvV4exMXoL0o"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n"
      ],
      "metadata": {
        "id": "fJPIh5T4aQPT"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%time\n",
        "# The quantification of single-cell RNA-seq with kallisto requires an index. \n",
        "# Indices are species specific and can be generated or downloaded directly with `kb`. \n",
        "# Here we download a pre-made index for human (the genome.idx.gz file) along with an auxillary file (genes_t2g.txt) \n",
        "# that describes the relationship between transcripts and genes.\n",
        "\n",
        "index_url = \"https://github.com/pachterlab/kallisto-transcriptome-indices/releases/download/94/Homo_sapiens.GRCh38.cdna.all.release-94_k31.idx.gz\"\n",
        "t2g_url = \"https://github.com/BUStools/getting_started/releases/download/species_mixing/transcripts_to_genes_hg19mm10.txt\"\n",
        "\n",
        "!wget {index_url} -O \"{work_dir}/genome.idx.gz\"\n",
        "!wget {t2g_url} -O \"{work_dir}/genes_t2g.txt\"\n",
        "\n",
        "# Unzip the genome index file\n",
        "!gunzip -f \"{work_dir}/genome.idx.gz\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "QdhoTJ10cCaz",
        "outputId": "940be59f-7e06-427b-8cdb-29ac0f292596"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2023-04-20 22:45:21--  https://github.com/pachterlab/kallisto-transcriptome-indices/releases/download/94/Homo_sapiens.GRCh38.cdna.all.release-94_k31.idx.gz\n",
            "Resolving github.com (github.com)... 20.27.177.113\n",
            "Connecting to github.com (github.com)|20.27.177.113|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/160138161/82cc8d80-f68f-11e8-8faf-e4656ce5fc8f?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230420%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230420T224521Z&X-Amz-Expires=300&X-Amz-Signature=c351a0eb38f9fc90f497a4d4b7b71942ab193f2c5889140dacd1fd7f1359e76e&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=160138161&response-content-disposition=attachment%3B%20filename%3DHomo_sapiens.GRCh38.cdna.all.release-94_k31.idx.gz&response-content-type=application%2Foctet-stream [following]\n",
            "--2023-04-20 22:45:22--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/160138161/82cc8d80-f68f-11e8-8faf-e4656ce5fc8f?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230420%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230420T224521Z&X-Amz-Expires=300&X-Amz-Signature=c351a0eb38f9fc90f497a4d4b7b71942ab193f2c5889140dacd1fd7f1359e76e&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=160138161&response-content-disposition=attachment%3B%20filename%3DHomo_sapiens.GRCh38.cdna.all.release-94_k31.idx.gz&response-content-type=application%2Foctet-stream\n",
            "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.110.133, 185.199.109.133, 185.199.111.133, ...\n",
            "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.110.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1841920271 (1.7G) [application/octet-stream]\n",
            "Saving to: ‘/content/drive/MyDrive/Pate_Lab/DNP/Bioinformatics/H17_LungMk/LungMk/genome.idx.gz’\n",
            "\n",
            "/content/drive/MyDr 100%[===================>]   1.71G  79.7MB/s    in 24s     \n",
            "\n",
            "2023-04-20 22:45:46 (72.7 MB/s) - ‘/content/drive/MyDrive/Pate_Lab/DNP/Bioinformatics/H17_LungMk/LungMk/genome.idx.gz’ saved [1841920271/1841920271]\n",
            "\n",
            "--2023-04-20 22:45:46--  https://github.com/BUStools/getting_started/releases/download/species_mixing/transcripts_to_genes_hg19mm10.txt\n",
            "Resolving github.com (github.com)... 20.27.177.113\n",
            "Connecting to github.com (github.com)|20.27.177.113|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/191064839/37b22d80-913c-11e9-9025-bca8558040be?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230420%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230420T224546Z&X-Amz-Expires=300&X-Amz-Signature=09b311ac2df5499efe19c3c5e47afffd68ce44fb93fbcf5c7929bc2108f21d4e&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=191064839&response-content-disposition=attachment%3B%20filename%3Dtranscripts_to_genes_hg19mm10.txt&response-content-type=application%2Foctet-stream [following]\n",
            "--2023-04-20 22:45:46--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/191064839/37b22d80-913c-11e9-9025-bca8558040be?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230420%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230420T224546Z&X-Amz-Expires=300&X-Amz-Signature=09b311ac2df5499efe19c3c5e47afffd68ce44fb93fbcf5c7929bc2108f21d4e&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=191064839&response-content-disposition=attachment%3B%20filename%3Dtranscripts_to_genes_hg19mm10.txt&response-content-type=application%2Foctet-stream\n",
            "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.110.133, 185.199.109.133, 185.199.111.133, ...\n",
            "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.110.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 18953747 (18M) [application/octet-stream]\n",
            "Saving to: ‘/content/drive/MyDrive/Pate_Lab/DNP/Bioinformatics/H17_LungMk/LungMk/genes_t2g.txt’\n",
            "\n",
            "/content/drive/MyDr 100%[===================>]  18.08M  71.5MB/s    in 0.3s    \n",
            "\n",
            "2023-04-20 22:45:47 (71.5 MB/s) - ‘/content/drive/MyDrive/Pate_Lab/DNP/Bioinformatics/H17_LungMk/LungMk/genes_t2g.txt’ saved [18953747/18953747]\n",
            "\n",
            "CPU times: user 575 ms, sys: 107 ms, total: 682 ms\n",
            "Wall time: 1min 5s\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!ls \"{work_dir}\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Sm755ZnoVrPI",
        "outputId": "ecd15c0d-e76c-4ab6-c6f1-3a02a48c1afc"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            " genes_t2g.txt\t genome.idx   GRCh38genome.idx\t'GRCh38genome.idx 2'\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# inspect the index files to ensure that they were downloaded properly\n",
        "genome_idx_size = os.path.getsize(f\"{work_dir}/genome.idx\") / (1024*1024*1024)\n",
        "genes_t2g_size = os.path.getsize(f\"{work_dir}/genes_t2g.txt\") / (1024*1024*1024)\n",
        "\n",
        "print(f\"genome.idx size: {genome_idx_size:.2f} GB\")\n",
        "print(f\"genes_t2g.txt size: {genes_t2g_size:.2f} GB\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "m0JCFX30XrD3",
        "outputId": "0f9b4bf9-b245-4a5c-d0d6-11f1e544bc89"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "genome.idx size: 2.27 GB\n",
            "genes_t2g.txt size: 0.02 GB\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!wget -S -O /dev/null \"https://github.com/pachterlab/kallisto-transcriptome-indices/releases/download/94/Homo_sapiens.GRCh38.cdna.all.release-94_k31.idx.gz\" 2>&1 | grep 'Content-Length' | awk '{print $2 / (1024^3) \" GB\"}'\n",
        "!wget -S -O /dev/null \"https://github.com/BUStools/getting_started/releases/download/species_mixing/transcripts_to_genes_hg19mm10.txt\" 2>&1 | grep 'Content-Length' | awk '{print $2 / (1024^3) \" GB\"}'"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "QeCWup4PYtKl",
        "outputId": "c6e57ab8-f111-46fa-ad9e-7c2b247a69ca"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "0 GB\n",
            "1.71542 GB\n",
            "0 GB\n",
            "0.0176521 GB\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!head \"{work_dir}/genome.idx\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "hY1OEbgjcUKm",
        "outputId": "4e617260-b886-4ef5-cb49-77c00adeb5b2"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u001f\u0000\u0000\u0000��\u0002\u0000\f\u0000\u0000\u0000\t\u0000\u0000\u0000\r\u0000\u0000\u0000\b\u0000\u0000\u0000\f\u0000\u0000\u0000\u001f\u0000\u0000\u0000%\u0000\u0000\u0000\u0011\u0000\u0000\u0000\u001f\u0000\u0000\u0000\u001c\u0000\u0000\u0000\u0015\u0000\u0000\u0000\u000b\u0000\u0000\u0000\u001f\u0000\u0000\u0000\u0014\u0000\u0000\u0000\u0010\u0000\u0000\u0000\u0010\u0000\u0000\u0000\u0014\u0000\u0000\u0000\u001f\u0000\u0000\u0000\u0013\u0000\u0000\u0000\u0014\u0000\u0000\u0000\u001f\u0000\u0000\u0000\u0010\u0000\u0000\u0000\u001f\u0000\u0000\u0000\u0017\u0000\u0000\u0000\u0015\u0000\u0000\u0000\u0015\u0000\u0000\u0000\u0012\u0000\u0000\u0000\u0011\u0000\u0000\u0000\u0012\u0000\u0000\u0000\u0011\u0000\u0000\u0000\u001f\u0000\u0000\u0000\u0011\u0000\u0000\u0000\u0012\u0000\u0000\u0000\u0011\u0000\u0000\u0000\u001f\u0000\u0000\u0000\u0017\u0000\u0000\u0000\u0011\u0000\u0000\u0000\u0014\u0000\u0000\u0000\u001f\u0000\u0000\u0000\u0011\u0000\u0000\u0000\u0014\u0000\u0000\u0000\u000b\u0000\u0000\u0000\u001f\u0000\u0000\u0000\u0011\u0000\u0000\u0000\u0011\u0000\u0000\u0000\u001f\u0000\u0000\u0000\u0013\u0000\u0000\u0000\u001f\u0000\u0000\u0000\u0013\u0000\u0000\u0000\u001f\u0000\u0000\u0000\u0010\u0000\u0000\u0000\u0017\u0000\u0000\u0000\u001f\u0000\u0000\u0000\u0014\u0000\u0000\u0000\u001f\u0000\u0000\u0000\u001f\u0000\u0000\u0000\u001c\u0000\u0000\u0000\u001f\u0000\u0000\u0000\u0013\u0000\u0000\u0000\u001f\u0000\u0000\u0000\u0014\u0000\u0000\u0000\u0010\u0000\u0000\u0000\u0017\u0000\u0000\u0000%\u0000\u0000\u0000\u0015\u0000\u0000\u0000\u0010\u0000\u0000\u0000\u0012\u0000\u0000\u0000\u0014\u0000\u0000\u0000\u0011\u0000\u0000\u0000[\u0001\u0000\u0000l\u0001\u0000\u0000�\u0001\u0000\u0000^\u0001\u0000\u0000[\u0001\u0000\u0000P\u0001\u0000\u0000\\\u0001\u0000\u0000d\u0001\u0000\u0000�\u0001\u0000\u0000h\u0001\u0000\u0000U\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000p\u0001\u0000\u0000[\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000}\u0001\u0000\u0000W\u0001\u0000\u0000 \u0002\u0000\u0000m\u0001\u0000\u0000\u0007\u0002\u0000\u0000�\u0001\u0000\u0000X\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000u\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000v\u0001\u0000\u0000U\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000X\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000l\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000a\u0001\u0000\u0000�\u0001\u0000\u0000G\u0001\u0000\u0000\u0019\u0002\u0000\u0000W\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000R\u0001\u0000\u0000�\u0001\u0000\u0000=\u0001\u0000\u0000�\u0001\u0000\u0000W\u0001\u0000\u0000I\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000}\u0001\u0000\u0000W\u0001\u0000\u0000�\u0002\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000Q\u0001\u0000\u0000U\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000m\u0001\u0000\u0000f\u0000\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000[\u0001\u0000\u0000K\u0001\u0000\u0000]\u0001\u0000\u0000�\u0001\u0000\u00001\u0002\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000(\u0002\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000|\u0001\u0000\u0000�\u0001\u0000\u0000T\u0001\u0000\u0000{\u0001\u0000\u0000�\u0001\u0000\u0000Q\u0001\u0000\u0000`\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000[\u0001\u0000\u0000\b\u0002\u0000\u0000�\u0001\u0000\u00006\u0002\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000l\u0001\u0000\u0000�\u0001\u0000\u0000}\u0001\u0000\u0000(\u0002\u0000\u0000X\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0002\u0000\u0000�\u0002\u0000\u0000[\u0001\u0000\u0000v\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000l\u0001\u0000\u0000|\u0001\u0000\u0000�\u0001\u0000\u0000}\u0001\u0000\u0000�\u0001\u0000\u0000[\u0001\u0000\u0000�\u0006\u0000\u0000\u0019\u0002\u0000\u0000W\u0001\u0000\u0000[\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000 \u0002\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000 \u0002\u0000\u0000�\u0001\u0000\u0000[\u0001\u0000\u0000�\u0001\u0000\u0000(\u0002\u0000\u0000^\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000m\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000X\u0001\u0000\u0000[\u0001\u0000\u0000W\u0001\u0000\u0000�\u0001\u0000\u0000\u0000\u0002\u0000\u0000X\u0001\u0000\u0000�\u0001\u0000\u0000[\u0001\u0000\u0000u\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000\u000f\u0005\u0000\u0000�\u0003\u0000\u0000�\u0003\u0000\u0000�\u0003\u0000\u0000�\u0003\u0000\u0000�\n",
            "\u0000\u0000�\u0003\u0000\u0000�\u0004\u0000\u0000�\u0006\u0000\u0000�\u0005\u0000\u0000�\u0002\u0000\u0000(\u0006\u0000\u0000�\u0006\u0000\u0000S\u0005\u0000\u0000=\u0002\u0000\u00004\u0002\u0000\u0000�\u0006\u0000\u0000�\u0005\u0000\u0000\u000f\u0005\u0000\u0000Q\r\u0000\u0000�\r\u0000\u0000\u0015\u0004\u0000\u0000�\u0003\u0000\u0000^\u0002\u0000\u0000�\u0001\u0000\u0000\u001b\u0006\u0000\u0000\u0018\u0013\u0000\u0000\u0018\u0011\u0000\u0000\u0017\u0001\u0000\u0000X\u0004\u0000\u0000>\u0001\u0000\u0000�\u0006\u0000\u0000�\u0005\u0000\u0000�\u0005\u0000\u0000e\u0006\u0000\u0000l\u0004\u0000\u0000\u000b\u000e\u0000\u0000�\t\u0000\u0000�\u0006\u0000\u0000�9\u0000\u0000�\u0002\u0000\u0000�\u0001\u0000\u0000)\u0002\u0000\u0000C\u001f\u0000\u0000�\u0002\u0000\u0000\"\u0002\u0000\u00006\u0002\u0000\u0000*\u0002\u0000\u0000.\u0004\u0000\u0000\u0001\u0003\u0000\u0000`\u0006\u0000\u00007\u0002\u0000\u0000\u000f\u0002\u0000\u0000M\u0004\u0000\u0000)\u0002\u0000\u0000\u001e\u0002\u0000\u0000�\u0003\u0000\u0000�\u0005\u0000\u0000H\u0006\u0000\u0000�\u0002\u0000\u0000�\u0007\u0000\u0000!\u0005\u0000\u0000!\u0002\u0000\u0000�\u0004\u0000\u0000�\u0003\u0000\u0000�\u0003\u0000\u0000�\u0004\u0000\u0000X\u0004\u0000\u0000�\u0004\u0000\u0000�\u0003\u0000\u0000�\u0002\u0000\u0000�\u0003\u0000\u0000\u0006\u0004\u0000\u0000�\u0003\u0000\u0000�\u0005\u0000\u0000N\u0002\u0000\u0000n\u0002\u0000\u0000\u0005\u0000\u0000�\u0005\u0000\u0000�\n",
            "\u0000\u0000�\u0004\u0000\u0000*\u0005\u0000\u0000�\u0003\u0000\u0000�\u0003\u0000\u0000�\u0003\u0000\u0000�\u0003\u0000\u0000R\u000b\u0000\u0000\t2\u0000\u0000�\u0002\u0000\u0000�\u0001\u0000\u0000)\u0002\u0000\u0000C\u001f\u0000\u0000�\u0003\u0000\u0000�\u0003\u0000\u0000C\u000b\u0000\u0000�\u0003\u0000\u0000�\u0007\u0000\u0000�\b\u0000\u0000t\b\u0000\u0000�\u0004\u0000\u0000�\u0007\u0000\u0000X\u0002\u0000\u0000�\u0002\u0000\u0000�\u0004\u0000\u0000j\b\u0000\u0000�\u0005\u0000\u0000�\u0004\u0000\u0000\u001a\u0005\u0000\u00003\u0003\u0000\u0000�\u0004\u0000\u0000�\u0004\u0000\u0000�\u0002\u0000\u0000/\u0004\u0000\u0000y\u0005\u0000\u0000�\u0003\u0000\u0000\u0006\u0003\u0000\u0000�\u0006\u0000\u0000w\u0006\u0000\u0000�\u0006\u0000\u0000=\u0002\u0000\u00004\u0002\u0000\u0000�\u0005\u0000\u0000�\u0006\u0000\u0000�\u0003\u0000\u0000*\u0005\u0000\u0000\u0004\u0000\u0000o\u0004\u0000\u0000�\u0002\u0000\u0000M\u0002\u0000\u0000D\u0002\u0000\u0000�\u0003\u0000\u0000�\u0002\u0000\u0000w\u0006\u0000\u0000�\u0006\u0000\u0000S\u0005\u0000\u0000=\u0002\u0000\u00004\u0002\u0000\u0000�\u0005\u0000\u0000�\u0006\u0000\u0000I\u0006\u0000\u00001\u0005\u0000\u0000F\u0006\u0000\u0000�\u0003\u0000\u0000�\u0004\u0000\u0000�\u0002\u0000\u0000�\u0014\u0000\u0000�\u0007\u0000\u0000%\u0016\u0000\u0000�\u001c\u0000\u0000H\u0006\u0000\u0000�\u0016\u0000\u0000>\u0002\u0000\u0000�\u0005\u0000\u0000�\u0007\u0000\u0000E\u0010\u0000\u0000(\u0002\u0000\u0000c\r\u0000\u0000�\u000e\u0000\u0000�\u000e\u0000\u0000D\u0010\u0000\u0000e\u0002\u0000\u0000D\u0010\u0000\u0000R\u000b\u0000\u0000�\u0002\u0000\u0000w\u0006\u0000\u0000�\u0006\u0000\u0000S\u0005\u0000\u0000=\u0002\u0000\u00004\u0002\u0000\u0000�\u0005\u0000\u0000�\u0006\u0000\u0000H\u0006\u0000\u0000�\u0002\u0000\u0000H\u0006\u0000\u0000�\u0002\u0000\u0000H\u0006\u0000\u0000�\u0002\u0000\u0000H\u0006\u0000\u0000�\u0002\u0000\u0000H\u0006\u0000\u0000�\u0002\u0000\u0000H\u0006\u0000\u0000�\u0002\u0000\u0000�\u0002\u0000\u0000�\u0003\u0000\u0000�\u0002\u0000\u0000�\u0002\u0000\u0000�\u0002\u0000\u0000�\u0002\u0000\u0000�\n",
            "\u0000\u0000�\u0003\u0000\u00004\u0000\u0000\u0000!\u0000\u0000\u0000%\u0000\u0000\u00001\u0000\u0000\u00001\u0000\u0000\u0000.\u0000\u0000\u00002\u0000\u0000\u0000=\u0000\u0000\u0000&\u0000\u0000\u0000F\u0000\u0000\u00004\u0000\u0000\u0000&\u0000\u0000\u0000'\u0000\u0000\u00004\u0000\u0000\u0000&\u0000\u0000\u0000\u0000\u0000\u0000�\u0000\u0000\u0000�\u0000\u0000\u00003\u0000\u0000\u00000\u0000\u0000\u0000H\u0000\u0000\u0000.\u0000\u0000\u0000?\u0000\u0000\u00005\u0000\u0000\u0000\"\n",
            "\u0000\u0000\\\u0004\u0000\u0000%\n",
            "\u0000\u0000_\u0004\u0000\u0000�\u0005\u0000\u0000�\u0005\u0000\u0000;\n",
            "\u0000\u0000h\u0004\u0000\u0000�\u0003\u0000\u0000Z\u0004\u0000\u0000�\u0006\u0000\u0000�\u0005\u0000\u0000\u001e\u0005\u0000\u0000X\u0004\u0000\u0000�\n",
            "\u0000\u0000�\u0004\u0000\u0000�\u0006\u0000\u0000�\u0004\u0000\u0000(\u0005\u0000\u0000/\u0004\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000\u000b\u0002\u0000\u0000�\u0001\u0000\u0000X\u0004\u0000\u0000�\u0005\u0000\u0000�\u0004\u0000\u0000/\u0004\u0000\u0000�\u0005\u0000\u0000�\u0006\u0000\u0000;\n",
            "\u0000\u0000h\u0004\u0000\u0000�\u0003\u0000\u0000_\u0004\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000:\u0001\u0000\u00008\u0001\u0000\u0000M\u0002\u0000\u0000f\u0001\u0000\u0000a\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000\u0016\u0002\u0000\u0000a\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000a\u0001\u0000\u0000�\u0001\u0000\u0000a\u0001\u0000\u0000�\u0001\u0000\u0000z\u0001\u0000\u0000;\u0002\u0000\u0000�\u0002\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000K\u0002\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000}\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000a\u0001\u0000\u0000^\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000q\u0001\u0000\u0000�\u0001\u0000\u0000v\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000D\u0001\u0000\u0000x\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000/\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000z\u0001\u0000\u0000a\u0001\u0000\u0000�\u0001\u0000\u0000a\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000n\u0001\u0000\u0000�\u0001\u0000\u0000u\u0001\u0000\u0000\u0014\u0002\u0000\u0000\u0016\u0002\u0000\u0000�\u0001\u0000\u0000a\u0001\u0000\u0000�\u0001\u0000\u0000\u0001\u0000\u0000�\u0001\u0000\u0000x\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000z\u0001\u0000\u0000a\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000~\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000\u0014\u0002\u0000\u0000a\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000y\u0001\u0000\u0000�\u0001\u0000\u0000\u0001\u0000\u0000|\u0001\u0000\u0000\u0014\u0002\u0000\u0000�\u0001\u0000\u0000|\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000s\u0001\u0000\u0000�\u0001\u0000\u00006\u0001\u0000\u0000~\u0001\u0000\u0000�\u0001\u0000\u0000&\u0001\u0000\u0000 \u0001\u0000\u0000�\u0001\u0000\u0000;\u0002\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000a\u0001\u0000\u0000\u001a\u0002\u0000\u0000s\u0001\u0000\u0000�\u0001\u0000\u0000L\u0001\u0000\u00008\u0001\u0000\u0000z\u0001\u0000\u0000�\u0001\u0000\u0000a\u0001\u0000\u0000�\u0001\u0000\u0000y\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000a\u0001\u0000\u0000�\u0001\u0000\u0000m\u0001\u0000\u0000_\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u00002\u0001\u0000\u0000d\u0001\u0000\u0000�\u0001\u0000\u0000^\u0001\u0000\u0000�\u0001\u0000\u0000/\u0001\u0000\u0000d\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000a\u0001\u0000\u0000a\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000{\u0001\u0000\u0000�\u0001\u0000\u0000]\u0001\u0000\u0000�\u0001\u0000\u0000]\u0001\u0000\u0000�\u0001\u0000\u0000Z\u0002\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000~\u0001\u0000\u0000D\u0001\u0000\u0000�\u0001\u0000\u0000�\u0002\u0000\u0000�\u0001\u0000\u0000~\u0001\u0000\u0000�\u0001\u0000\u0000p\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000a\u0001\u0000\u0000v\u0001\u0000\u0000g\u0001\u0000\u0000�\u0001\u0000\u0000f\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000\u0005\u0002\u0000\u0000o\u0001\u0000\u0000z\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000q\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000}\u0001\u0000\u0000j\u0001\u0000\u0000�\u0001\u0000\u0000a\u0001\u0000\u0000a\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000\u0014\u0002\u0000\u0000|\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000z\u0001\u0000\u0000z\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000~\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u0000D\u0001\u0000\u0000~\u0001\u0000\u0000�\u0001\u0000\u0000�\u0001\u0000\u00008\u0001\u0000\u0000:\u0001\u0000\u0000;\u0000\u0000\u0000<\u0000\u0000\u0000<\u0000\u0000\u00009\u0000\u0000\u00009\u0000\u0000\u0000>\u0000\u0000\u0000?\u0000\u0000\u0000>\u0000\u0000\u0000B\u0000\u0000\u0000E\u0000\u0000\u0000<\u0000\u0000\u00001\u0000\u0000\u00003\u0000\u0000\u00002\u0000\u0000\u00000\u0000\u0000\u00002\u0000\u0000\u00007\u0000\u0000\u0000?\u0000\u0000\u0000<\u0000\u0000\u0000B\u0000\u0000\u00009\u0000\u0000\u0000B\u0000\u0000\u0000?\u0000\u0000\u00009\u0000\u0000\u00008\u0000\u0000\u0000=\u0000\u0000\u0000<\u0000\u0000\u0000?\u0000\u0000\u00000\u0000\u0000\u0000?\u0000\u0000\u0000?\u0000\u0000\u00006\u0000\u0000\u0000?\u0000\u0000\u0000B\u0000\u0000\u0000?\u0000\u0000\u00000\u0000\u0000\u00000\u0000\u0000\u0000/\u0000\u0000\u00002\u0000\u0000\u00005\u0000\u0000\u0000>\u0000\u0000\u00006\u0000\u0000\u0000.\u0000\u0000\u0000/\u0000\u0000\u00000\u0000\u0000\u00002\u0000\u0000\u00005\u0000\u0000\u0000;\u0000\u0000\u0000?\u0000\u0000\u0000<\u0000\u0000\u0000<\u0000\u0000\u0000B\u0000\u0000\u00001\u0000\u0000\u0000>\u0000\u0000\u0000=\u0000\u0000\u0000B\u0000\u0000\u00005\u0000\u0000\u00009\u0000\u0000\u0000?\u0000\u0000\u00003\u0000\u0000\u0000B\u0000\u0000\u00004\u0000\u0000\u0000A\u0000\u0000\u0000?\u0000\u0000\u0000;\u0000\u0000\u00006\u0000\u0000\u0000<\u0000\u0000\u0000@\u0000\u0000\u0000?\u0000\u0000\u0000<\u0000\u0000\u0000<\u0000\u0000\u00000\u0000\u0000\u00002\u0000\u0000\u00003\u0000\u0000\u00002\u0000\u0000\u0000>\u0000\u0000\u0000<\u0000\u0000\u0000:\u0000\u0000\u0000;\u0000\u0000\u0000:\u0000\u0000\u0000>\u0000\u0000\u00005\u0000\u0000\u00002\u0000\u0000\u0000.\u0000\u0000\u00003\u0000\u0000\u00000\u0000\u0000\u00003\u0000\u0000\u00002\u0000\u0000\u00002\u0000\u0000\u0000=\u0000\u0000\u0000<\u0000\u0000\u0000<\u0000\u0000\u00002\u0000\u0000\u0000%\u0014\u0000\u0000\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Kallisto index files are binary files. They are generated from the reference transcriptome sequences in FASTA format, and the resulting index file is used to perform fast and memory-efficient quantification of RNA-seq data."
      ],
      "metadata": {
        "id": "O33Rl9Gcdq7x"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Inspect the first few lines of the files\n",
        "!head \"{work_dir}/genes_t2g.txt\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "X2foASIVbuzs",
        "outputId": "c5a57eb8-69c6-4811-ea51-64fd50ed0cba"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "hg19_ENST00000456328\thg19_ENSG00000223972\thg19_DDX11L1\n",
            "hg19_ENST00000515242\thg19_ENSG00000223972\thg19_DDX11L1\n",
            "hg19_ENST00000518655\thg19_ENSG00000223972\thg19_DDX11L1\n",
            "hg19_ENST00000450305\thg19_ENSG00000223972\thg19_DDX11L1\n",
            "hg19_ENST00000438504\thg19_ENSG00000227232\thg19_WASH7P\n",
            "hg19_ENST00000541675\thg19_ENSG00000227232\thg19_WASH7P\n",
            "hg19_ENST00000423562\thg19_ENSG00000227232\thg19_WASH7P\n",
            "hg19_ENST00000488147\thg19_ENSG00000227232\thg19_WASH7P\n",
            "hg19_ENST00000538476\thg19_ENSG00000227232\thg19_WASH7P\n",
            "hg19_ENST00000473358\thg19_ENSG00000243485\thg19_MIR1302-10\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%time\n",
        "# This step runs `kb` to quantify the reads. `kb` can take as input URLs where the reads are located, and will stream the data \n",
        "# to Google Colab where it is quantified as it is downloaded. This allows for quantifying very large datasets without first \n",
        "# downloading them and saving them to disk. \n",
        "!kb count -i {gd_root}/LungMk/genome.idx -g {gd_root}/LungMk/genes_t2g.txt --overwrite -t 2 -x 10xv2 {data_root}/HRR339729_f1.fastq.gz {data_root}/HRR339729_r2.fastq.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "dolul8Y92KGm",
        "outputId": "b1487292-97ab-4879-f58c-c1bced32ee34"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[2023-04-20 22:46:39,736]    INFO Generating BUS file from\n",
            "[2023-04-20 22:46:39,736]    INFO         /content/drive/MyDrive/Pate_Lab/DNP/Bioinformatics/H17_LungMk/Data_files/HRA001149/HRR339729/HRR339729_f1.fastq.gz\n",
            "[2023-04-20 22:46:39,736]    INFO         /content/drive/MyDrive/Pate_Lab/DNP/Bioinformatics/H17_LungMk/Data_files/HRA001149/HRR339729/HRR339729_r2.fastq.gz\n",
            "[2023-04-20 23:29:07,255]    INFO Sorting BUS file ./output.bus to tmp/output.s.bus\n",
            "[2023-04-20 23:29:53,793]    INFO Whitelist not provided\n",
            "[2023-04-20 23:29:53,793]    INFO Copying pre-packaged 10XV2 whitelist to .\n",
            "[2023-04-20 23:29:53,933]    INFO Inspecting BUS file tmp/output.s.bus\n",
            "[2023-04-20 23:30:15,363]    INFO Correcting BUS records in tmp/output.s.bus to tmp/output.s.c.bus with whitelist ./10xv2_whitelist.txt\n",
            "[2023-04-20 23:30:35,037]    INFO Sorting BUS file tmp/output.s.c.bus to ./output.unfiltered.bus\n",
            "[2023-04-20 23:30:37,839]    INFO Generating count matrix ./counts_unfiltered/cells_x_genes from BUS file ./output.unfiltered.bus\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!ls \"{gd_root}/LungMk\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OeAvHxCWCLep",
        "outputId": "7c967a67-c75c-434c-863f-fa268cfb50b4"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            " genes_t2g.txt\t genome.idx   GRCh38genome.idx\t'GRCh38genome.idx 2'\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "It doesn't seem like the kb-wrapper is generating bustools files"
      ],
      "metadata": {
        "id": "gjnWKDMAEVhX"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Install Kallisto "
      ],
      "metadata": {
        "id": "QbelVNKWf4X_"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!sudo apt-get remove kallisto"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "R3cNPVGCf772",
        "outputId": "3d19fb3f-ce36-4275-a7ae-89eac1415935"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "Package 'kallisto' is not installed, so not removed\n",
            "0 upgraded, 0 newly installed, 0 to remove and 24 not upgraded.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!rm /usr/local/bin/kallisto"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8dnnEF5ygAVG",
        "outputId": "78bea836-2480-4c34-86ed-f4fa2d93cd49"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "rm: cannot remove '/usr/local/bin/kallisto': No such file or directory\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://github.com/pachterlab/kallisto/releases/download/v0.46.2/kallisto_linux-v0.46.2.tar.gz\n",
        "!tar -zxvf kallisto_linux-v0.46.2.tar.gz\n",
        "!mv kallisto /usr/local/bin/"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "UI3o060-gP--",
        "outputId": "22f5ec5c-918a-42bc-85a0-606f80dca5c3"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2023-04-20 21:10:24--  https://github.com/pachterlab/kallisto/releases/download/v0.46.2/kallisto_linux-v0.46.2.tar.gz\n",
            "Resolving github.com (github.com)... 20.27.177.113\n",
            "Connecting to github.com (github.com)|20.27.177.113|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/26562905/b87f5a00-510d-11ea-9bfc-64cef1470625?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230420%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230420T211024Z&X-Amz-Expires=300&X-Amz-Signature=83ab01a49071c9744e69790c3bc555d1251de59ff3e539d7a86bd67b8dc5ce16&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=26562905&response-content-disposition=attachment%3B%20filename%3Dkallisto_linux-v0.46.2.tar.gz&response-content-type=application%2Foctet-stream [following]\n",
            "--2023-04-20 21:10:24--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/26562905/b87f5a00-510d-11ea-9bfc-64cef1470625?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230420%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230420T211024Z&X-Amz-Expires=300&X-Amz-Signature=83ab01a49071c9744e69790c3bc555d1251de59ff3e539d7a86bd67b8dc5ce16&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=26562905&response-content-disposition=attachment%3B%20filename%3Dkallisto_linux-v0.46.2.tar.gz&response-content-type=application%2Foctet-stream\n",
            "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 7303987 (7.0M) [application/octet-stream]\n",
            "Saving to: ‘kallisto_linux-v0.46.2.tar.gz’\n",
            "\n",
            "kallisto_linux-v0.4 100%[===================>]   6.96M  30.1MB/s    in 0.2s    \n",
            "\n",
            "2023-04-20 21:10:25 (30.1 MB/s) - ‘kallisto_linux-v0.46.2.tar.gz’ saved [7303987/7303987]\n",
            "\n",
            "kallisto/\n",
            "kallisto/test/\n",
            "kallisto/README.md\n",
            "kallisto/kallisto\n",
            "kallisto/license.txt\n",
            "kallisto/test/reads_1.fastq.gz\n",
            "kallisto/test/transcripts.fasta.gz\n",
            "kallisto/test/README.md\n",
            "kallisto/test/chrom.txt\n",
            "kallisto/test/Snakefile\n",
            "kallisto/test/reads_2.fastq.gz\n",
            "kallisto/test/transcripts.gtf.gz\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!cd /usr/local/bin/kallisto"
      ],
      "metadata": {
        "id": "iOD3_MGjgUcl"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!ls -al /usr/local/bin/kallisto"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "g7bVk1KggXnL",
        "outputId": "18ca7c08-dcc6-4ca2-efec-8fa3269df311"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "total 21972\n",
            "drwxr-xr-x 3  501 staff     4096 Feb 16  2020 .\n",
            "drwxr-xr-x 1 root root      4096 Apr 20 21:10 ..\n",
            "-rwxr-xr-x 1  501 staff 22477379 Feb 16  2020 kallisto\n",
            "-rw-r--r-- 1  501 staff     1357 Mar 20  2017 license.txt\n",
            "-rw-r--r-- 1  501 staff     2250 Mar 20  2017 README.md\n",
            "drwxr-xr-x 2  501 staff     4096 Apr 20 21:10 test\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!sudo apt-get install kallisto"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0HfMuI-ZgbRW",
        "outputId": "a76c4dd8-37fe-4188-d7fa-66fff7ba24c4"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  libhts3\n",
            "The following NEW packages will be installed:\n",
            "  kallisto libhts3\n",
            "0 upgraded, 2 newly installed, 0 to remove and 24 not upgraded.\n",
            "Need to get 598 kB of archives.\n",
            "After this operation, 1,637 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu focal-updates/universe amd64 libhts3 amd64 1.10.2-3ubuntu0.1 [350 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu focal/universe amd64 kallisto amd64 0.46.1+dfsg-2build1 [248 kB]\n",
            "Fetched 598 kB in 2s (351 kB/s)\n",
            "debconf: unable to initialize frontend: Dialog\n",
            "debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 76, <> line 2.)\n",
            "debconf: falling back to frontend: Readline\n",
            "debconf: unable to initialize frontend: Readline\n",
            "debconf: (This frontend requires a controlling tty.)\n",
            "debconf: falling back to frontend: Teletype\n",
            "dpkg-preconfigure: unable to re-open stdin: \n",
            "Selecting previously unselected package libhts3:amd64.\n",
            "(Reading database ... 122352 files and directories currently installed.)\n",
            "Preparing to unpack .../libhts3_1.10.2-3ubuntu0.1_amd64.deb ...\n",
            "Unpacking libhts3:amd64 (1.10.2-3ubuntu0.1) ...\n",
            "Selecting previously unselected package kallisto.\n",
            "Preparing to unpack .../kallisto_0.46.1+dfsg-2build1_amd64.deb ...\n",
            "Unpacking kallisto (0.46.1+dfsg-2build1) ...\n",
            "Setting up libhts3:amd64 (1.10.2-3ubuntu0.1) ...\n",
            "Setting up kallisto (0.46.1+dfsg-2build1) ...\n",
            "Processing triggers for libc-bin (2.31-0ubuntu9.9) ...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!git clone https://github.com/BUStools/bustools.git"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "_KErYagBnPWk",
        "outputId": "6cd68bc9-1626-444b-b638-dc789baa4583"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Cloning into 'bustools'...\n",
            "remote: Enumerating objects: 3419, done.\u001b[K\n",
            "remote: Counting objects: 100% (631/631), done.\u001b[K\n",
            "remote: Compressing objects: 100% (224/224), done.\u001b[K\n",
            "remote: Total 3419 (delta 472), reused 473 (delta 405), pack-reused 2788\u001b[K\n",
            "Receiving objects: 100% (3419/3419), 3.36 MiB | 10.04 MiB/s, done.\n",
            "Resolving deltas: 100% (1351/1351), done.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!cd bustools\n",
        "# !mkdir build\n",
        "!cp bustools/CMakeLists.txt bustools/build/.\n",
        "!cd bustools/build & cmake .. -S bustools & make"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "EvLTXehancul",
        "outputId": "e2c9d1a2-c304-43cf-b51a-2ebd32fc653b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "cp: cannot create regular file 'bustools/build/.': No such file or directory\n",
            "/bin/bash: line 0: cd: bustools/build: No such file or directory\n",
            "make: *** No targets specified and no makefile found.  Stop.\n",
            "\u001b[33mCMake Warning:\n",
            "  Ignoring extra path from command line:\n",
            "\n",
            "   \"/\"\n",
            "\n",
            "\u001b[0m\n",
            "\u001b[33mCMake Warning:\n",
            "  Ignoring extra path from command line:\n",
            "\n",
            "   \"..\"\n",
            "\n",
            "\u001b[0m\n",
            "-- The C compiler identification is GNU 9.4.0\n",
            "-- The CXX compiler identification is GNU 9.4.0\n",
            "-- Detecting C compiler ABI info\n",
            "-- Detecting C compiler ABI info - done\n",
            "-- Check for working C compiler: /usr/bin/cc - skipped\n",
            "-- Detecting C compile features\n",
            "-- Detecting C compile features - done\n",
            "-- Detecting CXX compiler ABI info\n",
            "-- Detecting CXX compiler ABI info - done\n",
            "-- Check for working CXX compiler: /usr/bin/c++ - skipped\n",
            "-- Detecting CXX compile features\n",
            "-- Detecting CXX compile features - done\n",
            "\u001b[0mrelease mode\u001b[0m\n",
            "-- Performing Test CMAKE_HAVE_LIBC_PTHREAD\n",
            "-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed\n",
            "-- Looking for pthread_create in pthreads\n",
            "-- Looking for pthread_create in pthreads - not found\n",
            "-- Looking for pthread_create in pthread\n",
            "-- Looking for pthread_create in pthread - found\n",
            "-- Found Threads: TRUE  \n",
            "-- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version \"1.2.11\") \n",
            "-- Configuring done\n",
            "-- Generating done\n",
            "-- Build files have been written to: /content\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!make install"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ovL0qz9cnhin",
        "outputId": "540f4978-5120-4b0a-bdb5-6cab8c3f1fc5"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[  4%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/BUSData.cpp.o\u001b[0m\n",
            "[  8%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/Common.cpp.o\u001b[0m\n",
            "[ 12%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_capture.cpp.o\u001b[0m\n",
            "[ 16%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_clusterhist.cpp.o\u001b[0m\n",
            "[ 20%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_collapse.cpp.o\u001b[0m\n",
            "[ 24%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_compress.cpp.o\u001b[0m\n",
            "[ 28%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_correct.cpp.o\u001b[0m\n",
            "[ 32%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_count.cpp.o\u001b[0m\n",
            "[ 36%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_decompress.cpp.o\u001b[0m\n",
            "[ 40%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_extract.cpp.o\u001b[0m\n",
            "[ 44%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_inspect.cpp.o\u001b[0m\n",
            "[ 48%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_linker.cpp.o\u001b[0m\n",
            "[ 52%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_main.cpp.o\u001b[0m\n",
            "[ 56%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_mash.cpp.o\u001b[0m\n",
            "[ 60%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_merge.cpp.o\u001b[0m\n",
            "[ 64%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_predict.cpp.o\u001b[0m\n",
            "[ 68%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_project.cpp.o\u001b[0m\n",
            "[ 72%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_sort.cpp.o\u001b[0m\n",
            "[ 76%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_text.cpp.o\u001b[0m\n",
            "[ 80%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_umicorrect.cpp.o\u001b[0m\n",
            "[ 84%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools_core.dir/bustools_whitelist.cpp.o\u001b[0m\n",
            "[ 88%] \u001b[32mBuilding C object src/CMakeFiles/bustools_core.dir/roaring.c.o\u001b[0m\n",
            "[ 92%] \u001b[32m\u001b[1mLinking CXX static library libbustools_core.a\u001b[0m\n",
            "[ 92%] Built target bustools_core\n",
            "[ 96%] \u001b[32mBuilding CXX object src/CMakeFiles/bustools.dir/bustools_main.cpp.o\u001b[0m\n",
            "[100%] \u001b[32m\u001b[1mLinking CXX executable bustools\u001b[0m\n",
            "[100%] Built target bustools\n",
            "\u001b[36mInstall the project...\u001b[0m\n",
            "-- Install configuration: \"Release\"\n",
            "-- Installing: /usr/local/bin/bustools\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%env PATH=/usr/bin:/usr/local/bin:$PATH:/opt/longranger-2.2.2\n",
        "!echo $PATH"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Ff9xDaMTnyBL",
        "outputId": "13dd7f01-d79e-41bb-d34f-9af3c25ed268"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "env: PATH=/usr/bin:/usr/local/bin:$PATH:/opt/longranger-2.2.2\n",
            "/usr/bin:/usr/local/bin:$PATH:/opt/longranger-2.2.2\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!bustools"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "35y14QI0n1yV",
        "outputId": "21f50f12-9632-4a8c-d5cb-6654f98ff268"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "bustools 0.42.0\n",
            "\n",
            "Usage: bustools <CMD> [arguments] ..\n",
            "\n",
            "Where <CMD> can be one of: \n",
            "\n",
            "sort            Sort a BUS file by barcodes and UMIs\n",
            "correct         Error correct a BUS file\n",
            "umicorrect      Error correct the UMIs in a BUS file\n",
            "count           Generate count matrices from a BUS file\n",
            "inspect         Produce a report summarizing a BUS file\n",
            "whitelist       Generate a whitelist from a BUS file\n",
            "project         Project a BUS file to gene sets\n",
            "capture         Capture records from a BUS file\n",
            "merge           Merge bus files from same experiment\n",
            "text            Convert a binary BUS file to a tab-delimited text file\n",
            "extract         Extract FASTQ reads correspnding to reads in BUS file\n",
            "predict         Correct the count matrix using prediction of unseen species\n",
            "collapse        Turn BUS files into a BUG file\n",
            "clusterhist     Create UMI histograms per cluster\n",
            "linker          Remove section of barcodes in BUS files\n",
            "compress        Compress a BUS file\n",
            "inflate         Decompress a BUSZ (compressed BUS) file\n",
            "version         Prints version number\n",
            "cite            Prints citation information\n",
            "\n",
            "Running bustools <CMD> without arguments prints usage information for <CMD>\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!wget -c https://repo.anaconda.com/archive/Anaconda3-2021.11-Linux-x86_64.sh\n",
        "!chmod +x Anaconda3-2021.11-Linux-x86_64.sh\n",
        "!bash ./Anaconda3-2021.11-Linux-x86_64.sh -b -f -p /usr/local"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "iWIMzLvjoCSr",
        "outputId": "f08a2a3b-ac79-4b8d-e31a-7330d9b6eca1"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2023-04-20 21:44:24--  https://repo.anaconda.com/archive/Anaconda3-2021.11-Linux-x86_64.sh\n",
            "Resolving repo.anaconda.com (repo.anaconda.com)... 104.16.131.3, 104.16.130.3, 2606:4700::6810:8303, ...\n",
            "Connecting to repo.anaconda.com (repo.anaconda.com)|104.16.131.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 608680744 (580M) [application/x-sh]\n",
            "Saving to: ‘Anaconda3-2021.11-Linux-x86_64.sh’\n",
            "\n",
            "Anaconda3-2021.11-L 100%[===================>] 580.48M   186MB/s    in 3.1s    \n",
            "\n",
            "2023-04-20 21:44:27 (186 MB/s) - ‘Anaconda3-2021.11-Linux-x86_64.sh’ saved [608680744/608680744]\n",
            "\n",
            "PREFIX=/usr/local\n",
            "Unpacking payload ...\n",
            "Collecting package metadata (current_repodata.json): - \b\b\\ \b\b| \b\bdone\n",
            "Solving environment: - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local\n",
            "\n",
            "  added / updated specs:\n",
            "    - _ipyw_jlab_nb_ext_conf==0.1.0=py39h06a4308_0\n",
            "    - _libgcc_mutex==0.1=main\n",
            "    - _openmp_mutex==4.5=1_gnu\n",
            "    - alabaster==0.7.12=pyhd3eb1b0_0\n",
            "    - anaconda-client==1.9.0=py39h06a4308_0\n",
            "    - anaconda-navigator==2.1.1=py39_0\n",
            "    - anaconda-project==0.10.1=pyhd3eb1b0_0\n",
            "    - anaconda==2021.11=py39_0\n",
            "    - anyio==2.2.0=py39h06a4308_1\n",
            "    - appdirs==1.4.4=pyhd3eb1b0_0\n",
            "    - argh==0.26.2=py39h06a4308_0\n",
            "    - argon2-cffi==20.1.0=py39h27cfd23_1\n",
            "    - arrow==0.13.1=py39h06a4308_0\n",
            "    - asn1crypto==1.4.0=py_0\n",
            "    - astroid==2.6.6=py39h06a4308_0\n",
            "    - astropy==4.3.1=py39h09021b7_0\n",
            "    - async_generator==1.10=pyhd3eb1b0_0\n",
            "    - atomicwrites==1.4.0=py_0\n",
            "    - attrs==21.2.0=pyhd3eb1b0_0\n",
            "    - autopep8==1.5.7=pyhd3eb1b0_0\n",
            "    - babel==2.9.1=pyhd3eb1b0_0\n",
            "    - backcall==0.2.0=pyhd3eb1b0_0\n",
            "    - backports.functools_lru_cache==1.6.4=pyhd3eb1b0_0\n",
            "    - backports.shutil_get_terminal_size==1.0.0=pyhd3eb1b0_3\n",
            "    - backports.tempfile==1.0=pyhd3eb1b0_1\n",
            "    - backports.weakref==1.0.post1=py_1\n",
            "    - backports==1.0=pyhd3eb1b0_2\n",
            "    - beautifulsoup4==4.10.0=pyh06a4308_0\n",
            "    - binaryornot==0.4.4=pyhd3eb1b0_1\n",
            "    - bitarray==2.3.0=py39h7f8727e_1\n",
            "    - bkcharts==0.2=py39h06a4308_0\n",
            "    - black==19.10b0=py_0\n",
            "    - blas==1.0=mkl\n",
            "    - bleach==4.0.0=pyhd3eb1b0_0\n",
            "    - blosc==1.21.0=h8c45485_0\n",
            "    - bokeh==2.4.1=py39h06a4308_0\n",
            "    - boto==2.49.0=py39h06a4308_0\n",
            "    - bottleneck==1.3.2=py39hdd57654_1\n",
            "    - brotli==1.0.9=he6710b0_2\n",
            "    - brotlipy==0.7.0=py39h27cfd23_1003\n",
            "    - brunsli==0.1=h2531618_0\n",
            "    - bzip2==1.0.8=h7b6447c_0\n",
            "    - c-ares==1.17.1=h27cfd23_0\n",
            "    - ca-certificates==2021.10.26=h06a4308_2\n",
            "    - cached-property==1.5.2=py_0\n",
            "    - cairo==1.16.0=hf32fb01_1\n",
            "    - certifi==2021.10.8=py39h06a4308_0\n",
            "    - cffi==1.14.6=py39h400218f_0\n",
            "    - cfitsio==3.470=hf0d0db6_6\n",
            "    - chardet==4.0.0=py39h06a4308_1003\n",
            "    - charls==2.2.0=h2531618_0\n",
            "    - charset-normalizer==2.0.4=pyhd3eb1b0_0\n",
            "    - click==8.0.3=pyhd3eb1b0_0\n",
            "    - cloudpickle==2.0.0=pyhd3eb1b0_0\n",
            "    - clyent==1.2.2=py39h06a4308_1\n",
            "    - colorama==0.4.4=pyhd3eb1b0_0\n",
            "    - conda-build==3.21.5=py39h06a4308_0\n",
            "    - conda-content-trust==0.1.1=pyhd3eb1b0_0\n",
            "    - conda-env==2.6.0=1\n",
            "    - conda-pack==0.6.0=pyhd3eb1b0_0\n",
            "    - conda-package-handling==1.7.3=py39h27cfd23_1\n",
            "    - conda-repo-cli==1.0.4=pyhd3eb1b0_0\n",
            "    - conda-token==0.3.0=pyhd3eb1b0_0\n",
            "    - conda-verify==3.4.2=py_1\n",
            "    - conda==4.10.3=py39h06a4308_0\n",
            "    - contextlib2==0.6.0.post1=pyhd3eb1b0_0\n",
            "    - cookiecutter==1.7.2=pyhd3eb1b0_0\n",
            "    - cryptography==3.4.8=py39hd23ed53_0\n",
            "    - curl==7.78.0=h1ccaba5_0\n",
            "    - cycler==0.10.0=py39h06a4308_0\n",
            "    - cython==0.29.24=py39hdbfa776_0\n",
            "    - cytoolz==0.11.0=py39h27cfd23_0\n",
            "    - daal4py==2021.3.0=py39hae6d005_0\n",
            "    - dal==2021.3.0=h06a4308_557\n",
            "    - dask-core==2021.10.0=pyhd3eb1b0_0\n",
            "    - dask==2021.10.0=pyhd3eb1b0_0\n",
            "    - dataclasses==0.8=pyh6d0b6a4_7\n",
            "    - dbus==1.13.18=hb2f20db_0\n",
            "    - debugpy==1.4.1=py39h295c915_0\n",
            "    - decorator==5.1.0=pyhd3eb1b0_0\n",
            "    - defusedxml==0.7.1=pyhd3eb1b0_0\n",
            "    - diff-match-patch==20200713=pyhd3eb1b0_0\n",
            "    - distributed==2021.10.0=py39h06a4308_0\n",
            "    - docutils==0.17.1=py39h06a4308_1\n",
            "    - entrypoints==0.3=py39h06a4308_0\n",
            "    - et_xmlfile==1.1.0=py39h06a4308_0\n",
            "    - expat==2.4.1=h2531618_2\n",
            "    - fastcache==1.1.0=py39he8ac12f_0\n",
            "    - filelock==3.3.1=pyhd3eb1b0_1\n",
            "    - flake8==3.9.2=pyhd3eb1b0_0\n",
            "    - flask==1.1.2=pyhd3eb1b0_0\n",
            "    - fontconfig==2.13.1=h6c09931_0\n",
            "    - fonttools==4.25.0=pyhd3eb1b0_0\n",
            "    - freetype==2.10.4=h5ab3b9f_0\n",
            "    - fribidi==1.0.10=h7b6447c_0\n",
            "    - fsspec==2021.8.1=pyhd3eb1b0_0\n",
            "    - future==0.18.2=py39h06a4308_1\n",
            "    - get_terminal_size==1.0.0=haa9412d_0\n",
            "    - gevent==21.8.0=py39h7f8727e_1\n",
            "    - giflib==5.2.1=h7b6447c_0\n",
            "    - glib==2.69.1=h5202010_0\n",
            "    - glob2==0.7=pyhd3eb1b0_0\n",
            "    - gmp==6.2.1=h2531618_2\n",
            "    - gmpy2==2.0.8=py39h8083e48_3\n",
            "    - graphite2==1.3.14=h23475e2_0\n",
            "    - greenlet==1.1.1=py39h295c915_0\n",
            "    - gst-plugins-base==1.14.0=h8213a91_2\n",
            "    - gstreamer==1.14.0=h28cd5cc_2\n",
            "    - h5py==3.3.0=py39h930cdd6_0\n",
            "    - harfbuzz==2.8.1=h6f93f22_0\n",
            "    - hdf5==1.10.6=hb1b8bf9_0\n",
            "    - heapdict==1.0.1=pyhd3eb1b0_0\n",
            "    - html5lib==1.1=pyhd3eb1b0_0\n",
            "    - icu==58.2=he6710b0_3\n",
            "    - idna==3.2=pyhd3eb1b0_0\n",
            "    - imagecodecs==2021.8.26=py39h4cda21f_0\n",
            "    - imageio==2.9.0=pyhd3eb1b0_0\n",
            "    - imagesize==1.2.0=pyhd3eb1b0_0\n",
            "    - importlib-metadata==4.8.1=py39h06a4308_0\n",
            "    - importlib_metadata==4.8.1=hd3eb1b0_0\n",
            "    - inflection==0.5.1=py39h06a4308_0\n",
            "    - iniconfig==1.1.1=pyhd3eb1b0_0\n",
            "    - intel-openmp==2021.4.0=h06a4308_3561\n",
            "    - intervaltree==3.1.0=pyhd3eb1b0_0\n",
            "    - ipykernel==6.4.1=py39h06a4308_1\n",
            "    - ipython==7.29.0=py39hb070fc8_0\n",
            "    - ipython_genutils==0.2.0=pyhd3eb1b0_1\n",
            "    - ipywidgets==7.6.5=pyhd3eb1b0_1\n",
            "    - isort==5.9.3=pyhd3eb1b0_0\n",
            "    - itsdangerous==2.0.1=pyhd3eb1b0_0\n",
            "    - jbig==2.1=hdba287a_0\n",
            "    - jdcal==1.4.1=pyhd3eb1b0_0\n",
            "    - jedi==0.18.0=py39h06a4308_1\n",
            "    - jeepney==0.7.1=pyhd3eb1b0_0\n",
            "    - jinja2-time==0.2.0=pyhd3eb1b0_2\n",
            "    - jinja2==2.11.3=pyhd3eb1b0_0\n",
            "    - joblib==1.1.0=pyhd3eb1b0_0\n",
            "    - jpeg==9d=h7f8727e_0\n",
            "    - json5==0.9.6=pyhd3eb1b0_0\n",
            "    - jsonschema==3.2.0=pyhd3eb1b0_2\n",
            "    - jupyter==1.0.0=py39h06a4308_7\n",
            "    - jupyter_client==6.1.12=pyhd3eb1b0_0\n",
            "    - jupyter_console==6.4.0=pyhd3eb1b0_0\n",
            "    - jupyter_core==4.8.1=py39h06a4308_0\n",
            "    - jupyter_server==1.4.1=py39h06a4308_0\n",
            "    - jupyterlab==3.2.1=pyhd3eb1b0_1\n",
            "    - jupyterlab_pygments==0.1.2=py_0\n",
            "    - jupyterlab_server==2.8.2=pyhd3eb1b0_0\n",
            "    - jupyterlab_widgets==1.0.0=pyhd3eb1b0_1\n",
            "    - jxrlib==1.1=h7b6447c_2\n",
            "    - keyring==23.1.0=py39h06a4308_0\n",
            "    - kiwisolver==1.3.1=py39h2531618_0\n",
            "    - krb5==1.19.2=hac12032_0\n",
            "    - lazy-object-proxy==1.6.0=py39h27cfd23_0\n",
            "    - lcms2==2.12=h3be6417_0\n",
            "    - ld_impl_linux-64==2.35.1=h7274673_9\n",
            "    - lerc==3.0=h295c915_0\n",
            "    - libaec==1.0.4=he6710b0_1\n",
            "    - libarchive==3.4.2=h62408e4_0\n",
            "    - libcurl==7.78.0=h0b77cf5_0\n",
            "    - libdeflate==1.8=h7f8727e_5\n",
            "    - libedit==3.1.20210910=h7f8727e_0\n",
            "    - libev==4.33=h7f8727e_1\n",
            "    - libffi==3.3=he6710b0_2\n",
            "    - libgcc-ng==9.3.0=h5101ec6_17\n",
            "    - libgfortran-ng==7.5.0=ha8ba4b0_17\n",
            "    - libgfortran4==7.5.0=ha8ba4b0_17\n",
            "    - libgomp==9.3.0=h5101ec6_17\n",
            "    - liblief==0.10.1=h2531618_1\n",
            "    - libllvm11==11.1.0=h3826bc1_0\n",
            "    - libnghttp2==1.41.0=hf8bcb03_2\n",
            "    - libpng==1.6.37=hbc83047_0\n",
            "    - libsodium==1.0.18=h7b6447c_0\n",
            "    - libspatialindex==1.9.3=h2531618_0\n",
            "    - libssh2==1.9.0=h1ba5d50_1\n",
            "    - libstdcxx-ng==9.3.0=hd4cf53a_17\n",
            "    - libtiff==4.2.0=h85742a9_0\n",
            "    - libtool==2.4.6=h7b6447c_1005\n",
            "    - libuuid==1.0.3=h7f8727e_2\n",
            "    - libuv==1.40.0=h7b6447c_0\n",
            "    - libwebp-base==1.2.0=h27cfd23_0\n",
            "    - libwebp==1.2.0=h89dd481_0\n",
            "    - libxcb==1.14=h7b6447c_0\n",
            "    - libxml2==2.9.12=h03d6c58_0\n",
            "    - libxslt==1.1.34=hc22bd24_0\n",
            "    - libzopfli==1.0.3=he6710b0_0\n",
            "    - llvmlite==0.37.0=py39h295c915_1\n",
            "    - locket==0.2.1=py39h06a4308_1\n",
            "    - lxml==4.6.3=py39h9120a33_0\n",
            "    - lz4-c==1.9.3=h295c915_1\n",
            "    - lzo==2.10=h7b6447c_2\n",
            "    - markupsafe==1.1.1=py39h27cfd23_0\n",
            "    - matplotlib-base==3.4.3=py39hbbc1b5f_0\n",
            "    - matplotlib-inline==0.1.2=pyhd3eb1b0_2\n",
            "    - matplotlib==3.4.3=py39h06a4308_0\n",
            "    - mccabe==0.6.1=py39h06a4308_1\n",
            "    - mistune==0.8.4=py39h27cfd23_1000\n",
            "    - mkl-service==2.4.0=py39h7f8727e_0\n",
            "    - mkl==2021.4.0=h06a4308_640\n",
            "    - mkl_fft==1.3.1=py39hd3c417c_0\n",
            "    - mkl_random==1.2.2=py39h51133e4_0\n",
            "    - mock==4.0.3=pyhd3eb1b0_0\n",
            "    - more-itertools==8.10.0=pyhd3eb1b0_0\n",
            "    - mpc==1.1.0=h10f8cd9_1\n",
            "    - mpfr==4.0.2=hb69a4c5_1\n",
            "    - mpi==1.0=mpich\n",
            "    - mpich==3.3.2=hc856adb_0\n",
            "    - mpmath==1.2.1=py39h06a4308_0\n",
            "    - msgpack-python==1.0.2=py39hff7bd54_1\n",
            "    - multipledispatch==0.6.0=py39h06a4308_0\n",
            "    - munkres==1.1.4=py_0\n",
            "    - mypy_extensions==0.4.3=py39h06a4308_0\n",
            "    - navigator-updater==0.2.1=py39h06a4308_0\n",
            "    - nbclassic==0.2.6=pyhd3eb1b0_0\n",
            "    - nbclient==0.5.3=pyhd3eb1b0_0\n",
            "    - nbconvert==6.1.0=py39h06a4308_0\n",
            "    - nbformat==5.1.3=pyhd3eb1b0_0\n",
            "    - ncurses==6.3=heee7806_1\n",
            "    - nest-asyncio==1.5.1=pyhd3eb1b0_0\n",
            "    - networkx==2.6.3=pyhd3eb1b0_0\n",
            "    - nltk==3.6.5=pyhd3eb1b0_0\n",
            "    - nose==1.3.7=pyhd3eb1b0_1006\n",
            "    - notebook==6.4.5=py39h06a4308_0\n",
            "    - numba==0.54.1=py39h51133e4_0\n",
            "    - numexpr==2.7.3=py39h22e1b3c_1\n",
            "    - numpy-base==1.20.3=py39h74d4b33_0\n",
            "    - numpy==1.20.3=py39hf144106_0\n",
            "    - numpydoc==1.1.0=pyhd3eb1b0_1\n",
            "    - olefile==0.46=pyhd3eb1b0_0\n",
            "    - openjpeg==2.4.0=h3ad879b_0\n",
            "    - openpyxl==3.0.9=pyhd3eb1b0_0\n",
            "    - openssl==1.1.1l=h7f8727e_0\n",
            "    - packaging==21.0=pyhd3eb1b0_0\n",
            "    - pandas==1.3.4=py39h8c16a72_0\n",
            "    - pandocfilters==1.4.3=py39h06a4308_1\n",
            "    - pango==1.45.3=hd140c19_0\n",
            "    - parso==0.8.2=pyhd3eb1b0_0\n",
            "    - partd==1.2.0=pyhd3eb1b0_0\n",
            "    - patchelf==0.13=h295c915_0\n",
            "    - path.py==12.5.0=hd3eb1b0_0\n",
            "    - path==16.0.0=py39h06a4308_0\n",
            "    - pathlib2==2.3.6=py39h06a4308_2\n",
            "    - pathspec==0.7.0=py_0\n",
            "    - patsy==0.5.2=py39h06a4308_0\n",
            "    - pcre==8.45=h295c915_0\n",
            "    - pep8==1.7.1=py39h06a4308_0\n",
            "    - pexpect==4.8.0=pyhd3eb1b0_3\n",
            "    - pickleshare==0.7.5=pyhd3eb1b0_1003\n",
            "    - pillow==8.4.0=py39h5aabda8_0\n",
            "    - pip==21.2.4=py39h06a4308_0\n",
            "    - pixman==0.40.0=h7f8727e_1\n",
            "    - pkginfo==1.7.1=py39h06a4308_0\n",
            "    - pluggy==0.13.1=py39h06a4308_0\n",
            "    - ply==3.11=py39h06a4308_0\n",
            "    - poyo==0.5.0=pyhd3eb1b0_0\n",
            "    - prometheus_client==0.11.0=pyhd3eb1b0_0\n",
            "    - prompt-toolkit==3.0.20=pyhd3eb1b0_0\n",
            "    - prompt_toolkit==3.0.20=hd3eb1b0_0\n",
            "    - psutil==5.8.0=py39h27cfd23_1\n",
            "    - ptyprocess==0.7.0=pyhd3eb1b0_2\n",
            "    - py-lief==0.10.1=py39h2531618_1\n",
            "    - py==1.10.0=pyhd3eb1b0_0\n",
            "    - pycodestyle==2.7.0=pyhd3eb1b0_0\n",
            "    - pycosat==0.6.3=py39h27cfd23_0\n",
            "    - pycparser==2.20=py_2\n",
            "    - pycurl==7.44.1=py39h8f2d780_1\n",
            "    - pydocstyle==6.1.1=pyhd3eb1b0_0\n",
            "    - pyerfa==2.0.0=py39h27cfd23_0\n",
            "    - pyflakes==2.3.1=pyhd3eb1b0_0\n",
            "    - pygments==2.10.0=pyhd3eb1b0_0\n",
            "    - pyjwt==2.1.0=py39h06a4308_0\n",
            "    - pylint==2.9.6=py39h06a4308_1\n",
            "    - pyls-spyder==0.4.0=pyhd3eb1b0_0\n",
            "    - pyodbc==4.0.31=py39h295c915_0\n",
            "    - pyopenssl==21.0.0=pyhd3eb1b0_1\n",
            "    - pyparsing==3.0.4=pyhd3eb1b0_0\n",
            "    - pyqt==5.9.2=py39h2531618_6\n",
            "    - pyrsistent==0.18.0=py39heee7806_0\n",
            "    - pysocks==1.7.1=py39h06a4308_0\n",
            "    - pytables==3.6.1=py39h77479fe_1\n",
            "    - pytest==6.2.4=py39h06a4308_2\n",
            "    - python-dateutil==2.8.2=pyhd3eb1b0_0\n",
            "    - python-libarchive-c==2.9=pyhd3eb1b0_1\n",
            "    - python-lsp-black==1.0.0=pyhd3eb1b0_0\n",
            "    - python-lsp-jsonrpc==1.0.0=pyhd3eb1b0_0\n",
            "    - python-lsp-server==1.2.4=pyhd3eb1b0_0\n",
            "    - python-slugify==5.0.2=pyhd3eb1b0_0\n",
            "    - python==3.9.7=h12debd9_1\n",
            "    - pytz==2021.3=pyhd3eb1b0_0\n",
            "    - pywavelets==1.1.1=py39h6323ea4_4\n",
            "    - pyxdg==0.27=pyhd3eb1b0_0\n",
            "    - pyyaml==6.0=py39h7f8727e_1\n",
            "    - pyzmq==22.2.1=py39h295c915_1\n",
            "    - qdarkstyle==3.0.2=pyhd3eb1b0_0\n",
            "    - qstylizer==0.1.10=pyhd3eb1b0_0\n",
            "    - qt==5.9.7=h5867ecd_1\n",
            "    - qtawesome==1.0.2=pyhd3eb1b0_0\n",
            "    - qtconsole==5.1.1=pyhd3eb1b0_0\n",
            "    - qtpy==1.10.0=pyhd3eb1b0_0\n",
            "    - readline==8.1=h27cfd23_0\n",
            "    - regex==2021.8.3=py39h7f8727e_0\n",
            "    - requests==2.26.0=pyhd3eb1b0_0\n",
            "    - ripgrep==12.1.1=0\n",
            "    - rope==0.19.0=pyhd3eb1b0_0\n",
            "    - rtree==0.9.7=py39h06a4308_1\n",
            "    - ruamel_yaml==0.15.100=py39h27cfd23_0\n",
            "    - scikit-image==0.18.3=py39h51133e4_0\n",
            "    - scikit-learn-intelex==2021.3.0=py39h06a4308_0\n",
            "    - scikit-learn==0.24.2=py39ha9443f7_0\n",
            "    - scipy==1.7.1=py39h292c36d_2\n",
            "    - seaborn==0.11.2=pyhd3eb1b0_0\n",
            "    - secretstorage==3.3.1=py39h06a4308_0\n",
            "    - send2trash==1.8.0=pyhd3eb1b0_1\n",
            "    - setuptools==58.0.4=py39h06a4308_0\n",
            "    - simplegeneric==0.8.1=py39h06a4308_2\n",
            "    - singledispatch==3.7.0=pyhd3eb1b0_1001\n",
            "    - sip==4.19.13=py39h2531618_0\n",
            "    - six==1.16.0=pyhd3eb1b0_0\n",
            "    - snappy==1.1.8=he6710b0_0\n",
            "    - sniffio==1.2.0=py39h06a4308_1\n",
            "    - snowballstemmer==2.1.0=pyhd3eb1b0_0\n",
            "    - sortedcollections==2.1.0=pyhd3eb1b0_0\n",
            "    - sortedcontainers==2.4.0=pyhd3eb1b0_0\n",
            "    - soupsieve==2.2.1=pyhd3eb1b0_0\n",
            "    - sphinx==4.2.0=pyhd3eb1b0_1\n",
            "    - sphinxcontrib-applehelp==1.0.2=pyhd3eb1b0_0\n",
            "    - sphinxcontrib-devhelp==1.0.2=pyhd3eb1b0_0\n",
            "    - sphinxcontrib-htmlhelp==2.0.0=pyhd3eb1b0_0\n",
            "    - sphinxcontrib-jsmath==1.0.1=pyhd3eb1b0_0\n",
            "    - sphinxcontrib-qthelp==1.0.3=pyhd3eb1b0_0\n",
            "    - sphinxcontrib-serializinghtml==1.1.5=pyhd3eb1b0_0\n",
            "    - sphinxcontrib-websupport==1.2.4=py_0\n",
            "    - sphinxcontrib==1.0=py39h06a4308_1\n",
            "    - spyder-kernels==2.1.3=py39h06a4308_0\n",
            "    - spyder==5.1.5=py39h06a4308_1\n",
            "    - sqlalchemy==1.4.22=py39h7f8727e_0\n",
            "    - sqlite==3.36.0=hc218d9a_0\n",
            "    - statsmodels==0.12.2=py39h27cfd23_0\n",
            "    - sympy==1.9=py39h06a4308_0\n",
            "    - tbb4py==2021.4.0=py39hd09550d_0\n",
            "    - tbb==2021.4.0=hd09550d_0\n",
            "    - tblib==1.7.0=pyhd3eb1b0_0\n",
            "    - terminado==0.9.4=py39h06a4308_0\n",
            "    - testpath==0.5.0=pyhd3eb1b0_0\n",
            "    - text-unidecode==1.3=pyhd3eb1b0_0\n",
            "    - textdistance==4.2.1=pyhd3eb1b0_0\n",
            "    - threadpoolctl==2.2.0=pyh0d69192_0\n",
            "    - three-merge==0.1.1=pyhd3eb1b0_0\n",
            "    - tifffile==2021.7.2=pyhd3eb1b0_2\n",
            "    - tinycss==0.4=pyhd3eb1b0_1002\n",
            "    - tk==8.6.11=h1ccaba5_0\n",
            "    - toml==0.10.2=pyhd3eb1b0_0\n",
            "    - toolz==0.11.1=pyhd3eb1b0_0\n",
            "    - tornado==6.1=py39h27cfd23_0\n",
            "    - tqdm==4.62.3=pyhd3eb1b0_1\n",
            "    - traitlets==5.1.0=pyhd3eb1b0_0\n",
            "    - typed-ast==1.4.3=py39h7f8727e_1\n",
            "    - typing_extensions==3.10.0.2=pyh06a4308_0\n",
            "    - tzdata==2021e=hda174b7_0\n",
            "    - ujson==4.0.2=py39h2531618_0\n",
            "    - unicodecsv==0.14.1=py39h06a4308_0\n",
            "    - unidecode==1.2.0=pyhd3eb1b0_0\n",
            "    - unixodbc==2.3.9=h7b6447c_0\n",
            "    - urllib3==1.26.7=pyhd3eb1b0_0\n",
            "    - watchdog==2.1.3=py39h06a4308_0\n",
            "    - wcwidth==0.2.5=pyhd3eb1b0_0\n",
            "    - webencodings==0.5.1=py39h06a4308_1\n",
            "    - werkzeug==2.0.2=pyhd3eb1b0_0\n",
            "    - wheel==0.37.0=pyhd3eb1b0_1\n",
            "    - whichcraft==0.6.1=pyhd3eb1b0_0\n",
            "    - widgetsnbextension==3.5.1=py39h06a4308_0\n",
            "    - wrapt==1.12.1=py39he8ac12f_1\n",
            "    - wurlitzer==2.1.1=py39h06a4308_0\n",
            "    - xlrd==2.0.1=pyhd3eb1b0_0\n",
            "    - xlsxwriter==3.0.1=pyhd3eb1b0_0\n",
            "    - xlwt==1.3.0=py39h06a4308_0\n",
            "    - xmltodict==0.12.0=pyhd3eb1b0_0\n",
            "    - xz==5.2.5=h7b6447c_0\n",
            "    - yaml==0.2.5=h7b6447c_0\n",
            "    - yapf==0.31.0=pyhd3eb1b0_0\n",
            "    - zeromq==4.3.4=h2531618_0\n",
            "    - zfp==0.5.5=h2531618_6\n",
            "    - zict==2.0.0=pyhd3eb1b0_0\n",
            "    - zipp==3.6.0=pyhd3eb1b0_0\n",
            "    - zlib==1.2.11=h7b6447c_3\n",
            "    - zope.event==4.5.0=py39h06a4308_0\n",
            "    - zope.interface==5.4.0=py39h7f8727e_0\n",
            "    - zope==1.0=py39h06a4308_1\n",
            "    - zstd==1.4.9=haebb681_0\n",
            "\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  _ipyw_jlab_nb_ext~ pkgs/main/linux-64::_ipyw_jlab_nb_ext_conf-0.1.0-py39h06a4308_0\n",
            "  _libgcc_mutex      pkgs/main/linux-64::_libgcc_mutex-0.1-main\n",
            "  _openmp_mutex      pkgs/main/linux-64::_openmp_mutex-4.5-1_gnu\n",
            "  alabaster          pkgs/main/noarch::alabaster-0.7.12-pyhd3eb1b0_0\n",
            "  anaconda           pkgs/main/linux-64::anaconda-2021.11-py39_0\n",
            "  anaconda-client    pkgs/main/linux-64::anaconda-client-1.9.0-py39h06a4308_0\n",
            "  anaconda-navigator pkgs/main/linux-64::anaconda-navigator-2.1.1-py39_0\n",
            "  anaconda-project   pkgs/main/noarch::anaconda-project-0.10.1-pyhd3eb1b0_0\n",
            "  anyio              pkgs/main/linux-64::anyio-2.2.0-py39h06a4308_1\n",
            "  appdirs            pkgs/main/noarch::appdirs-1.4.4-pyhd3eb1b0_0\n",
            "  argh               pkgs/main/linux-64::argh-0.26.2-py39h06a4308_0\n",
            "  argon2-cffi        pkgs/main/linux-64::argon2-cffi-20.1.0-py39h27cfd23_1\n",
            "  arrow              pkgs/main/linux-64::arrow-0.13.1-py39h06a4308_0\n",
            "  asn1crypto         pkgs/main/noarch::asn1crypto-1.4.0-py_0\n",
            "  astroid            pkgs/main/linux-64::astroid-2.6.6-py39h06a4308_0\n",
            "  astropy            pkgs/main/linux-64::astropy-4.3.1-py39h09021b7_0\n",
            "  async_generator    pkgs/main/noarch::async_generator-1.10-pyhd3eb1b0_0\n",
            "  atomicwrites       pkgs/main/noarch::atomicwrites-1.4.0-py_0\n",
            "  attrs              pkgs/main/noarch::attrs-21.2.0-pyhd3eb1b0_0\n",
            "  autopep8           pkgs/main/noarch::autopep8-1.5.7-pyhd3eb1b0_0\n",
            "  babel              pkgs/main/noarch::babel-2.9.1-pyhd3eb1b0_0\n",
            "  backcall           pkgs/main/noarch::backcall-0.2.0-pyhd3eb1b0_0\n",
            "  backports          pkgs/main/noarch::backports-1.0-pyhd3eb1b0_2\n",
            "  backports.functoo~ pkgs/main/noarch::backports.functools_lru_cache-1.6.4-pyhd3eb1b0_0\n",
            "  backports.shutil_~ pkgs/main/noarch::backports.shutil_get_terminal_size-1.0.0-pyhd3eb1b0_3\n",
            "  backports.tempfile pkgs/main/noarch::backports.tempfile-1.0-pyhd3eb1b0_1\n",
            "  backports.weakref  pkgs/main/noarch::backports.weakref-1.0.post1-py_1\n",
            "  beautifulsoup4     pkgs/main/noarch::beautifulsoup4-4.10.0-pyh06a4308_0\n",
            "  binaryornot        pkgs/main/noarch::binaryornot-0.4.4-pyhd3eb1b0_1\n",
            "  bitarray           pkgs/main/linux-64::bitarray-2.3.0-py39h7f8727e_1\n",
            "  bkcharts           pkgs/main/linux-64::bkcharts-0.2-py39h06a4308_0\n",
            "  black              pkgs/main/noarch::black-19.10b0-py_0\n",
            "  blas               pkgs/main/linux-64::blas-1.0-mkl\n",
            "  bleach             pkgs/main/noarch::bleach-4.0.0-pyhd3eb1b0_0\n",
            "  blosc              pkgs/main/linux-64::blosc-1.21.0-h8c45485_0\n",
            "  bokeh              pkgs/main/linux-64::bokeh-2.4.1-py39h06a4308_0\n",
            "  boto               pkgs/main/linux-64::boto-2.49.0-py39h06a4308_0\n",
            "  bottleneck         pkgs/main/linux-64::bottleneck-1.3.2-py39hdd57654_1\n",
            "  brotli             pkgs/main/linux-64::brotli-1.0.9-he6710b0_2\n",
            "  brotlipy           pkgs/main/linux-64::brotlipy-0.7.0-py39h27cfd23_1003\n",
            "  brunsli            pkgs/main/linux-64::brunsli-0.1-h2531618_0\n",
            "  bzip2              pkgs/main/linux-64::bzip2-1.0.8-h7b6447c_0\n",
            "  c-ares             pkgs/main/linux-64::c-ares-1.17.1-h27cfd23_0\n",
            "  ca-certificates    pkgs/main/linux-64::ca-certificates-2021.10.26-h06a4308_2\n",
            "  cached-property    pkgs/main/noarch::cached-property-1.5.2-py_0\n",
            "  cairo              pkgs/main/linux-64::cairo-1.16.0-hf32fb01_1\n",
            "  certifi            pkgs/main/linux-64::certifi-2021.10.8-py39h06a4308_0\n",
            "  cffi               pkgs/main/linux-64::cffi-1.14.6-py39h400218f_0\n",
            "  cfitsio            pkgs/main/linux-64::cfitsio-3.470-hf0d0db6_6\n",
            "  chardet            pkgs/main/linux-64::chardet-4.0.0-py39h06a4308_1003\n",
            "  charls             pkgs/main/linux-64::charls-2.2.0-h2531618_0\n",
            "  charset-normalizer pkgs/main/noarch::charset-normalizer-2.0.4-pyhd3eb1b0_0\n",
            "  click              pkgs/main/noarch::click-8.0.3-pyhd3eb1b0_0\n",
            "  cloudpickle        pkgs/main/noarch::cloudpickle-2.0.0-pyhd3eb1b0_0\n",
            "  clyent             pkgs/main/linux-64::clyent-1.2.2-py39h06a4308_1\n",
            "  colorama           pkgs/main/noarch::colorama-0.4.4-pyhd3eb1b0_0\n",
            "  conda              pkgs/main/linux-64::conda-4.10.3-py39h06a4308_0\n",
            "  conda-build        pkgs/main/linux-64::conda-build-3.21.5-py39h06a4308_0\n",
            "  conda-content-tru~ pkgs/main/noarch::conda-content-trust-0.1.1-pyhd3eb1b0_0\n",
            "  conda-env          pkgs/main/linux-64::conda-env-2.6.0-1\n",
            "  conda-pack         pkgs/main/noarch::conda-pack-0.6.0-pyhd3eb1b0_0\n",
            "  conda-package-han~ pkgs/main/linux-64::conda-package-handling-1.7.3-py39h27cfd23_1\n",
            "  conda-repo-cli     pkgs/main/noarch::conda-repo-cli-1.0.4-pyhd3eb1b0_0\n",
            "  conda-token        pkgs/main/noarch::conda-token-0.3.0-pyhd3eb1b0_0\n",
            "  conda-verify       pkgs/main/noarch::conda-verify-3.4.2-py_1\n",
            "  contextlib2        pkgs/main/noarch::contextlib2-0.6.0.post1-pyhd3eb1b0_0\n",
            "  cookiecutter       pkgs/main/noarch::cookiecutter-1.7.2-pyhd3eb1b0_0\n",
            "  cryptography       pkgs/main/linux-64::cryptography-3.4.8-py39hd23ed53_0\n",
            "  curl               pkgs/main/linux-64::curl-7.78.0-h1ccaba5_0\n",
            "  cycler             pkgs/main/linux-64::cycler-0.10.0-py39h06a4308_0\n",
            "  cython             pkgs/main/linux-64::cython-0.29.24-py39hdbfa776_0\n",
            "  cytoolz            pkgs/main/linux-64::cytoolz-0.11.0-py39h27cfd23_0\n",
            "  daal4py            pkgs/main/linux-64::daal4py-2021.3.0-py39hae6d005_0\n",
            "  dal                pkgs/main/linux-64::dal-2021.3.0-h06a4308_557\n",
            "  dask               pkgs/main/noarch::dask-2021.10.0-pyhd3eb1b0_0\n",
            "  dask-core          pkgs/main/noarch::dask-core-2021.10.0-pyhd3eb1b0_0\n",
            "  dataclasses        pkgs/main/noarch::dataclasses-0.8-pyh6d0b6a4_7\n",
            "  dbus               pkgs/main/linux-64::dbus-1.13.18-hb2f20db_0\n",
            "  debugpy            pkgs/main/linux-64::debugpy-1.4.1-py39h295c915_0\n",
            "  decorator          pkgs/main/noarch::decorator-5.1.0-pyhd3eb1b0_0\n",
            "  defusedxml         pkgs/main/noarch::defusedxml-0.7.1-pyhd3eb1b0_0\n",
            "  diff-match-patch   pkgs/main/noarch::diff-match-patch-20200713-pyhd3eb1b0_0\n",
            "  distributed        pkgs/main/linux-64::distributed-2021.10.0-py39h06a4308_0\n",
            "  docutils           pkgs/main/linux-64::docutils-0.17.1-py39h06a4308_1\n",
            "  entrypoints        pkgs/main/linux-64::entrypoints-0.3-py39h06a4308_0\n",
            "  et_xmlfile         pkgs/main/linux-64::et_xmlfile-1.1.0-py39h06a4308_0\n",
            "  expat              pkgs/main/linux-64::expat-2.4.1-h2531618_2\n",
            "  fastcache          pkgs/main/linux-64::fastcache-1.1.0-py39he8ac12f_0\n",
            "  filelock           pkgs/main/noarch::filelock-3.3.1-pyhd3eb1b0_1\n",
            "  flake8             pkgs/main/noarch::flake8-3.9.2-pyhd3eb1b0_0\n",
            "  flask              pkgs/main/noarch::flask-1.1.2-pyhd3eb1b0_0\n",
            "  fontconfig         pkgs/main/linux-64::fontconfig-2.13.1-h6c09931_0\n",
            "  fonttools          pkgs/main/noarch::fonttools-4.25.0-pyhd3eb1b0_0\n",
            "  freetype           pkgs/main/linux-64::freetype-2.10.4-h5ab3b9f_0\n",
            "  fribidi            pkgs/main/linux-64::fribidi-1.0.10-h7b6447c_0\n",
            "  fsspec             pkgs/main/noarch::fsspec-2021.8.1-pyhd3eb1b0_0\n",
            "  future             pkgs/main/linux-64::future-0.18.2-py39h06a4308_1\n",
            "  get_terminal_size  pkgs/main/linux-64::get_terminal_size-1.0.0-haa9412d_0\n",
            "  gevent             pkgs/main/linux-64::gevent-21.8.0-py39h7f8727e_1\n",
            "  giflib             pkgs/main/linux-64::giflib-5.2.1-h7b6447c_0\n",
            "  glib               pkgs/main/linux-64::glib-2.69.1-h5202010_0\n",
            "  glob2              pkgs/main/noarch::glob2-0.7-pyhd3eb1b0_0\n",
            "  gmp                pkgs/main/linux-64::gmp-6.2.1-h2531618_2\n",
            "  gmpy2              pkgs/main/linux-64::gmpy2-2.0.8-py39h8083e48_3\n",
            "  graphite2          pkgs/main/linux-64::graphite2-1.3.14-h23475e2_0\n",
            "  greenlet           pkgs/main/linux-64::greenlet-1.1.1-py39h295c915_0\n",
            "  gst-plugins-base   pkgs/main/linux-64::gst-plugins-base-1.14.0-h8213a91_2\n",
            "  gstreamer          pkgs/main/linux-64::gstreamer-1.14.0-h28cd5cc_2\n",
            "  h5py               pkgs/main/linux-64::h5py-3.3.0-py39h930cdd6_0\n",
            "  harfbuzz           pkgs/main/linux-64::harfbuzz-2.8.1-h6f93f22_0\n",
            "  hdf5               pkgs/main/linux-64::hdf5-1.10.6-hb1b8bf9_0\n",
            "  heapdict           pkgs/main/noarch::heapdict-1.0.1-pyhd3eb1b0_0\n",
            "  html5lib           pkgs/main/noarch::html5lib-1.1-pyhd3eb1b0_0\n",
            "  icu                pkgs/main/linux-64::icu-58.2-he6710b0_3\n",
            "  idna               pkgs/main/noarch::idna-3.2-pyhd3eb1b0_0\n",
            "  imagecodecs        pkgs/main/linux-64::imagecodecs-2021.8.26-py39h4cda21f_0\n",
            "  imageio            pkgs/main/noarch::imageio-2.9.0-pyhd3eb1b0_0\n",
            "  imagesize          pkgs/main/noarch::imagesize-1.2.0-pyhd3eb1b0_0\n",
            "  importlib-metadata pkgs/main/linux-64::importlib-metadata-4.8.1-py39h06a4308_0\n",
            "  importlib_metadata pkgs/main/noarch::importlib_metadata-4.8.1-hd3eb1b0_0\n",
            "  inflection         pkgs/main/linux-64::inflection-0.5.1-py39h06a4308_0\n",
            "  iniconfig          pkgs/main/noarch::iniconfig-1.1.1-pyhd3eb1b0_0\n",
            "  intel-openmp       pkgs/main/linux-64::intel-openmp-2021.4.0-h06a4308_3561\n",
            "  intervaltree       pkgs/main/noarch::intervaltree-3.1.0-pyhd3eb1b0_0\n",
            "  ipykernel          pkgs/main/linux-64::ipykernel-6.4.1-py39h06a4308_1\n",
            "  ipython            pkgs/main/linux-64::ipython-7.29.0-py39hb070fc8_0\n",
            "  ipython_genutils   pkgs/main/noarch::ipython_genutils-0.2.0-pyhd3eb1b0_1\n",
            "  ipywidgets         pkgs/main/noarch::ipywidgets-7.6.5-pyhd3eb1b0_1\n",
            "  isort              pkgs/main/noarch::isort-5.9.3-pyhd3eb1b0_0\n",
            "  itsdangerous       pkgs/main/noarch::itsdangerous-2.0.1-pyhd3eb1b0_0\n",
            "  jbig               pkgs/main/linux-64::jbig-2.1-hdba287a_0\n",
            "  jdcal              pkgs/main/noarch::jdcal-1.4.1-pyhd3eb1b0_0\n",
            "  jedi               pkgs/main/linux-64::jedi-0.18.0-py39h06a4308_1\n",
            "  jeepney            pkgs/main/noarch::jeepney-0.7.1-pyhd3eb1b0_0\n",
            "  jinja2             pkgs/main/noarch::jinja2-2.11.3-pyhd3eb1b0_0\n",
            "  jinja2-time        pkgs/main/noarch::jinja2-time-0.2.0-pyhd3eb1b0_2\n",
            "  joblib             pkgs/main/noarch::joblib-1.1.0-pyhd3eb1b0_0\n",
            "  jpeg               pkgs/main/linux-64::jpeg-9d-h7f8727e_0\n",
            "  json5              pkgs/main/noarch::json5-0.9.6-pyhd3eb1b0_0\n",
            "  jsonschema         pkgs/main/noarch::jsonschema-3.2.0-pyhd3eb1b0_2\n",
            "  jupyter            pkgs/main/linux-64::jupyter-1.0.0-py39h06a4308_7\n",
            "  jupyter_client     pkgs/main/noarch::jupyter_client-6.1.12-pyhd3eb1b0_0\n",
            "  jupyter_console    pkgs/main/noarch::jupyter_console-6.4.0-pyhd3eb1b0_0\n",
            "  jupyter_core       pkgs/main/linux-64::jupyter_core-4.8.1-py39h06a4308_0\n",
            "  jupyter_server     pkgs/main/linux-64::jupyter_server-1.4.1-py39h06a4308_0\n",
            "  jupyterlab         pkgs/main/noarch::jupyterlab-3.2.1-pyhd3eb1b0_1\n",
            "  jupyterlab_pygmen~ pkgs/main/noarch::jupyterlab_pygments-0.1.2-py_0\n",
            "  jupyterlab_server  pkgs/main/noarch::jupyterlab_server-2.8.2-pyhd3eb1b0_0\n",
            "  jupyterlab_widgets pkgs/main/noarch::jupyterlab_widgets-1.0.0-pyhd3eb1b0_1\n",
            "  jxrlib             pkgs/main/linux-64::jxrlib-1.1-h7b6447c_2\n",
            "  keyring            pkgs/main/linux-64::keyring-23.1.0-py39h06a4308_0\n",
            "  kiwisolver         pkgs/main/linux-64::kiwisolver-1.3.1-py39h2531618_0\n",
            "  krb5               pkgs/main/linux-64::krb5-1.19.2-hac12032_0\n",
            "  lazy-object-proxy  pkgs/main/linux-64::lazy-object-proxy-1.6.0-py39h27cfd23_0\n",
            "  lcms2              pkgs/main/linux-64::lcms2-2.12-h3be6417_0\n",
            "  ld_impl_linux-64   pkgs/main/linux-64::ld_impl_linux-64-2.35.1-h7274673_9\n",
            "  lerc               pkgs/main/linux-64::lerc-3.0-h295c915_0\n",
            "  libaec             pkgs/main/linux-64::libaec-1.0.4-he6710b0_1\n",
            "  libarchive         pkgs/main/linux-64::libarchive-3.4.2-h62408e4_0\n",
            "  libcurl            pkgs/main/linux-64::libcurl-7.78.0-h0b77cf5_0\n",
            "  libdeflate         pkgs/main/linux-64::libdeflate-1.8-h7f8727e_5\n",
            "  libedit            pkgs/main/linux-64::libedit-3.1.20210910-h7f8727e_0\n",
            "  libev              pkgs/main/linux-64::libev-4.33-h7f8727e_1\n",
            "  libffi             pkgs/main/linux-64::libffi-3.3-he6710b0_2\n",
            "  libgcc-ng          pkgs/main/linux-64::libgcc-ng-9.3.0-h5101ec6_17\n",
            "  libgfortran-ng     pkgs/main/linux-64::libgfortran-ng-7.5.0-ha8ba4b0_17\n",
            "  libgfortran4       pkgs/main/linux-64::libgfortran4-7.5.0-ha8ba4b0_17\n",
            "  libgomp            pkgs/main/linux-64::libgomp-9.3.0-h5101ec6_17\n",
            "  liblief            pkgs/main/linux-64::liblief-0.10.1-h2531618_1\n",
            "  libllvm11          pkgs/main/linux-64::libllvm11-11.1.0-h3826bc1_0\n",
            "  libnghttp2         pkgs/main/linux-64::libnghttp2-1.41.0-hf8bcb03_2\n",
            "  libpng             pkgs/main/linux-64::libpng-1.6.37-hbc83047_0\n",
            "  libsodium          pkgs/main/linux-64::libsodium-1.0.18-h7b6447c_0\n",
            "  libspatialindex    pkgs/main/linux-64::libspatialindex-1.9.3-h2531618_0\n",
            "  libssh2            pkgs/main/linux-64::libssh2-1.9.0-h1ba5d50_1\n",
            "  libstdcxx-ng       pkgs/main/linux-64::libstdcxx-ng-9.3.0-hd4cf53a_17\n",
            "  libtiff            pkgs/main/linux-64::libtiff-4.2.0-h85742a9_0\n",
            "  libtool            pkgs/main/linux-64::libtool-2.4.6-h7b6447c_1005\n",
            "  libuuid            pkgs/main/linux-64::libuuid-1.0.3-h7f8727e_2\n",
            "  libuv              pkgs/main/linux-64::libuv-1.40.0-h7b6447c_0\n",
            "  libwebp            pkgs/main/linux-64::libwebp-1.2.0-h89dd481_0\n",
            "  libwebp-base       pkgs/main/linux-64::libwebp-base-1.2.0-h27cfd23_0\n",
            "  libxcb             pkgs/main/linux-64::libxcb-1.14-h7b6447c_0\n",
            "  libxml2            pkgs/main/linux-64::libxml2-2.9.12-h03d6c58_0\n",
            "  libxslt            pkgs/main/linux-64::libxslt-1.1.34-hc22bd24_0\n",
            "  libzopfli          pkgs/main/linux-64::libzopfli-1.0.3-he6710b0_0\n",
            "  llvmlite           pkgs/main/linux-64::llvmlite-0.37.0-py39h295c915_1\n",
            "  locket             pkgs/main/linux-64::locket-0.2.1-py39h06a4308_1\n",
            "  lxml               pkgs/main/linux-64::lxml-4.6.3-py39h9120a33_0\n",
            "  lz4-c              pkgs/main/linux-64::lz4-c-1.9.3-h295c915_1\n",
            "  lzo                pkgs/main/linux-64::lzo-2.10-h7b6447c_2\n",
            "  markupsafe         pkgs/main/linux-64::markupsafe-1.1.1-py39h27cfd23_0\n",
            "  matplotlib         pkgs/main/linux-64::matplotlib-3.4.3-py39h06a4308_0\n",
            "  matplotlib-base    pkgs/main/linux-64::matplotlib-base-3.4.3-py39hbbc1b5f_0\n",
            "  matplotlib-inline  pkgs/main/noarch::matplotlib-inline-0.1.2-pyhd3eb1b0_2\n",
            "  mccabe             pkgs/main/linux-64::mccabe-0.6.1-py39h06a4308_1\n",
            "  mistune            pkgs/main/linux-64::mistune-0.8.4-py39h27cfd23_1000\n",
            "  mkl                pkgs/main/linux-64::mkl-2021.4.0-h06a4308_640\n",
            "  mkl-service        pkgs/main/linux-64::mkl-service-2.4.0-py39h7f8727e_0\n",
            "  mkl_fft            pkgs/main/linux-64::mkl_fft-1.3.1-py39hd3c417c_0\n",
            "  mkl_random         pkgs/main/linux-64::mkl_random-1.2.2-py39h51133e4_0\n",
            "  mock               pkgs/main/noarch::mock-4.0.3-pyhd3eb1b0_0\n",
            "  more-itertools     pkgs/main/noarch::more-itertools-8.10.0-pyhd3eb1b0_0\n",
            "  mpc                pkgs/main/linux-64::mpc-1.1.0-h10f8cd9_1\n",
            "  mpfr               pkgs/main/linux-64::mpfr-4.0.2-hb69a4c5_1\n",
            "  mpi                pkgs/main/linux-64::mpi-1.0-mpich\n",
            "  mpich              pkgs/main/linux-64::mpich-3.3.2-hc856adb_0\n",
            "  mpmath             pkgs/main/linux-64::mpmath-1.2.1-py39h06a4308_0\n",
            "  msgpack-python     pkgs/main/linux-64::msgpack-python-1.0.2-py39hff7bd54_1\n",
            "  multipledispatch   pkgs/main/linux-64::multipledispatch-0.6.0-py39h06a4308_0\n",
            "  munkres            pkgs/main/noarch::munkres-1.1.4-py_0\n",
            "  mypy_extensions    pkgs/main/linux-64::mypy_extensions-0.4.3-py39h06a4308_0\n",
            "  navigator-updater  pkgs/main/linux-64::navigator-updater-0.2.1-py39h06a4308_0\n",
            "  nbclassic          pkgs/main/noarch::nbclassic-0.2.6-pyhd3eb1b0_0\n",
            "  nbclient           pkgs/main/noarch::nbclient-0.5.3-pyhd3eb1b0_0\n",
            "  nbconvert          pkgs/main/linux-64::nbconvert-6.1.0-py39h06a4308_0\n",
            "  nbformat           pkgs/main/noarch::nbformat-5.1.3-pyhd3eb1b0_0\n",
            "  ncurses            pkgs/main/linux-64::ncurses-6.3-heee7806_1\n",
            "  nest-asyncio       pkgs/main/noarch::nest-asyncio-1.5.1-pyhd3eb1b0_0\n",
            "  networkx           pkgs/main/noarch::networkx-2.6.3-pyhd3eb1b0_0\n",
            "  nltk               pkgs/main/noarch::nltk-3.6.5-pyhd3eb1b0_0\n",
            "  nose               pkgs/main/noarch::nose-1.3.7-pyhd3eb1b0_1006\n",
            "  notebook           pkgs/main/linux-64::notebook-6.4.5-py39h06a4308_0\n",
            "  numba              pkgs/main/linux-64::numba-0.54.1-py39h51133e4_0\n",
            "  numexpr            pkgs/main/linux-64::numexpr-2.7.3-py39h22e1b3c_1\n",
            "  numpy              pkgs/main/linux-64::numpy-1.20.3-py39hf144106_0\n",
            "  numpy-base         pkgs/main/linux-64::numpy-base-1.20.3-py39h74d4b33_0\n",
            "  numpydoc           pkgs/main/noarch::numpydoc-1.1.0-pyhd3eb1b0_1\n",
            "  olefile            pkgs/main/noarch::olefile-0.46-pyhd3eb1b0_0\n",
            "  openjpeg           pkgs/main/linux-64::openjpeg-2.4.0-h3ad879b_0\n",
            "  openpyxl           pkgs/main/noarch::openpyxl-3.0.9-pyhd3eb1b0_0\n",
            "  openssl            pkgs/main/linux-64::openssl-1.1.1l-h7f8727e_0\n",
            "  packaging          pkgs/main/noarch::packaging-21.0-pyhd3eb1b0_0\n",
            "  pandas             pkgs/main/linux-64::pandas-1.3.4-py39h8c16a72_0\n",
            "  pandocfilters      pkgs/main/linux-64::pandocfilters-1.4.3-py39h06a4308_1\n",
            "  pango              pkgs/main/linux-64::pango-1.45.3-hd140c19_0\n",
            "  parso              pkgs/main/noarch::parso-0.8.2-pyhd3eb1b0_0\n",
            "  partd              pkgs/main/noarch::partd-1.2.0-pyhd3eb1b0_0\n",
            "  patchelf           pkgs/main/linux-64::patchelf-0.13-h295c915_0\n",
            "  path               pkgs/main/linux-64::path-16.0.0-py39h06a4308_0\n",
            "  path.py            pkgs/main/noarch::path.py-12.5.0-hd3eb1b0_0\n",
            "  pathlib2           pkgs/main/linux-64::pathlib2-2.3.6-py39h06a4308_2\n",
            "  pathspec           pkgs/main/noarch::pathspec-0.7.0-py_0\n",
            "  patsy              pkgs/main/linux-64::patsy-0.5.2-py39h06a4308_0\n",
            "  pcre               pkgs/main/linux-64::pcre-8.45-h295c915_0\n",
            "  pep8               pkgs/main/linux-64::pep8-1.7.1-py39h06a4308_0\n",
            "  pexpect            pkgs/main/noarch::pexpect-4.8.0-pyhd3eb1b0_3\n",
            "  pickleshare        pkgs/main/noarch::pickleshare-0.7.5-pyhd3eb1b0_1003\n",
            "  pillow             pkgs/main/linux-64::pillow-8.4.0-py39h5aabda8_0\n",
            "  pip                pkgs/main/linux-64::pip-21.2.4-py39h06a4308_0\n",
            "  pixman             pkgs/main/linux-64::pixman-0.40.0-h7f8727e_1\n",
            "  pkginfo            pkgs/main/linux-64::pkginfo-1.7.1-py39h06a4308_0\n",
            "  pluggy             pkgs/main/linux-64::pluggy-0.13.1-py39h06a4308_0\n",
            "  ply                pkgs/main/linux-64::ply-3.11-py39h06a4308_0\n",
            "  poyo               pkgs/main/noarch::poyo-0.5.0-pyhd3eb1b0_0\n",
            "  prometheus_client  pkgs/main/noarch::prometheus_client-0.11.0-pyhd3eb1b0_0\n",
            "  prompt-toolkit     pkgs/main/noarch::prompt-toolkit-3.0.20-pyhd3eb1b0_0\n",
            "  prompt_toolkit     pkgs/main/noarch::prompt_toolkit-3.0.20-hd3eb1b0_0\n",
            "  psutil             pkgs/main/linux-64::psutil-5.8.0-py39h27cfd23_1\n",
            "  ptyprocess         pkgs/main/noarch::ptyprocess-0.7.0-pyhd3eb1b0_2\n",
            "  py                 pkgs/main/noarch::py-1.10.0-pyhd3eb1b0_0\n",
            "  py-lief            pkgs/main/linux-64::py-lief-0.10.1-py39h2531618_1\n",
            "  pycodestyle        pkgs/main/noarch::pycodestyle-2.7.0-pyhd3eb1b0_0\n",
            "  pycosat            pkgs/main/linux-64::pycosat-0.6.3-py39h27cfd23_0\n",
            "  pycparser          pkgs/main/noarch::pycparser-2.20-py_2\n",
            "  pycurl             pkgs/main/linux-64::pycurl-7.44.1-py39h8f2d780_1\n",
            "  pydocstyle         pkgs/main/noarch::pydocstyle-6.1.1-pyhd3eb1b0_0\n",
            "  pyerfa             pkgs/main/linux-64::pyerfa-2.0.0-py39h27cfd23_0\n",
            "  pyflakes           pkgs/main/noarch::pyflakes-2.3.1-pyhd3eb1b0_0\n",
            "  pygments           pkgs/main/noarch::pygments-2.10.0-pyhd3eb1b0_0\n",
            "  pyjwt              pkgs/main/linux-64::pyjwt-2.1.0-py39h06a4308_0\n",
            "  pylint             pkgs/main/linux-64::pylint-2.9.6-py39h06a4308_1\n",
            "  pyls-spyder        pkgs/main/noarch::pyls-spyder-0.4.0-pyhd3eb1b0_0\n",
            "  pyodbc             pkgs/main/linux-64::pyodbc-4.0.31-py39h295c915_0\n",
            "  pyopenssl          pkgs/main/noarch::pyopenssl-21.0.0-pyhd3eb1b0_1\n",
            "  pyparsing          pkgs/main/noarch::pyparsing-3.0.4-pyhd3eb1b0_0\n",
            "  pyqt               pkgs/main/linux-64::pyqt-5.9.2-py39h2531618_6\n",
            "  pyrsistent         pkgs/main/linux-64::pyrsistent-0.18.0-py39heee7806_0\n",
            "  pysocks            pkgs/main/linux-64::pysocks-1.7.1-py39h06a4308_0\n",
            "  pytables           pkgs/main/linux-64::pytables-3.6.1-py39h77479fe_1\n",
            "  pytest             pkgs/main/linux-64::pytest-6.2.4-py39h06a4308_2\n",
            "  python             pkgs/main/linux-64::python-3.9.7-h12debd9_1\n",
            "  python-dateutil    pkgs/main/noarch::python-dateutil-2.8.2-pyhd3eb1b0_0\n",
            "  python-libarchive~ pkgs/main/noarch::python-libarchive-c-2.9-pyhd3eb1b0_1\n",
            "  python-lsp-black   pkgs/main/noarch::python-lsp-black-1.0.0-pyhd3eb1b0_0\n",
            "  python-lsp-jsonrpc pkgs/main/noarch::python-lsp-jsonrpc-1.0.0-pyhd3eb1b0_0\n",
            "  python-lsp-server  pkgs/main/noarch::python-lsp-server-1.2.4-pyhd3eb1b0_0\n",
            "  python-slugify     pkgs/main/noarch::python-slugify-5.0.2-pyhd3eb1b0_0\n",
            "  pytz               pkgs/main/noarch::pytz-2021.3-pyhd3eb1b0_0\n",
            "  pywavelets         pkgs/main/linux-64::pywavelets-1.1.1-py39h6323ea4_4\n",
            "  pyxdg              pkgs/main/noarch::pyxdg-0.27-pyhd3eb1b0_0\n",
            "  pyyaml             pkgs/main/linux-64::pyyaml-6.0-py39h7f8727e_1\n",
            "  pyzmq              pkgs/main/linux-64::pyzmq-22.2.1-py39h295c915_1\n",
            "  qdarkstyle         pkgs/main/noarch::qdarkstyle-3.0.2-pyhd3eb1b0_0\n",
            "  qstylizer          pkgs/main/noarch::qstylizer-0.1.10-pyhd3eb1b0_0\n",
            "  qt                 pkgs/main/linux-64::qt-5.9.7-h5867ecd_1\n",
            "  qtawesome          pkgs/main/noarch::qtawesome-1.0.2-pyhd3eb1b0_0\n",
            "  qtconsole          pkgs/main/noarch::qtconsole-5.1.1-pyhd3eb1b0_0\n",
            "  qtpy               pkgs/main/noarch::qtpy-1.10.0-pyhd3eb1b0_0\n",
            "  readline           pkgs/main/linux-64::readline-8.1-h27cfd23_0\n",
            "  regex              pkgs/main/linux-64::regex-2021.8.3-py39h7f8727e_0\n",
            "  requests           pkgs/main/noarch::requests-2.26.0-pyhd3eb1b0_0\n",
            "  ripgrep            pkgs/main/linux-64::ripgrep-12.1.1-0\n",
            "  rope               pkgs/main/noarch::rope-0.19.0-pyhd3eb1b0_0\n",
            "  rtree              pkgs/main/linux-64::rtree-0.9.7-py39h06a4308_1\n",
            "  ruamel_yaml        pkgs/main/linux-64::ruamel_yaml-0.15.100-py39h27cfd23_0\n",
            "  scikit-image       pkgs/main/linux-64::scikit-image-0.18.3-py39h51133e4_0\n",
            "  scikit-learn       pkgs/main/linux-64::scikit-learn-0.24.2-py39ha9443f7_0\n",
            "  scikit-learn-inte~ pkgs/main/linux-64::scikit-learn-intelex-2021.3.0-py39h06a4308_0\n",
            "  scipy              pkgs/main/linux-64::scipy-1.7.1-py39h292c36d_2\n",
            "  seaborn            pkgs/main/noarch::seaborn-0.11.2-pyhd3eb1b0_0\n",
            "  secretstorage      pkgs/main/linux-64::secretstorage-3.3.1-py39h06a4308_0\n",
            "  send2trash         pkgs/main/noarch::send2trash-1.8.0-pyhd3eb1b0_1\n",
            "  setuptools         pkgs/main/linux-64::setuptools-58.0.4-py39h06a4308_0\n",
            "  simplegeneric      pkgs/main/linux-64::simplegeneric-0.8.1-py39h06a4308_2\n",
            "  singledispatch     pkgs/main/noarch::singledispatch-3.7.0-pyhd3eb1b0_1001\n",
            "  sip                pkgs/main/linux-64::sip-4.19.13-py39h2531618_0\n",
            "  six                pkgs/main/noarch::six-1.16.0-pyhd3eb1b0_0\n",
            "  snappy             pkgs/main/linux-64::snappy-1.1.8-he6710b0_0\n",
            "  sniffio            pkgs/main/linux-64::sniffio-1.2.0-py39h06a4308_1\n",
            "  snowballstemmer    pkgs/main/noarch::snowballstemmer-2.1.0-pyhd3eb1b0_0\n",
            "  sortedcollections  pkgs/main/noarch::sortedcollections-2.1.0-pyhd3eb1b0_0\n",
            "  sortedcontainers   pkgs/main/noarch::sortedcontainers-2.4.0-pyhd3eb1b0_0\n",
            "  soupsieve          pkgs/main/noarch::soupsieve-2.2.1-pyhd3eb1b0_0\n",
            "  sphinx             pkgs/main/noarch::sphinx-4.2.0-pyhd3eb1b0_1\n",
            "  sphinxcontrib      pkgs/main/linux-64::sphinxcontrib-1.0-py39h06a4308_1\n",
            "  sphinxcontrib-app~ pkgs/main/noarch::sphinxcontrib-applehelp-1.0.2-pyhd3eb1b0_0\n",
            "  sphinxcontrib-dev~ pkgs/main/noarch::sphinxcontrib-devhelp-1.0.2-pyhd3eb1b0_0\n",
            "  sphinxcontrib-htm~ pkgs/main/noarch::sphinxcontrib-htmlhelp-2.0.0-pyhd3eb1b0_0\n",
            "  sphinxcontrib-jsm~ pkgs/main/noarch::sphinxcontrib-jsmath-1.0.1-pyhd3eb1b0_0\n",
            "  sphinxcontrib-qth~ pkgs/main/noarch::sphinxcontrib-qthelp-1.0.3-pyhd3eb1b0_0\n",
            "  sphinxcontrib-ser~ pkgs/main/noarch::sphinxcontrib-serializinghtml-1.1.5-pyhd3eb1b0_0\n",
            "  sphinxcontrib-web~ pkgs/main/noarch::sphinxcontrib-websupport-1.2.4-py_0\n",
            "  spyder             pkgs/main/linux-64::spyder-5.1.5-py39h06a4308_1\n",
            "  spyder-kernels     pkgs/main/linux-64::spyder-kernels-2.1.3-py39h06a4308_0\n",
            "  sqlalchemy         pkgs/main/linux-64::sqlalchemy-1.4.22-py39h7f8727e_0\n",
            "  sqlite             pkgs/main/linux-64::sqlite-3.36.0-hc218d9a_0\n",
            "  statsmodels        pkgs/main/linux-64::statsmodels-0.12.2-py39h27cfd23_0\n",
            "  sympy              pkgs/main/linux-64::sympy-1.9-py39h06a4308_0\n",
            "  tbb                pkgs/main/linux-64::tbb-2021.4.0-hd09550d_0\n",
            "  tbb4py             pkgs/main/linux-64::tbb4py-2021.4.0-py39hd09550d_0\n",
            "  tblib              pkgs/main/noarch::tblib-1.7.0-pyhd3eb1b0_0\n",
            "  terminado          pkgs/main/linux-64::terminado-0.9.4-py39h06a4308_0\n",
            "  testpath           pkgs/main/noarch::testpath-0.5.0-pyhd3eb1b0_0\n",
            "  text-unidecode     pkgs/main/noarch::text-unidecode-1.3-pyhd3eb1b0_0\n",
            "  textdistance       pkgs/main/noarch::textdistance-4.2.1-pyhd3eb1b0_0\n",
            "  threadpoolctl      pkgs/main/noarch::threadpoolctl-2.2.0-pyh0d69192_0\n",
            "  three-merge        pkgs/main/noarch::three-merge-0.1.1-pyhd3eb1b0_0\n",
            "  tifffile           pkgs/main/noarch::tifffile-2021.7.2-pyhd3eb1b0_2\n",
            "  tinycss            pkgs/main/noarch::tinycss-0.4-pyhd3eb1b0_1002\n",
            "  tk                 pkgs/main/linux-64::tk-8.6.11-h1ccaba5_0\n",
            "  toml               pkgs/main/noarch::toml-0.10.2-pyhd3eb1b0_0\n",
            "  toolz              pkgs/main/noarch::toolz-0.11.1-pyhd3eb1b0_0\n",
            "  tornado            pkgs/main/linux-64::tornado-6.1-py39h27cfd23_0\n",
            "  tqdm               pkgs/main/noarch::tqdm-4.62.3-pyhd3eb1b0_1\n",
            "  traitlets          pkgs/main/noarch::traitlets-5.1.0-pyhd3eb1b0_0\n",
            "  typed-ast          pkgs/main/linux-64::typed-ast-1.4.3-py39h7f8727e_1\n",
            "  typing_extensions  pkgs/main/noarch::typing_extensions-3.10.0.2-pyh06a4308_0\n",
            "  tzdata             pkgs/main/noarch::tzdata-2021e-hda174b7_0\n",
            "  ujson              pkgs/main/linux-64::ujson-4.0.2-py39h2531618_0\n",
            "  unicodecsv         pkgs/main/linux-64::unicodecsv-0.14.1-py39h06a4308_0\n",
            "  unidecode          pkgs/main/noarch::unidecode-1.2.0-pyhd3eb1b0_0\n",
            "  unixodbc           pkgs/main/linux-64::unixodbc-2.3.9-h7b6447c_0\n",
            "  urllib3            pkgs/main/noarch::urllib3-1.26.7-pyhd3eb1b0_0\n",
            "  watchdog           pkgs/main/linux-64::watchdog-2.1.3-py39h06a4308_0\n",
            "  wcwidth            pkgs/main/noarch::wcwidth-0.2.5-pyhd3eb1b0_0\n",
            "  webencodings       pkgs/main/linux-64::webencodings-0.5.1-py39h06a4308_1\n",
            "  werkzeug           pkgs/main/noarch::werkzeug-2.0.2-pyhd3eb1b0_0\n",
            "  wheel              pkgs/main/noarch::wheel-0.37.0-pyhd3eb1b0_1\n",
            "  whichcraft         pkgs/main/noarch::whichcraft-0.6.1-pyhd3eb1b0_0\n",
            "  widgetsnbextension pkgs/main/linux-64::widgetsnbextension-3.5.1-py39h06a4308_0\n",
            "  wrapt              pkgs/main/linux-64::wrapt-1.12.1-py39he8ac12f_1\n",
            "  wurlitzer          pkgs/main/linux-64::wurlitzer-2.1.1-py39h06a4308_0\n",
            "  xlrd               pkgs/main/noarch::xlrd-2.0.1-pyhd3eb1b0_0\n",
            "  xlsxwriter         pkgs/main/noarch::xlsxwriter-3.0.1-pyhd3eb1b0_0\n",
            "  xlwt               pkgs/main/linux-64::xlwt-1.3.0-py39h06a4308_0\n",
            "  xmltodict          pkgs/main/noarch::xmltodict-0.12.0-pyhd3eb1b0_0\n",
            "  xz                 pkgs/main/linux-64::xz-5.2.5-h7b6447c_0\n",
            "  yaml               pkgs/main/linux-64::yaml-0.2.5-h7b6447c_0\n",
            "  yapf               pkgs/main/noarch::yapf-0.31.0-pyhd3eb1b0_0\n",
            "  zeromq             pkgs/main/linux-64::zeromq-4.3.4-h2531618_0\n",
            "  zfp                pkgs/main/linux-64::zfp-0.5.5-h2531618_6\n",
            "  zict               pkgs/main/noarch::zict-2.0.0-pyhd3eb1b0_0\n",
            "  zipp               pkgs/main/noarch::zipp-3.6.0-pyhd3eb1b0_0\n",
            "  zlib               pkgs/main/linux-64::zlib-1.2.11-h7b6447c_3\n",
            "  zope               pkgs/main/linux-64::zope-1.0-py39h06a4308_1\n",
            "  zope.event         pkgs/main/linux-64::zope.event-4.5.0-py39h06a4308_0\n",
            "  zope.interface     pkgs/main/linux-64::zope.interface-5.4.0-py39h7f8727e_0\n",
            "  zstd               pkgs/main/linux-64::zstd-1.4.9-haebb681_0\n",
            "\n",
            "\n",
            "Preparing transaction: - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "Executing transaction: | \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \n",
            "\n",
            "    Installed package of scikit-learn can be accelerated using scikit-learn-intelex.\n",
            "    More details are available here: https://intel.github.io/scikit-learn-intelex\n",
            "\n",
            "    For example:\n",
            "\n",
            "        $ conda install scikit-learn-intelex\n",
            "        $ python -m sklearnex my_application.py\n",
            "\n",
            "    \n",
            "\n",
            "\b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n",
            "installation finished.\n",
            "WARNING:\n",
            "    You currently have a PYTHONPATH environment variable set. This may cause\n",
            "    unexpected behavior when running the Python interpreter in Anaconda3.\n",
            "    For best results, please verify that your PYTHONPATH only points to\n",
            "    directories of packages that are compatible with the Python interpreter\n",
            "    in Anaconda3: /usr/local\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Google drive root\n",
        "gd_root = \"/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients\"\n",
        "# Data root\n",
        "data_root = f\"{gd_root}/Data files/HRA001149\"\n",
        "# working directory\n",
        "work_dir = f\"{gd_root}/LungMk\"\n",
        "\n",
        "# create the directory \n",
        "!mkdir -p \"{work_dir}\"\n",
        "\n",
        "# download the cDNA fasta file\n",
        "!curl -O ftp://ftp.ensembl.org/pub/release-94/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz\n",
        "\n",
        "# build the kallisto index\n",
        "!kallisto index -i Homo_sapiens.GRCh38.cdna.all.release-94_k31.idx Homo_sapiens.GRCh38.cdna.all.fa.gz\n",
        "\n",
        "# download the transcripts to genes mapping file\n",
        "!curl -O https://github.com/BUStools/getting_started/releases/download/species_mixing/transcripts_to_genes_hg19mm10.txt\n",
        "\n",
        "# move the mapping file to the working directory\n",
        "!mv transcripts_to_genes_hg19mm10.txt \"{work_dir}/genes_t2g.txt\"\n",
        "\n",
        "# run kb count with the downloaded index and t2g file\n",
        "!kb count -i Homo_sapiens.GRCh38.cdna.all.release-94_k31.idx -g \"{work_dir}/genes_t2g.txt\" --overwrite -t 2 -x 10xv2 {data_root}/HRR339729_f1.fastq.gz {data_root}/HRR339729_r2.fastq.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "-FxZUlzut-oB",
        "outputId": "c49195be-9e59-4dfa-ad5a-392415058992"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "100 64.7M  100 64.7M    0     0  5827k      0  0:00:11  0:00:11 --:--:-- 11.3M\n",
            "\n",
            "[build] loading fasta file Homo_sapiens.GRCh38.cdna.all.fa.gz\n",
            "[build] k-mer length: 31\n",
            "[build] warning: clipped off poly-A tail (longer than 10)\n",
            "        from 1471 target sequences\n",
            "[build] warning: replaced 5 non-ACGUT characters in the input sequence\n",
            "        with pseudorandom nucleotides\n",
            "[build] counting k-mers ... done.\n",
            "[build] building target de Bruijn graph ...  done \n",
            "[build] creating equivalence classes ...  done\n",
            "[build] target de Bruijn graph has 1118780 contigs and contains 108619921 k-mers \n",
            "\n",
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\n",
            "[2023-04-20 22:17:42,552]    INFO Generating BUS file from\n",
            "[2023-04-20 22:17:42,552]    INFO         /content/drive/MyDrive/Pate\n",
            "[2023-04-20 22:17:42,552]    INFO         Lab/DNP\n",
            "[2023-04-20 22:17:42,552]    INFO         Folder/Bioinformatics/H17\n",
            "[2023-04-20 22:17:42,552]    INFO         Lung\n",
            "[2023-04-20 22:17:42,552]    INFO         Mk\n",
            "[2023-04-20 22:17:42,552]    INFO         from\n",
            "[2023-04-20 22:17:42,552]    INFO         COVID-19\n",
            "[2023-04-20 22:17:42,552]    INFO         Patients/Data\n",
            "[2023-04-20 22:17:42,552]    INFO         files/HRA001149/HRR339729_f1.fastq.gz\n",
            "[2023-04-20 22:17:42,552]    INFO         /content/drive/MyDrive/Pate\n",
            "[2023-04-20 22:17:42,552]    INFO         Lab/DNP\n",
            "[2023-04-20 22:17:42,552]    INFO         Folder/Bioinformatics/H17\n",
            "[2023-04-20 22:17:42,552]    INFO         Lung\n",
            "[2023-04-20 22:17:42,552]    INFO         Mk\n",
            "[2023-04-20 22:17:42,552]    INFO         from\n",
            "[2023-04-20 22:17:42,552]    INFO         COVID-19\n",
            "[2023-04-20 22:17:42,552]    INFO         Patients/Data\n",
            "[2023-04-20 22:17:42,552]    INFO         files/HRA001149/HRR339729_r2.fastq.gz\n",
            "[2023-04-20 22:17:42,559]   ERROR An exception occurred\n",
            "Traceback (most recent call last):\n",
            "  File \"/usr/local/lib/python3.9/site-packages/kb_python/main.py\", line 476, in main\n",
            "    COMMAND_TO_FUNCTION[args.command](args)\n",
            "  File \"/usr/local/lib/python3.9/site-packages/kb_python/main.py\", line 136, in parse_count\n",
            "    count(\n",
            "  File \"/usr/local/lib/python3.9/site-packages/kb_python/count.py\", line 404, in count\n",
            "    bus_result = kallisto_bus(\n",
            "  File \"/usr/local/lib/python3.9/site-packages/kb_python/count.py\", line 62, in kallisto_bus\n",
            "    run_executable(command)\n",
            "  File \"/usr/local/lib/python3.9/site-packages/kb_python/utils.py\", line 114, in run_executable\n",
            "    raise sp.CalledProcessError(p.returncode, ' '.join(command))\n",
            "subprocess.CalledProcessError: Command '/usr/local/lib/python3.9/site-packages/kb_python/bins/linux/kallisto/kallisto bus -i Homo_sapiens.GRCh38.cdna.all.release-94_k31.idx -o . -x 10xv2 -t 2 /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729_f1.fastq.gz /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729_r2.fastq.gz' returned non-zero exit status 1.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!curl -O ftp://ftp.ensembl.org/pub/release-94/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Mvx9Q2oYesgP",
        "outputId": "09828cd4-cdf3-4ae2-ccd5-3beb56d4b615"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "100 64.7M  100 64.7M    0     0  5908k      0  0:00:11  0:00:11 --:--:-- 11.5M\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Download the index file to the working directory\n",
        "!wget {index_url} -P \"{work_dir}\"\n",
        "\n",
        "# Unzip the index file\n",
        "!gzip -d \"{work_dir}/Homo_sapiens.GRCh38.cdna.all.fa.gz\"\n",
        "\n",
        "# Build the kallisto index\n",
        "!kallisto index -i \"{work_dir}/Homo_sapiens.GRCh38.cdna.all.release-94_k31.idx\" \"{work_dir}/Homo_sapiens.GRCh38.cdna.all.fa\""
      ],
      "metadata": {
        "id": "i_gyY5WzrJPA"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!kallisto index -i \tHomo_sapiens.GRCh38.cdna.all.release-94_k31.idx\tHomo_sapiens.GRCh38.cdna.all.fa.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "t-wgoGPge-e1",
        "outputId": "a6a11c28-61ca-4b0e-fa1c-6c6eaf020e18"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "[build] loading fasta file Homo_sapiens.GRCh38.cdna.all.fa.gz\n",
            "[build] k-mer length: 31\n",
            "[build] warning: clipped off poly-A tail (longer than 10)\n",
            "        from 1471 target sequences\n",
            "[build] warning: replaced 5 non-ACGUT characters in the input sequence\n",
            "        with pseudorandom nucleotides\n",
            "[build] counting k-mers ... done.\n",
            "[build] building target de Bruijn graph ... ^C\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!ls -lh Homo_sapiens.GRCh38.cdna.all.release-94_k31.idx"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "z6UZxORLl-wE",
        "outputId": "0cd7ad19-813a-4414-8ba8-4b6555a35e00"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "-rw-r--r-- 1 root root 2.3G Apr 20 21:17 Homo_sapiens.GRCh38.cdna.all.release-94_k31.idx\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "t2g_url = \"https://github.com/BUStools/getting_started/releases/download/species_mixing/transcripts_to_genes_hg19mm10.txt\"\n",
        "!wget {t2g_url} -O \"{work_dir}/genes_t2g.txt\"\n",
        "\n",
        "# Run kb count with the downloaded index and t2g file\n",
        "!kb count -i Homo_sapiens.GRCh38.cdna.all.release-94_k31.idx -g \"{work_dir}/genes_t2g.txt\" --overwrite -t 2 -x 10xv2 /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_f1.fastq.gz  /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_r2.fastq.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "hV2GxJKzkaPI",
        "outputId": "96bec42b-ce3d-416c-fcf7-c4784d050cf2"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2023-04-20 21:46:13--  https://github.com/BUStools/getting_started/releases/download/species_mixing/transcripts_to_genes_hg19mm10.txt\n",
            "Resolving github.com (github.com)... 20.27.177.113\n",
            "Connecting to github.com (github.com)|20.27.177.113|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/191064839/37b22d80-913c-11e9-9025-bca8558040be?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230420%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230420T214613Z&X-Amz-Expires=300&X-Amz-Signature=1bf8f40bba0344195d88488e1e17834f121558c97c762000139bdd23647fe463&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=191064839&response-content-disposition=attachment%3B%20filename%3Dtranscripts_to_genes_hg19mm10.txt&response-content-type=application%2Foctet-stream [following]\n",
            "--2023-04-20 21:46:13--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/191064839/37b22d80-913c-11e9-9025-bca8558040be?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230420%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230420T214613Z&X-Amz-Expires=300&X-Amz-Signature=1bf8f40bba0344195d88488e1e17834f121558c97c762000139bdd23647fe463&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=191064839&response-content-disposition=attachment%3B%20filename%3Dtranscripts_to_genes_hg19mm10.txt&response-content-type=application%2Foctet-stream\n",
            "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 18953747 (18M) [application/octet-stream]\n",
            "Saving to: ‘/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/genes_t2g.txt’\n",
            "\n",
            "/content/drive/MyDr 100%[===================>]  18.08M  75.2MB/s    in 0.2s    \n",
            "\n",
            "2023-04-20 21:46:14 (75.2 MB/s) - ‘/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/genes_t2g.txt’ saved [18953747/18953747]\n",
            "\n",
            "[2023-04-20 21:46:17,103]    INFO Generating BUS file from\n",
            "[2023-04-20 21:46:17,103]    INFO         /content/drive/MyDrive/Pate\n",
            "[2023-04-20 21:46:17,103]    INFO         Lab/DNP\n",
            "[2023-04-20 21:46:17,103]    INFO         Folder/Bioinformatics/H17\n",
            "[2023-04-20 21:46:17,103]    INFO         Lung\n",
            "[2023-04-20 21:46:17,103]    INFO         Mk\n",
            "[2023-04-20 21:46:17,103]    INFO         from\n",
            "[2023-04-20 21:46:17,103]    INFO         COVID-19\n",
            "[2023-04-20 21:46:17,103]    INFO         Patients/Data\n",
            "[2023-04-20 21:46:17,103]    INFO         files/HRA001149/HRR339729/HRR339729_f1.fastq.gz\n",
            "[2023-04-20 21:46:17,103]    INFO         /content/drive/MyDrive/Pate\n",
            "[2023-04-20 21:46:17,103]    INFO         Lab/DNP\n",
            "[2023-04-20 21:46:17,104]    INFO         Folder/Bioinformatics/H17\n",
            "[2023-04-20 21:46:17,104]    INFO         Lung\n",
            "[2023-04-20 21:46:17,104]    INFO         Mk\n",
            "[2023-04-20 21:46:17,104]    INFO         from\n",
            "[2023-04-20 21:46:17,104]    INFO         COVID-19\n",
            "[2023-04-20 21:46:17,104]    INFO         Patients/Data\n",
            "[2023-04-20 21:46:17,104]    INFO         files/HRA001149/HRR339729/HRR339729_r2.fastq.gz\n",
            "[2023-04-20 21:46:17,137]   ERROR An exception occurred\n",
            "Traceback (most recent call last):\n",
            "  File \"/usr/local/lib/python3.9/dist-packages/kb_python/main.py\", line 476, in main\n",
            "    COMMAND_TO_FUNCTION[args.command](args)\n",
            "  File \"/usr/local/lib/python3.9/dist-packages/kb_python/main.py\", line 136, in parse_count\n",
            "    count(\n",
            "  File \"/usr/local/lib/python3.9/dist-packages/kb_python/count.py\", line 404, in count\n",
            "    bus_result = kallisto_bus(\n",
            "  File \"/usr/local/lib/python3.9/dist-packages/kb_python/count.py\", line 62, in kallisto_bus\n",
            "    run_executable(command)\n",
            "  File \"/usr/local/lib/python3.9/dist-packages/kb_python/utils.py\", line 114, in run_executable\n",
            "    raise sp.CalledProcessError(p.returncode, ' '.join(command))\n",
            "subprocess.CalledProcessError: Command '/usr/local/lib/python3.9/dist-packages/kb_python/bins/linux/kallisto/kallisto bus -i Homo_sapiens.GRCh38.cdna.all.release-94_k31.idx -o . -x 10xv2 -t 2 /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_f1.fastq.gz /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_r2.fastq.gz' returned non-zero exit status 1.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Quantification\n",
        "\n",
        "This step runs `kb` to quantify the reads. `kb` can take as input URLs where the reads are located, and will stream the data to Google Colab where it is quantified as it is downloaded. This allows for quantifying very large datasets without first downloading them and saving them to disk."
      ],
      "metadata": {
        "id": "JX_cPaflgNQ1"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Define the function\n",
        "\n",
        "Let's first define the function to do the quantification.\n",
        "\n",
        "Aligning and quantifying using kallisto/bustool\n",
        "Kallisto is a program for quantifying abundances of transcripts from RNA-Seq data, which uses a novel idea of pseudoalignment for fast and accurate quantification of transcript abundances from RNA-Seq data. Kallisto can quantify expression levels of genes, transcripts, and isoforms. Kallisto generates an index from the reference transcriptome that allows fast pseudoalignment of RNA-Seq reads, followed by generation of gene- and transcript-level counts.\n",
        "\n",
        "Bustools is a set of tools for analyzing BUS files generated by kallisto. BUS files contain information about which barcodes were detected in which transcript and how many UMIs were associated with each barcode-transcript pair. Bustools can be used to correct and sort the barcode and UMI information in the BUS file, filter out low-quality reads and barcodes, count the number of unique molecular identifiers (UMIs) associated with each gene or transcript, and perform other downstream analyses.\n",
        "\n",
        "We will need to reference the GRCh38 Human Genome here to align it to our sequencing data\n",
        "\n",
        "The goal is to do the quantification given a patient ID.\n",
        "\n",
        "**Note:** You might need to create the folder /content/dive/MyDrive/LungMk\n",
        "\n",
        "**Note:** We are saving the files into the google drive so that they are not deleted."
      ],
      "metadata": {
        "id": "uRSuyYBdo2_p"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Define the function\n",
        "\n",
        "idx = f'{work_dir}/genome.idx'\n",
        "t2g = f'{work_dir}/genes_t2g.txt'\n",
        "\n",
        "def quantify(idx, t2g, r1, r2, out):\n",
        "  !kb count --verbose  -i \"{idx}\" -g \"{t2g}\" --overwrite -t 2 -x 10xv2 -o \"{out}\" \"{r1}\" \"{r2}\"\n",
        "\n",
        "def quantify_patient(patient_id):\n",
        "  r1 = f'{data_root}/{patient_id}/{patient_id}_f1.fastq.gz'\n",
        "  r2 = f'{data_root}/{patient_id}/{patient_id}_r2.fastq.gz'\n",
        "  out = f\"{work_dir}/counts/{patient_id}/\"\n",
        "  quantify(idx, t2g, r1, r2, out)"
      ],
      "metadata": {
        "id": "H9QKSZMVgFkV"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "There is an issue with this code, we are not getting the correct output for the bus files. "
      ],
      "metadata": {
        "id": "hB6fzK62AR9O"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "\n",
        "idx = f'{work_dir}/genome.idx'\n",
        "t2g = f'{work_dir}/genes_t2g.txt'\n",
        "\n",
        "def quantify(idx, t2g, r1, r2, out):\n",
        "  !kb count --verbose  -i \"{idx}\" -g \"{t2g}\" --overwrite -t 2 -x 10xv2 -o \"{out}\" \"{r1}\" \"{r2}\"\n",
        "  \n",
        "def quantify_patient(patient_id):\n",
        "  r1 = f'{data_root}/{patient_id}/{patient_id}_f1.fastq.gz'\n",
        "  r2 = f'{data_root}/{patient_id}/{patient_id}_r2.fastq.gz'\n",
        "  out = f\"{work_dir}/counts/{patient_id}/\"\n",
        "  os.makedirs(out, exist_ok=True)\n",
        "  quantify(idx, t2g, r1, r2, out)\n",
        "  !kb count -o {out}/output -g {t2g} -e {out}/matrix.ec -t {out}/transcripts.txt --genecounts {out}/output.counts.tsv {out}/output.bus\n"
      ],
      "metadata": {
        "id": "cQstZxXiAoAY"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Set working directory\n",
        "work_dir = \"/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk\"\n",
        "os.makedirs(f\"{work_dir}/counts\", exist_ok=True)\n",
        "\n",
        "# Set file paths\n",
        "idx = f\"{work_dir}/genome.idx\"\n",
        "t2g = f\"{work_dir}/genes_t2g.txt\"\n",
        "r1_path = \"/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_f1.fastq.gz\"\n",
        "r2_path = \"/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_r2.fastq.gz\"\n",
        "\n",
        "def quantify(idx, t2g, r1, r2, out):\n",
        "    !kb count --verbose -i \"{idx}\" -g \"{t2g}\" --overwrite -t 2 -x 10xv2 -o \"{out}\" \"{r1}\" \"{r2}\"\n",
        "\n",
        "def quantify_patient(patient_id):\n",
        "    out = f\"{work_dir}/counts/{patient_id}\"\n",
        "    os.makedirs(out, exist_ok=True)\n",
        "    r1 = r1_path\n",
        "    r2 = r2_path\n",
        "    quantify(idx, t2g, r1, r2, out)\n",
        "    !kb count -o {out}/output -g {t2g} -e {out}/matrix.ec -t {out}/transcripts.txt --genecounts {out}/output.counts.tsv {out}/output.bus\n",
        "\n",
        "# Call the function with patient ID \"HRR339729\"\n",
        "quantify_patient(\"HRR339729\")\n"
      ],
      "metadata": {
        "id": "mAcURvRT-cEq",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "a170777a-8d92-4c68-d4ac-3cee66ff5ae7"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[2023-04-19 19:47:14,134]   DEBUG Printing verbose output\n",
            "[2023-04-19 19:47:14,134]   DEBUG Creating tmp directory\n",
            "[2023-04-19 19:47:14,134]   DEBUG Namespace(list=False, command='count', keep_tmp=False, verbose=True, i='/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/genome.idx', g='/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/genes_t2g.txt', x='10xv2', o='/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729', w=None, t=2, m='4G', c1=None, c2=None, overwrite=True, lamanno=False, filter=None, loom=False, h5ad=False, fastqs=['/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_f1.fastq.gz', '/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_r2.fastq.gz'])\n",
            "[2023-04-19 19:47:14,136]    INFO Generating BUS file from\n",
            "[2023-04-19 19:47:14,137]    INFO         /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_f1.fastq.gz\n",
            "[2023-04-19 19:47:14,137]    INFO         /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_r2.fastq.gz\n",
            "[2023-04-19 19:47:14,137]   DEBUG /usr/local/lib/python3.9/dist-packages/kb_python/bins/linux/kallisto/kallisto bus -i /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/genome.idx -o /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729 -x 10xv2 -t 2 /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_f1.fastq.gz /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_r2.fastq.gz\n",
            "[2023-04-19 20:27:38,549]   DEBUG \n",
            "[2023-04-19 20:27:38,549]   DEBUG [index] k-mer length: 31\n",
            "[2023-04-19 20:27:38,549]   DEBUG [index] number of targets: 187,626\n",
            "[2023-04-19 20:27:38,549]   DEBUG [index] number of k-mers: 108,619,921\n",
            "[2023-04-19 20:27:38,549]   DEBUG [index] number of equivalence classes: 752,021\n",
            "[2023-04-19 20:27:38,549]   DEBUG [quant] will process sample 1: /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_f1.fastq.gz\n",
            "[2023-04-19 20:27:38,549]   DEBUG /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_r2.fastq.gz\n",
            "[2023-04-19 20:27:38,549]   DEBUG [quant] finding pseudoalignments for the reads ... done\n",
            "[2023-04-19 20:27:38,549]   DEBUG [quant] processed 227,781,009 reads, 97,711,650 reads pseudoaligned\n",
            "[2023-04-19 20:27:38,550]    INFO Sorting BUS file /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/output.bus to tmp/output.s.bus\n",
            "[2023-04-19 20:27:38,550]   DEBUG /usr/local/lib/python3.9/dist-packages/kb_python/bins/linux/bustools/bustools sort -o tmp/output.s.bus -T tmp -t 2 -m 4G /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/output.bus\n",
            "[2023-04-19 20:28:27,172]   DEBUG Read in 97711650 BUS records\n",
            "[2023-04-19 20:28:27,172]    INFO Whitelist not provided\n",
            "[2023-04-19 20:28:27,172]    INFO Copying pre-packaged 10XV2 whitelist to /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729\n",
            "[2023-04-19 20:28:27,290]    INFO Inspecting BUS file tmp/output.s.bus\n",
            "[2023-04-19 20:28:27,290]   DEBUG /usr/local/lib/python3.9/dist-packages/kb_python/bins/linux/bustools/bustools inspect -o /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/inspect.json -w /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/10xv2_whitelist.txt -e /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/matrix.ec tmp/output.s.bus\n",
            "[2023-04-19 20:28:48,946]    INFO Correcting BUS records in tmp/output.s.bus to tmp/output.s.c.bus with whitelist /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/10xv2_whitelist.txt\n",
            "[2023-04-19 20:28:48,947]   DEBUG /usr/local/lib/python3.9/dist-packages/kb_python/bins/linux/bustools/bustools correct -o tmp/output.s.c.bus -w /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/10xv2_whitelist.txt tmp/output.s.bus\n",
            "[2023-04-19 20:29:07,677]   DEBUG Found 737280 barcodes in the whitelist\n",
            "[2023-04-19 20:29:07,677]   DEBUG Number of hamming dist 1 barcodes = 20550336\n",
            "[2023-04-19 20:29:07,677]   DEBUG Processed 64724300 bus records\n",
            "[2023-04-19 20:29:07,677]   DEBUG In whitelist = 923192\n",
            "[2023-04-19 20:29:07,677]   DEBUG Corrected = 161639\n",
            "[2023-04-19 20:29:07,677]   DEBUG Uncorrected = 63639469\n",
            "[2023-04-19 20:29:07,678]    INFO Sorting BUS file tmp/output.s.c.bus to /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/output.unfiltered.bus\n",
            "[2023-04-19 20:29:07,678]   DEBUG /usr/local/lib/python3.9/dist-packages/kb_python/bins/linux/bustools/bustools sort -o /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/output.unfiltered.bus -T tmp -t 2 -m 4G tmp/output.s.c.bus\n",
            "[2023-04-19 20:29:10,499]   DEBUG Read in 1084831 BUS records\n",
            "[2023-04-19 20:29:10,503]    INFO Generating count matrix /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/counts_unfiltered/cells_x_genes from BUS file /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/output.unfiltered.bus\n",
            "[2023-04-19 20:29:10,503]   DEBUG /usr/local/lib/python3.9/dist-packages/kb_python/bins/linux/bustools/bustools count -o /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/counts_unfiltered/cells_x_genes -g /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/genes_t2g.txt -e /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/matrix.ec -t /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/transcripts.txt --genecounts /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/output.unfiltered.bus\n",
            "[2023-04-19 20:29:13,727]   DEBUG Removing tmp directory\n",
            "usage: kb count\n",
            "       [-h]\n",
            "       [--keep-tmp]\n",
            "       [--verbose]\n",
            "       -i\n",
            "       INDEX\n",
            "       -g\n",
            "       T2G\n",
            "       -x\n",
            "       TECHNOLOGY\n",
            "       [-o OUT]\n",
            "       [-w WHITELIST]\n",
            "       [-t THREADS]\n",
            "       [-m MEMORY]\n",
            "       [-c1 T2C]\n",
            "       [-c2 T2C]\n",
            "       [--overwrite]\n",
            "       [--lamanno | --filter [{bustools}]]\n",
            "       [--loom | --h5ad]\n",
            "       fastqs\n",
            "       [fastqs ...]\n",
            "kb count: error: argument -t: invalid int value: '/content/drive/MyDrive/Pate'\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Checking to see if there is an issue with the patient data"
      ],
      "metadata": {
        "id": "eI_1dmyHCFIH"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import gzip\n",
        "\n",
        "# Path to the input files\n",
        "r1_path = \"/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_f1.fastq.gz\"\n",
        "r2_path = \"/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_r2.fastq.gz\"\n",
        "\n",
        "# Check that the input files are in the correct format and can be read by kb\n",
        "with gzip.open(r1_path, \"rt\") as f:\n",
        "    print(f\"Contents of {r1_path}:\")\n",
        "    print(f.read(100))\n",
        "\n",
        "with gzip.open(r2_path, \"rt\") as f:\n",
        "    print(f\"Contents of {r2_path}:\")\n",
        "    print(f.read(100))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "AFaIlLw0Brno",
        "outputId": "49fbb04d-b217-4e69-8906-5af3b99468c5"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Contents of /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_f1.fastq.gz:\n",
            "@v300039845L1C001R0010000104/1\n",
            "GCTGAATAGTCGGCCTTTTTAATCTGTT\n",
            "+\n",
            "EFDED1?AE4EECAGFGFCBDG>@=EF?\n",
            "@v3000398\n",
            "Contents of /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_r2.fastq.gz:\n",
            "@v300039845L1C001R0010000104/2\n",
            "NTTTACTTTGTTCTTTTACTAGCATTTTGAACTGGGAATTTTAATTTATTTCCATTCTTCTATTAACAA\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "\n",
        "# Load the count matrix\n",
        "counts = pd.read_csv(\"/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/counts_unfiltered/cells_x_genes.mtx\", sep=\" \", skiprows=3, header=None)\n",
        "counts.columns = [\"barcode\", \"gene_id\", \"UMI\"]\n",
        "\n",
        "# Print the first few rows to inspect the data\n",
        "print(counts.head())"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 393
        },
        "id": "SrzT9thvaw1c",
        "outputId": "b23b0c1d-b1b9-4279-dbcf-7de246c10238"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "FileNotFoundError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-21-48c0f07c2510>\u001b[0m in \u001b[0;36m<cell line: 4>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;31m# Load the count matrix\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0mcounts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/counts_unfiltered/cells_x_genes.mtx\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\" \"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mskiprows\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheader\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0mcounts\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m\"barcode\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"gene_id\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"UMI\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.9/dist-packages/pandas/util/_decorators.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    209\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    210\u001b[0m                     \u001b[0mkwargs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mnew_arg_name\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnew_arg_value\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 211\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    212\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    213\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mcast\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mF\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.9/dist-packages/pandas/util/_decorators.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    329\u001b[0m                     \u001b[0mstacklevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfind_stack_level\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    330\u001b[0m                 )\n\u001b[0;32m--> 331\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    332\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    333\u001b[0m         \u001b[0;31m# error: \"Callable[[VarArg(Any), KwArg(Any)], Any]\" has no\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.9/dist-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, error_bad_lines, warn_bad_lines, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)\u001b[0m\n\u001b[1;32m    948\u001b[0m     \u001b[0mkwds\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkwds_defaults\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    949\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 950\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    951\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    952\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.9/dist-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    603\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    604\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 605\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    606\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    607\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.9/dist-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1440\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1441\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhandles\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mIOHandles\u001b[0m \u001b[0;34m|\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1442\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1443\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1444\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.9/dist-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1733\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0;34m\"b\"\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1734\u001b[0m                     \u001b[0mmode\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;34m\"b\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1735\u001b[0;31m             self.handles = get_handle(\n\u001b[0m\u001b[1;32m   1736\u001b[0m                 \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1737\u001b[0m                 \u001b[0mmode\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.9/dist-packages/pandas/io/common.py\u001b[0m in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    854\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mioargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mencoding\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;34m\"b\"\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mioargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmode\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    855\u001b[0m             \u001b[0;31m# Encoding\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 856\u001b[0;31m             handle = open(\n\u001b[0m\u001b[1;32m    857\u001b[0m                 \u001b[0mhandle\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    858\u001b[0m                 \u001b[0mioargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmode\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/counts_unfiltered/cells_x_genes.mtx'"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import struct\n",
        "\n",
        "bus_path = \"/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/output.bus\"\n",
        "\n",
        "# Check the number of lines in the output.bus file\n",
        "num_lines = 0\n",
        "with open(bus_path, \"rb\") as f:\n",
        "    while True:\n",
        "        binary = f.read(4)\n",
        "        if not binary:\n",
        "            break\n",
        "        (record_size,) = struct.unpack(\"<I\", binary)\n",
        "        f.read(record_size)\n",
        "        num_lines += 1\n",
        "\n",
        "print(f\"Number of lines in the output.bus file: {num_lines}\")\n",
        "\n",
        "# Print the first few lines of the output.bus file\n",
        "with open(bus_path, \"rb\") as f:\n",
        "    for i in range(10):\n",
        "        binary = f.read(4)\n",
        "        (record_size,) = struct.unpack(\"<I\", binary)\n",
        "        record = f.read(record_size)\n",
        "        print(record)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6CbEBqmc4p8Z",
        "outputId": "f5299237-34e8-4fd8-b031-805225ff2b97"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Number of lines in the output.bus file: 6\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "IOPub data rate exceeded.\n",
            "The notebook server will temporarily stop sending output\n",
            "to the client in order to avoid crashing it.\n",
            "To change this limit, set the config variable\n",
            "`--NotebookApp.iopub_data_rate_limit`.\n",
            "\n",
            "Current values:\n",
            "NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
            "NotebookApp.rate_limit_window=3.0 (secs)\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Load the gene annotation file\n",
        "genes = pd.read_csv(\"/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/genes_t2g.txt\", sep=\"\\t\", header=None, names=[\"transcript_id\", \"gene_id\", \"gene_name\"])\n",
        "\n",
        "# Print the first few rows of the dataframe\n",
        "print(genes.head())\n",
        "\n",
        "# Print the shape of the dataframe\n",
        "print(\"Shape of dataframe:\", genes.shape)\n",
        "\n",
        "# Print the unique values in the gene_id column\n",
        "print(\"Unique gene IDs:\", genes[\"gene_id\"].unique())"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "zXqtOz3ubpEG",
        "outputId": "f322cb2c-1e28-4b20-e328-414d0e73efe0"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "          transcript_id               gene_id     gene_name\n",
            "0  hg19_ENST00000456328  hg19_ENSG00000223972  hg19_DDX11L1\n",
            "1  hg19_ENST00000515242  hg19_ENSG00000223972  hg19_DDX11L1\n",
            "2  hg19_ENST00000518655  hg19_ENSG00000223972  hg19_DDX11L1\n",
            "3  hg19_ENST00000450305  hg19_ENSG00000223972  hg19_DDX11L1\n",
            "4  hg19_ENST00000438504  hg19_ENSG00000227232   hg19_WASH7P\n",
            "Shape of dataframe: (333131, 3)\n",
            "Unique gene IDs: ['hg19_ENSG00000223972' 'hg19_ENSG00000227232' 'hg19_ENSG00000243485' ...\n",
            " 'mm10_ENSMUSG00000098647' 'mm10_ENSMUSG00000096730'\n",
            " 'mm10_ENSMUSG00000095742']\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Looping\n",
        "\n",
        "Here is the code that allows us to loop through all the patients when we are ready for it."
      ],
      "metadata": {
        "id": "xrxMvj6XpQAM"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Here we list all the patient folder. We can use this when we need to loop\n",
        "# through all the patients.\n",
        "\n",
        "folders = [f for f in os.listdir(data_root) if os.path.isdir(os.path.join(data_root, f))]\n",
        "print(folders)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "HMCxM7xziZEz",
        "outputId": "54744c45-48eb-4c21-9651-962c0c9c5ade"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "['HRR339728', 'HRR339729', 'HRR339730', 'HRR339731', 'HRR339732', 'HRR339733', 'HRR339734', 'HRR339735', 'HRR339736', 'HRR339737', 'HRR339738', 'HRR339740', 'HRR339739', 'HRR339741', 'HRR339743', 'HRR339748', 'HRR339753', 'HRR339751', 'HRR339752', 'HRR339755', 'HRR339756', 'HRR339754', 'HRR339749', 'HRR339750', 'HRR339742', 'HRR339746', 'HRR339747', 'HRR339744', 'HRR339759', 'HRR339757', 'HRR339761', 'HRR339760', 'HRR339762', 'HRR339765', 'HRR339764', 'HRR339763', 'HRR339770', 'HRR339769', 'HRR339771', 'HRR339772', 'HRR339775', 'HRR339777', 'HRR339774', 'HRR339776', 'HRR339773', 'HRR339778', 'HRR339786', 'HRR339787', 'HRR339795', 'HRR339766', 'HRR339767', 'HRR339768', 'HRR339790', 'HRR339791', 'HRR339792', 'HRR339789', 'HRR339788', 'HRR339758', 'HRR339794', 'HRR339793', 'HRR339805', 'HRR339801', 'HRR339802', 'HRR339799', 'HRR339804', 'HRR339803', 'HRR339798', 'HRR339797', 'HRR339796', 'HRR339800', 'HRR339808', 'HRR339806', 'HRR339815', 'HRR339812', 'HRR339811', 'HRR339814', 'HRR339807', 'HRR339813', 'HRR339810', 'HRR339809', 'HRR339816', 'HRR339824', 'HRR339821', 'HRR339823', 'HRR339822', 'HRR339817', 'HRR339819', 'HRR339820', 'HRR339825', 'HRR339818', 'HRR339830', 'HRR339834', 'HRR339832', 'HRR339835', 'HRR339833', 'HRR339829', 'HRR339828', 'HRR339826', 'HRR339831', 'HRR339827', 'HRR339842', 'HRR339840', 'HRR339837', 'HRR339838', 'HRR339839', 'HRR339836', 'HRR339843', 'HRR339844', 'HRR339841', 'HRR339846', 'HRR339845', 'HRR339847', 'HRR339848', 'HRR339849', 'HRR339782 (1)', 'HRR339783 (1)', 'HRR339781 (1)', 'HRR339745', 'HRR339779', 'HRR339780', 'HRR339785', 'HRR339783', 'HRR339781', 'HRR339784', 'HRR339782', 'HRA001149']\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Process for one patient\n",
        "\n",
        "Here we try out for one single patient as a POC."
      ],
      "metadata": {
        "id": "mzQpCn9Jpa0Q"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%time\n",
        "\n",
        "# Let's do quatification for one patient HRR339728\n",
        "patient_id = 'HRR339729'\n",
        "quantify_patient(patient_id)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "kovvl1XKgz2H",
        "outputId": "cc0a91cd-cc03-42b4-d685-6acebb59c41f"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[2023-04-18 20:14:02,140]   DEBUG Printing verbose output\n",
            "[2023-04-18 20:14:02,141]   DEBUG Creating tmp directory\n",
            "[2023-04-18 20:14:02,141]   DEBUG Namespace(list=False, command='count', keep_tmp=False, verbose=True, i='/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/genome.idx', g='/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/genes_t2g.txt', x='10xv2', o='/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/', w=None, t=2, m='4G', c1=None, c2=None, overwrite=True, lamanno=False, filter=None, loom=False, h5ad=False, fastqs=['/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_S1_L001_R1_001.fastq.gz', '/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_S1_L001_R2_001.fastq.gz'])\n",
            "[2023-04-18 20:14:02,144]    INFO Generating BUS file from\n",
            "[2023-04-18 20:14:02,144]    INFO         /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_S1_L001_R1_001.fastq.gz\n",
            "[2023-04-18 20:14:02,145]    INFO         /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_S1_L001_R2_001.fastq.gz\n",
            "[2023-04-18 20:14:02,145]   DEBUG /usr/local/lib/python3.9/dist-packages/kb_python/bins/linux/kallisto/kallisto bus -i /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/genome.idx -o /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/ -x 10xv2 -t 2 /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_S1_L001_R1_001.fastq.gz /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_S1_L001_R2_001.fastq.gz\n",
            "[2023-04-18 20:14:02,161]   DEBUG kallisto 0.46.0\n",
            "[2023-04-18 20:14:02,161]   DEBUG Generates BUS files for single-cell sequencing\n",
            "[2023-04-18 20:14:02,161]   DEBUG \n",
            "[2023-04-18 20:14:02,161]   DEBUG Usage: kallisto bus [arguments] FASTQ-files\n",
            "[2023-04-18 20:14:02,161]   DEBUG \n",
            "[2023-04-18 20:14:02,161]   DEBUG Required arguments:\n",
            "[2023-04-18 20:14:02,161]   DEBUG -i, --index=STRING            Filename for the kallisto index to be used for\n",
            "[2023-04-18 20:14:02,161]   DEBUG pseudoalignment\n",
            "[2023-04-18 20:14:02,161]   DEBUG -o, --output-dir=STRING       Directory to write output to\n",
            "[2023-04-18 20:14:02,161]   DEBUG -x, --technology=STRING       Single-cell technology used\n",
            "[2023-04-18 20:14:02,161]   DEBUG \n",
            "[2023-04-18 20:14:02,161]   DEBUG Optional arguments:\n",
            "[2023-04-18 20:14:02,162]   DEBUG -l, --list                    List all single-cell technologies supported\n",
            "[2023-04-18 20:14:02,162]   DEBUG -t, --threads=INT             Number of threads to use (default: 1)\n",
            "[2023-04-18 20:14:02,162]   DEBUG \n",
            "[2023-04-18 20:14:02,162]   DEBUG Error: file not found /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_S1_L001_R1_001.fastq.gz\n",
            "[2023-04-18 20:14:02,162]   DEBUG Error: file not found /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_S1_L001_R2_001.fastq.gz\n",
            "[2023-04-18 20:14:02,162]   ERROR An exception occurred\n",
            "Traceback (most recent call last):\n",
            "  File \"/usr/local/lib/python3.9/dist-packages/kb_python/main.py\", line 476, in main\n",
            "    COMMAND_TO_FUNCTION[args.command](args)\n",
            "  File \"/usr/local/lib/python3.9/dist-packages/kb_python/main.py\", line 136, in parse_count\n",
            "    count(\n",
            "  File \"/usr/local/lib/python3.9/dist-packages/kb_python/count.py\", line 404, in count\n",
            "    bus_result = kallisto_bus(\n",
            "  File \"/usr/local/lib/python3.9/dist-packages/kb_python/count.py\", line 62, in kallisto_bus\n",
            "    run_executable(command)\n",
            "  File \"/usr/local/lib/python3.9/dist-packages/kb_python/utils.py\", line 114, in run_executable\n",
            "    raise sp.CalledProcessError(p.returncode, ' '.join(command))\n",
            "subprocess.CalledProcessError: Command '/usr/local/lib/python3.9/dist-packages/kb_python/bins/linux/kallisto/kallisto bus -i /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/genome.idx -o /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/ -x 10xv2 -t 2 /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_S1_L001_R1_001.fastq.gz /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_S1_L001_R2_001.fastq.gz' returned non-zero exit status 1.\n",
            "[2023-04-18 20:14:02,162]   DEBUG Removing tmp directory\n",
            "usage: kb count\n",
            "       [-h]\n",
            "       [--keep-tmp]\n",
            "       [--verbose]\n",
            "       -i\n",
            "       INDEX\n",
            "       -g\n",
            "       T2G\n",
            "       -x\n",
            "       TECHNOLOGY\n",
            "       [-o OUT]\n",
            "       [-w WHITELIST]\n",
            "       [-t THREADS]\n",
            "       [-m MEMORY]\n",
            "       [-c1 T2C]\n",
            "       [-c2 T2C]\n",
            "       [--overwrite]\n",
            "       [--lamanno | --filter [{bustools}]]\n",
            "       [--loom | --h5ad]\n",
            "       fastqs\n",
            "       [fastqs ...]\n",
            "kb count: error: argument -t: invalid int value: '/content/drive/MyDrive/Pate'\n",
            "CPU times: user 42.7 ms, sys: 28.4 ms, total: 71.2 ms\n",
            "Wall time: 6.46 s\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Quality Control\n",
        "\n",
        "* https://colab.research.google.com/github/pachterlab/kallistobustools/blob/master/docs/tutorials/kb_getting_started/python/kb_intro_2_python.ipynb\n",
        "\n",
        "## Filtering cells based on count\n",
        "Preliminary counts were then used for downstream analysis. Quality control was applied to cells based on three metrics step by step: the total UMI counts, number of detected genes and proportion of mitochondrial gene counts per cell. Specifically, cells with less than 1500 UMI counts and 500 detected genes were filtered, as well as cells with more than 10% mitochondrial gene counts. "
      ],
      "metadata": {
        "id": "B0-TRTDjHFCa"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Load the output from kallisto bus\n",
        "matrix = pd.read_csv(\"/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/counts_unfiltered/cells_x_genes.mtx\", sep=\" \", skiprows=3, header=None)\n",
        "matrix.columns = [\"barcode\", \"gene_id\", \"UMI\"]\n",
        "\n",
        "# Set the barcode column as the index\n",
        "matrix = matrix.pivot(index=\"barcode\", columns=\"gene_id\", values=\"UMI\")\n",
        "\n",
        "# Calculate the total UMI counts and number of detected genes per cell\n",
        "umi_counts = matrix.sum(axis=1)\n",
        "gene_counts = (matrix > 0).sum(axis=1)\n",
        "\n",
        "# Load the gene annotation file\n",
        "genes = pd.read_csv(\"/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/genes_t2g.txt\", sep=\"\\t\", header=None, names=[\"transcript_id\", \"gene_id\", \"gene_name\"])\n",
        "\n",
        "# Extract mitochondrial gene names\n",
        "mito_genes = genes[genes.gene_name.str.startswith(\"MT-\")].gene_id.tolist()\n",
        "\n",
        "# Calculate the proportion of mitochondrial gene counts per cell\n",
        "mito_counts = matrix[mito_genes].sum(axis=1)\n",
        "mito_prop = mito_counts / umi_counts\n",
        "\n",
        "# Filter cells with less than 1500 UMI counts, less than 500 detected genes, or more than 10% mitochondrial gene counts\n",
        "cells_to_keep = (umi_counts >= 1500) & (gene_counts >= 500) & (mito_prop <= 0.1)\n",
        "filtered_matrix = matrix.loc[cells_to_keep]\n",
        "\n",
        "# Calculate total UMI counts and number of detected genes for the filtered matrix\n",
        "filtered_umi_counts = filtered_matrix.sum(axis=1)\n",
        "filtered_gene_counts = (filtered_matrix > 0).sum(axis=1)"
      ],
      "metadata": {
        "id": "hLZCU5GrK6zn"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Validate XML"
      ],
      "metadata": {
        "id": "I4GyN9gFTFh8"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "xml_gz = '/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_sta.xml'\n",
        "\n",
        "with open(xml_gz, 'rt') as f:\n",
        "  HRR339729_sta = f.read()\n",
        "\n",
        "print(HRR339729_sta)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "XzN2pcZOT8Jv",
        "outputId": "d5fd30a0-1e41-4519-bd4b-9428bea7a44f"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "<Run accession=\"/p300/HRA-Process/temp/HRA001149/HRR339729/\" read_length=\"fixed\" spot_count=\"227781009\" base_count=\"29155969152\" \n",
            "base_count_bio=\"29155969152\" spot_count_mates=\"227781009\" base_count_bio_mates=\"29155969152\" spot_count_bad=\"0\" base_count_bio_bad=\"0\" spot_count_filtered=\"0\" \n",
            "base_count_bio_filtered=\"0\" cmp_base_count=\"29155969152\">\n",
            "  <Size value=\"74256608934\" units=\"bytes\"/>\n",
            "  <Bases cs_native=\"false\" count=\"29155969152\">\n",
            "    <Base value=\"A\" count=\"8270024956\"/>\n",
            "    <Base value=\"C\" count=\"6735888857\"/>\n",
            "    <Base value=\"G\" count=\"6335923237\"/>\n",
            "    <Base value=\"T\" count=\"7804688716\"/>\n",
            "    <Base value=\"N\" count=\"9443386\"/>\n",
            "  </Bases>\n",
            "  <GC-Content=\"44.83%\"/>\n",
            "  <AlignInfo>\n",
            "  </AlignInfo>\n",
            "  <Statistics nreads=\"2\" nspots=\"227781009\">\n",
            "    <Read index=\"0\" count=\"227781009\" average=\"28\" stdev=\"0\"/>\n",
            "    <Read index=\"1\" count=\"227781009\" average=\"100\" stdev=\"0\"/>\n",
            "  </Statistics>\n",
            "  <QualityCount>\n",
            "<Quality value=\"0\" count=\"9443386\"/>\n",
            "<Quality value=\"4\" count=\"3469229\"/>\n",
            "<Quality value=\"5\" count=\"40900687\"/>\n",
            "<Quality value=\"6\" count=\"53533223\"/>\n",
            "<Quality value=\"7\" count=\"50026344\"/>\n",
            "<Quality value=\"8\" count=\"49175535\"/>\n",
            "<Quality value=\"9\" count=\"49112665\"/>\n",
            "<Quality value=\"10\" count=\"51769266\"/>\n",
            "<Quality value=\"11\" count=\"55832350\"/>\n",
            "<Quality value=\"12\" count=\"56516200\"/>\n",
            "<Quality value=\"13\" count=\"61073062\"/>\n",
            "<Quality value=\"14\" count=\"69720733\"/>\n",
            "<Quality value=\"15\" count=\"73194458\"/>\n",
            "<Quality value=\"16\" count=\"86042181\"/>\n",
            "<Quality value=\"17\" count=\"87752960\"/>\n",
            "<Quality value=\"18\" count=\"101934206\"/>\n",
            "<Quality value=\"19\" count=\"101852905\"/>\n",
            "<Quality value=\"20\" count=\"125861428\"/>\n",
            "<Quality value=\"21\" count=\"143939211\"/>\n",
            "<Quality value=\"22\" count=\"158518829\"/>\n",
            "<Quality value=\"23\" count=\"177098634\"/>\n",
            "<Quality value=\"24\" count=\"212029874\"/>\n",
            "<Quality value=\"25\" count=\"229146057\"/>\n",
            "<Quality value=\"26\" count=\"256013745\"/>\n",
            "<Quality value=\"27\" count=\"291428404\"/>\n",
            "<Quality value=\"28\" count=\"333034052\"/>\n",
            "<Quality value=\"29\" count=\"393668076\"/>\n",
            "<Quality value=\"30\" count=\"432984572\"/>\n",
            "<Quality value=\"31\" count=\"522522417\"/>\n",
            "<Quality value=\"32\" count=\"631247554\"/>\n",
            "<Quality value=\"33\" count=\"744592611\"/>\n",
            "<Quality value=\"34\" count=\"906699801\"/>\n",
            "<Quality value=\"35\" count=\"1435638116\"/>\n",
            "<Quality value=\"36\" count=\"3463853320\"/>\n",
            "<Quality value=\"37\" count=\"14616189933\"/>\n",
            "<Quality value=\"38\" count=\"3080153128\"/>\n",
            "  </QualityCount>\n",
            "  <Databases>\n",
            "    <Database>\n",
            "      <Table name=\"SEQUENCE\">\n",
            "        <Statistics source=\"meta\">\n",
            "          <Rows count=\"227781009\"/>\n",
            "          <Elements count=\"29155969152\"/>\n",
            "        </Statistics>\n",
            "      </Table>\n",
            "    </Database>\n",
            "  </Databases>\n",
            "</Run>\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "validate_xml(xml_gz)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "1vXAZkC0UClL",
        "outputId": "786425fb-3024-44f8-bde9-cfaadc496d49"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "XML syntax error at line 12, column 14: error parsing attribute name, line 12, column 14\n",
            "Error line:   <GC-Content=\"44.83%\"/>\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Fix maleformed XML and print XML Structure\n",
        "This XML has a maleformed tag per GC-Content and we have to fix it before it parsing it to understand the XML structure.\n",
        "\n",
        "* The presence of an invalid token in an XML file could potentially cause issues with downstream RNA-seq analysis. If the invalid token affects the metadata or annotation information contained in the XML file, it may result in incorrect or incomplete information being used in downstream analysis. This could result in incorrect interpretations or conclusions being drawn from the analysis. It is important to ensure that all input files, including XML files, are correctly formatted and free of errors to minimize the risk of such issues occurring."
      ],
      "metadata": {
        "id": "gwchrovSUMn3"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "\n",
        "# Define paths\n",
        "xml_path = '/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_sta.xml'\n",
        "output_dir = '/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/'\n",
        "\n",
        "# Read XML file\n",
        "with open(xml_path, 'rt') as f:\n",
        "    xml_string = f.read()\n",
        "\n",
        "# Replace the attribute value in the XML string\n",
        "new_xml_string = xml_string.replace('<GC-Content=\"44.83%\"/>', '<GC-Content value=\"44.83%\"/>')\n",
        "\n",
        "# Write the new XML string to a file\n",
        "output_path = os.path.join(output_dir, 'HRR339729_sta_new.xml')\n",
        "with open(output_path, 'w') as f:\n",
        "    f.write(new_xml_string)\n",
        "    \n",
        "print(f\"New XML file saved to {output_path}\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ANIU_oZPZF0r",
        "outputId": "b405a333-b97b-45d6-c0f0-7531f33b03d4"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "New XML file saved to /content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_sta_new.xml\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Remove potential doublets (double balloon effect)\n",
        "\n",
        "This is what the investigators did in the original paper:\n",
        "\n",
        "\n",
        "*   To remove potential doublets, for PBMC samples, cells with UMI counts above 25,000 and detected genes above 5,000 are filtered out. For other tissues, cells with UMI counts above 70,000 and detected genes above 7,500 are filtered out. Additionally, we applied Scrublet (Wolock et al., 2019 link text) to identify potential doublets. The doublet score for each single cell and the threshold based on the bimodal distribution was calculated using default parameters. The expected doublet rate was set to be 0.08, and cells predicted to be doublets or with doubletScore larger than 0.25 were filtered. After quality control, a total of 1,598,708 cells were remained.\n",
        "*   for now we will not be using onliy the PBMC filter methods applied to all tissues\n",
        "*  *We may revisit this later*"
      ],
      "metadata": {
        "id": "-AXEfyzQHUO1"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import xml.etree.ElementTree as ET\n",
        "\n",
        "# Load the count matrix\n",
        "counts = pd.read_csv(\"/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/LungMk/counts/HRR339729/counts_unfiltered/cells_x_genes.mtx\", sep=\" \", skiprows=3, header=None)\n",
        "counts.columns = [\"barcode\", \"gene_id\", \"UMI\"]\n",
        "\n",
        "# Set the barcode column as the index\n",
        "counts = counts.pivot(index=\"barcode\", columns=\"gene_id\", values=\"UMI\")\n",
        "\n",
        "# Load the metadata file from the XML\n",
        "xml_tree = ET.parse('/content/drive/MyDrive/Pate Lab/DNP Folder/Bioinformatics/H17 Lung Mk from COVID-19 Patients/Data files/HRA001149/HRR339729/HRR339729_sta_new.xml')\n",
        "root = xml_tree.getroot()\n",
        "metadata_dict = {}\n",
        "for child in root.iter():\n",
        "    metadata_dict[child.tag] = child.text\n",
        "metadata = pd.DataFrame([metadata_dict])\n",
        "\n",
        "# Determine the filtering thresholds\n",
        "umi_threshold = 25000\n",
        "gene_threshold = 5000\n",
        "\n",
        "# Compute the UMI counts and number of detected genes for each cell\n",
        "umi_counts = counts.sum(axis=0)\n",
        "detected_genes = (counts > 0).sum(axis=0)\n",
        "\n",
        "# Filter out cells with UMI counts or detected genes above the thresholds\n",
        "mask = (umi_counts <= umi_threshold) & (detected_genes <= gene_threshold)\n",
        "filtered_counts = counts.loc[:, mask]\n",
        "\n",
        "# Print some statistics about the filtering\n",
        "print(\"Before filtering:\")\n",
        "print(f\"Number of cells: {counts.shape[1]}\")\n",
        "print(f\"Max UMI count: {umi_counts.max()}\")\n",
        "print(f\"Max detected genes: {detected_genes.max()}\")\n",
        "print(\"\")\n",
        "print(\"After filtering:\")\n",
        "print(f\"Number of cells: {filtered_counts.shape[1]}\")\n",
        "print(f\"Max UMI count: {filtered_counts.sum(axis=0).max()}\")\n",
        "print(f\"Max detected genes: {(filtered_counts > 0).sum(axis=0).max()}\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "xsXp_WrLOiMD",
        "outputId": "d7b36add-d2e3-4a54-c7d2-d5430ad1084f"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Before filtering:\n",
            "Number of cells: 1\n",
            "Max UMI count: 0\n",
            "Max detected genes: 0\n",
            "\n",
            "After filtering:\n",
            "Number of cells: 1\n",
            "Max UMI count: 0\n",
            "Max detected genes: 0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import xml.etree.ElementTree as ET\n",
        "\n",
        "# Load the count matrix\n",
        "counts = pd.read_csv(\"output/counts/bus_output/output.bus.count.txt\", index_col=0)\n",
        "\n",
        "# Load the metadata file from the XML\n",
        "xml_tree = ET.parse('/content/drive/MyDrive/Colab_Notebooks/Lung_Mk/HRR339742/HRR339742_sta.xml')\n",
        "root = xml_tree.getroot()\n",
        "metadata_dict = {}\n",
        "for child in root.iter():\n",
        "    metadata_dict[child.tag] = child.text\n",
        "metadata = pd.DataFrame([metadata_dict])\n",
        "\n",
        "# Determine the filtering thresholds\n",
        "umi_threshold = 25000\n",
        "gene_threshold = 5000\n",
        "\n",
        "# Compute the UMI counts and number of detected genes for each cell\n",
        "umi_counts = counts.sum(axis=0)\n",
        "detected_genes = (counts > 0).sum(axis=0)\n",
        "\n",
        "# Filter out cells with UMI counts or detected genes above the thresholds\n",
        "mask = (umi_counts <= umi_threshold) & (detected_genes <= gene_threshold)\n",
        "filtered_counts = counts.loc[:, mask]\n",
        "\n",
        "# Print some statistics about the filtering\n",
        "print(\"Before filtering:\")\n",
        "print(f\"Number of cells: {counts.shape[1]}\")\n",
        "print(f\"Max UMI count: {umi_counts.max()}\")\n",
        "print(f\"Max detected genes: {detected_genes.max()}\")\n",
        "print(\"\")\n",
        "print(\"After filtering:\")\n",
        "print(f\"Number of cells: {filtered_counts.shape[1]}\")\n",
        "print(f\"Max UMI count: {filtered_counts.sum(axis=0).max()}\")\n",
        "print(f\"Max detected genes: {(filtered_counts > 0).sum(axis=0).max()}\")"
      ],
      "metadata": {
        "id": "S3Ifg7CRHVrZ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Data visualization\n",
        "\n",
        "The stepwise quality control metrics used for individual samples were listed in Table S1. The resulting distribution of UMI counts, gene counts as well as mitochondrial gene percent- age were shown in Figures S1C–S1E. "
      ],
      "metadata": {
        "id": "7z2MClPOHYeE"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import seaborn as sns\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "# Load the count matrix\n",
        "counts = pd.read_csv(\"output/counts/bus_output/output.bus.count.txt\", sep=\"\\t\", index_col=0)\n",
        "\n",
        "# Load the gene annotation file from the metadata XML\n",
        "metadata = pd.read_xml(\"/content/drive/MyDrive/Colab_Notebooks/Lung_Mk/HRR339742/HRR339742_sta.xml\")\n",
        "genes = pd.DataFrame(metadata[\"transcript\"].apply(lambda x: x.get(\"gene_name\")).unique(), columns=[\"gene_name\"])\n",
        "\n",
        "# Identify mitochondrial genes\n",
        "mito_genes = genes[genes[\"gene_name\"].str.startswith(\"MT-\")].index\n",
        "mito_counts = counts.loc[mito_genes].sum(axis=0)\n",
        "total_counts = counts.sum(axis=0)\n",
        "mito_percentage = mito_counts / total_counts * 100\n",
        "\n",
        "# Compute the UMI counts and number of detected genes for each cell\n",
        "umi_counts = counts.sum(axis=0)\n",
        "detected_genes = (counts > 0).sum(axis=0)\n",
        "\n",
        "# Plot the distribution of UMI counts, detected genes, and mitochondrial gene percentage\n",
        "fig, axes = plt.subplots(ncols=3, figsize=(15,5))\n",
        "sns.histplot(umi_counts, ax=axes[0])\n",
        "sns.histplot(detected_genes, ax=axes[1])\n",
        "sns.histplot(mito_percentage, ax=axes[2])\n",
        "axes[0].set_xlabel(\"UMI counts\")\n",
        "axes[1].set_xlabel(\"Number of detected genes\")\n",
        "axes[2].set_xlabel(\"Mitochondrial gene percentage\")\n",
        "plt.show()\n",
        "\n",
        "# before and after plots"
      ],
      "metadata": {
        "id": "ITWo0WojHbA1"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Normaliazed UMI counts\n",
        "\n",
        "This is what the paper did:\n",
        "\n",
        "*  We normalized the UMI counts with the deconvolution strategy implemented in the  R package scran. Specifically, cell-specific size factors were computed by computeSumFactors function and further used to scale the counts for each cell. Then the logarithmic normalized counts were used for the downstream analysis.\n",
        "*  We can use Scnapy instead"
      ],
      "metadata": {
        "id": "IBiB2Qm6He-g"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import scanpy as sc\n",
        "\n",
        "# Load the count matrix\n",
        "adata = sc.read_text(\"output/counts/bus_output/output.bus.count.txt\", delimiter=\"\\t\").T\n",
        "\n",
        "# Normalize the data using Total Count Normalization (TCN)\n",
        "sc.pp.normalize_total(adata, target_sum=1e4)\n",
        "\n",
        "# Scale the data by cell-specific size factors\n",
        "sc.pp.scale(adata, max_value=10)\n",
        "\n",
        "# Logarithmically transform the data\n",
        "sc.pp.log1p(adata)"
      ],
      "metadata": {
        "id": "0XPsngbEHgFL"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "For normalization of UMI counts, the Scanpy package provides several normalization methods, including the Total Count Normalization (TCN) and Normalization by Logarithm (LogNormalize) methods, which are commonly used in single-cell RNA-seq analysis. Here, we first load the count matrix using Scanpy's read_text function. We then normalize the data using the normalize_total function, which scales the counts for each cell so that they have the same total count (in this case, 10,000). We then scale the data by cell-specific size factors using the scale function, and logarithmically transform the data using the log1p function.\n",
        "\n"
      ],
      "metadata": {
        "id": "v9pYnaE_HjvU"
      }
    }
  ]
}