{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/DPariser/DataScience/blob/main/Preprocessing/Creating_H5AD_and_Loom_Files_ID_missing_patients.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "7546582a-5750-47ba-b75e-0419a2caff4e",
      "metadata": {
        "id": "7546582a-5750-47ba-b75e-0419a2caff4e"
      },
      "source": [
        "# Creating H5AD and Loom Files"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "5c790bd3-2371-44c7-aae4-01e086b2ea7f",
      "metadata": {
        "id": "5c790bd3-2371-44c7-aae4-01e086b2ea7f"
      },
      "source": [
        "The following command will generate an RNA count matrix of cells (rows) by genes (columns) in H5AD format, which is a binary format used to store Anndata (https://anndata.readthedocs.io/en/stable/) objects. Also, since the reads were generated with the 10x Genomics Chromium Single Cell v2 &v3 Chemistry, the -x 10xv2 and v3 argument is used (see the alignment file). To view other supported technologies, run kb --list.\n",
        "\n",
        "Note: To output a Loom (https://linnarssonlab.org/loompy/format/index.html) file instead, replace the --h5ad flag with --loom. To obtain the raw matrix output by kb instead of the H5AD or Loom converted files, omit these flags.\n",
        "\n",
        "https://colab.research.google.com/github/pachterlab/kallistobustools/blob/master/docs/tutorials/kb_getting_started/python/kb_intro_2_python.ipynb#scrollTo=ijU_u6uj3Sio\n",
        "\n",
        "https://www.kallistobus.tools/tutorials/kb_getting_started/python/kb_intro_2_python/"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "d621c5b0-7998-4d4e-8272-73e7b52baebd",
      "metadata": {
        "id": "d621c5b0-7998-4d4e-8272-73e7b52baebd"
      },
      "outputs": [],
      "source": [
        "%run Commonimports.ipynb"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "ee837fa9-8d3d-43e0-b3cf-b5363db56cec",
      "metadata": {
        "tags": [],
        "id": "ee837fa9-8d3d-43e0-b3cf-b5363db56cec"
      },
      "outputs": [],
      "source": [
        "import os\n",
        "import numpy as np\n",
        "import anndata"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "4b627cde-9a01-4b99-b41c-f0b323f72f2c",
      "metadata": {
        "tags": [],
        "id": "4b627cde-9a01-4b99-b41c-f0b323f72f2c",
        "outputId": "7a6ae3af-a52d-46ee-bc4d-4379993027a0"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "/home/dpariser\n"
          ]
        }
      ],
      "source": [
        "cd"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "46adcf63-33f7-4ef5-88ca-51638d80b2fc",
      "metadata": {
        "tags": [],
        "id": "46adcf63-33f7-4ef5-88ca-51638d80b2fc"
      },
      "outputs": [],
      "source": [
        "!pip install anndata loompy"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "a4723d8b-fe7e-4bc6-80b0-f357e178ecf7",
      "metadata": {
        "id": "a4723d8b-fe7e-4bc6-80b0-f357e178ecf7"
      },
      "source": [
        "Our matrix, scipy.sparse._coo.coo_matrix, is a COO (Coordinate List) sparse matrix. While Anndata can generally work with COO matrices, it often works more efficiently with CSR (Compressed Sparse Row) matrices. Converting your COO matrix to a CSR matrix before creating the Anndata object could potentially resolve the IORegistryError we have encountering.\n",
        "\n",
        "H5AD Files: Anndata, used for handling H5AD files, typically expects the data in a cell x gene format. Our code works without transposing the matrix for H5AD file creation, it implies that the original matrix is already in the cell x gene format (rows represent cells, columns represent genes).\n",
        "\n",
        "Loom Files: Loom files conventionally expect data in a gene x cell format. If transposing the matrix is necessary for the Loom file creation to work correctly, this further suggests that the original matrix is in the cell x gene format. By transposing it, we are converting it to the gene x cell format, which is expected by Loom."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "3a25ba51-c835-4272-85ac-4c4346d538cc",
      "metadata": {
        "tags": [],
        "id": "3a25ba51-c835-4272-85ac-4c4346d538cc"
      },
      "source": [
        "# H5AD File for One Patient"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "b88c1f86-f189-4228-9cf5-a195d4c6cab6",
      "metadata": {
        "tags": [],
        "id": "b88c1f86-f189-4228-9cf5-a195d4c6cab6"
      },
      "outputs": [],
      "source": [
        "# print(matrix.shape)  # Should be genes x cells\n",
        "# print(len(genes))    # Should match the number of genes\n",
        "# print(len(barcodes)) # Should match the number of cells"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "63456c67-cc6a-44d1-a39f-6670fc5655fb",
      "metadata": {
        "tags": [],
        "id": "63456c67-cc6a-44d1-a39f-6670fc5655fb"
      },
      "outputs": [],
      "source": [
        "# import anndata\n",
        "\n",
        "# # Paths to your files\n",
        "# matrix_path = \"Datafiles/HRR339728/HRR339728_output/counts_unfiltered/cells_x_genes.mtx\"\n",
        "# genes_path = \"Datafiles/HRR339728/HRR339728_output/counts_unfiltered/cells_x_genes.genes.txt\"\n",
        "# barcodes_path = \"Datafiles/HRR339728/HRR339728_output/counts_unfiltered/cells_x_genes.barcodes.txt\"\n",
        "\n",
        "# # Load data\n",
        "# matrix = scipy.io.mmread(matrix_path)\n",
        "\n",
        "# # Convert the COO matrix to CSR format\n",
        "# matrix_csr = csr_matrix(matrix)\n",
        "\n",
        "# genes = pd.read_csv(genes_path, header=None)\n",
        "# barcodes = pd.read_csv(barcodes_path, header=None)\n",
        "\n",
        "# # Create an Anndata object with the CSR matrix\n",
        "# adata = anndata.AnnData(X=matrix_csr)\n",
        "# adata.var_names = genes[0]\n",
        "# adata.obs_names = barcodes[0]\n",
        "\n",
        "# # Ensure uniqueness\n",
        "# adata.var_names_make_unique()\n",
        "# adata.obs_names_make_unique()\n",
        "\n",
        "# # Save as H5AD\n",
        "# adata.write(\"Datafiles/HRR339728/HRR339728_output/HRR339728_H5AD.h5ad\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "920e9bf0-4e13-4d9e-8083-1d12942dbf6b",
      "metadata": {
        "tags": [],
        "id": "920e9bf0-4e13-4d9e-8083-1d12942dbf6b"
      },
      "outputs": [],
      "source": [
        "# adata = anndata.read_h5ad(\"Datafiles/HRR339728/HRR339728_output/HRR339728_H5AD.h5ad\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "8700b2c8-2ca6-4dbe-99db-81df752b6cd0",
      "metadata": {
        "tags": [],
        "id": "8700b2c8-2ca6-4dbe-99db-81df752b6cd0"
      },
      "outputs": [],
      "source": [
        "# adata.shape"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "8e64aa26-61b1-4b3e-b374-c9ba518ef1dc",
      "metadata": {
        "tags": [],
        "id": "8e64aa26-61b1-4b3e-b374-c9ba518ef1dc"
      },
      "outputs": [],
      "source": [
        "# adata.obs.head()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "0f8b37e9-1e14-40f6-a9db-29b685f96235",
      "metadata": {
        "tags": [],
        "id": "0f8b37e9-1e14-40f6-a9db-29b685f96235"
      },
      "outputs": [],
      "source": [
        "# adata.var.head()"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "36a68428-4f30-419e-8db4-1bc1d8742dca",
      "metadata": {
        "tags": [],
        "id": "36a68428-4f30-419e-8db4-1bc1d8742dca"
      },
      "source": [
        "# Loom for one Patient"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "cec55bb1-4e4d-4aa3-a4bf-064d51c3055c",
      "metadata": {
        "tags": [],
        "id": "cec55bb1-4e4d-4aa3-a4bf-064d51c3055c",
        "outputId": "6638e094-c7f6-419a-b28b-95e59e334ca4"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Defaulting to user installation because normal site-packages is not writeable\n",
            "Requirement already satisfied: loompy in ./.local/lib/python3.11/site-packages (3.0.7)\n",
            "Requirement already satisfied: numba in ./.local/lib/python3.11/site-packages (0.58.1)\n",
            "Requirement already satisfied: h5py in ./.local/lib/python3.11/site-packages (from loompy) (3.10.0)\n",
            "Requirement already satisfied: numpy in /home/software/anaconda3/2023.07/lib/python3.11/site-packages (from loompy) (1.24.3)\n",
            "Requirement already satisfied: scipy in /home/software/anaconda3/2023.07/lib/python3.11/site-packages (from loompy) (1.10.1)\n",
            "Requirement already satisfied: setuptools in /home/software/anaconda3/2023.07/lib/python3.11/site-packages (from loompy) (67.8.0)\n",
            "Requirement already satisfied: click in /home/software/anaconda3/2023.07/lib/python3.11/site-packages (from loompy) (8.0.4)\n",
            "Requirement already satisfied: numpy-groupies in ./.local/lib/python3.11/site-packages (from loompy) (0.10.1)\n",
            "Requirement already satisfied: llvmlite<0.42,>=0.41.0dev0 in ./.local/lib/python3.11/site-packages (from numba) (0.41.1)\n"
          ]
        }
      ],
      "source": [
        "!pip install --upgrade loompy numba"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "ab29641d-90b7-4638-9292-0e86b7b6b507",
      "metadata": {
        "tags": [],
        "id": "ab29641d-90b7-4638-9292-0e86b7b6b507"
      },
      "outputs": [],
      "source": [
        "# # Paths to your files\n",
        "# matrix_path = \"Datafiles/HRR339728/HRR339728_output/counts_unfiltered/cells_x_genes.mtx\"\n",
        "# genes_path = \"Datafiles/HRR339728/HRR339728_output/counts_unfiltered/cells_x_genes.genes.txt\"\n",
        "# barcodes_path = \"Datafiles/HRR339728/HRR339728_output/counts_unfiltered/cells_x_genes.barcodes.txt\"\n",
        "\n",
        "# # Load data\n",
        "# matrix = scipy.io.mmread(matrix_path)\n",
        "\n",
        "# # Convert the COO matrix to CSR format and transpose\n",
        "# matrix_csr = csr_matrix(matrix).T\n",
        "\n",
        "# # Load and convert genes and barcodes to numpy arrays\n",
        "# genes = np.array(pd.read_csv(genes_path, header=None)[0])\n",
        "# barcodes = np.array(pd.read_csv(barcodes_path, header=None)[0])\n",
        "\n",
        "# # Create a Loom file with the transposed matrix\n",
        "# loompy.create(\"Datafiles/HRR339728/HRR339728_output/HRR339728.loom\", matrix_csr, {\"Gene\": genes}, {\"CellID\": barcodes})"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "576ef438-125f-44f4-96ab-e013371d7f46",
      "metadata": {
        "tags": [],
        "id": "576ef438-125f-44f4-96ab-e013371d7f46"
      },
      "outputs": [],
      "source": [
        "# # Open the Loom file\n",
        "# with loompy.connect(\"Datafiles/HRR339728/HRR339728_output/HRR339728.loom\") as ds:\n",
        "#     # Print the shape of the dataset (genes x cells)\n",
        "#     print(\"Shape of the dataset:\", ds.shape)\n",
        "\n",
        "#     # Print row attributes keys (usually genes)\n",
        "#     print(\"Row attributes (genes):\", ds.ra.keys())\n",
        "\n",
        "#     # Print column attributes keys (usually cells)\n",
        "#     print(\"Column attributes (cells):\", ds.ca.keys())\n",
        "\n",
        "#     # Optionally, view a subset of the data\n",
        "#     if ds.shape[0] > 0 and ds.shape[1] > 0:\n",
        "#         subset = ds[:, :5]  # First 5 cells\n",
        "#         print(\"Subset of data (first 5 cells):\", subset)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "d056648f-2d74-4238-949e-336f2c56d80b",
      "metadata": {
        "tags": [],
        "id": "d056648f-2d74-4238-949e-336f2c56d80b"
      },
      "outputs": [],
      "source": [
        "# import os\n",
        "# import pandas as pd\n",
        "# import scipy.io\n",
        "# from scipy.sparse import csr_matrix\n",
        "# import anndata\n",
        "# import loompy\n",
        "# import numpy as np\n",
        "# import datetime\n",
        "# import gc\n",
        "# import concurrent.futures\n",
        "\n",
        "# def print_with_timestamp(message):\n",
        "#     timestamp = datetime.datetime.now().strftime(\"%Y-%m-%d %H:%M:%S\")\n",
        "#     print(f\"[{timestamp}] {message}\")\n",
        "\n",
        "# def get_file_size(file_path):\n",
        "#     # Get the file size in bytes\n",
        "#     size_bytes = os.path.getsize(file_path)\n",
        "#     # Convert bytes to megabytes\n",
        "#     size_megabytes = size_bytes / (1024 * 1024)\n",
        "#     return size_megabytes\n",
        "\n",
        "# # Base directory where patient folders are located\n",
        "# base_dir = \"Datafiles\"\n",
        "\n",
        "# # List all patient folders\n",
        "# patient_folders = [folder for folder in os.listdir(base_dir) if folder.startswith(\"HRR\")]\n",
        "# patient_folders.sort()\n",
        "\n",
        "# # Specify the start patient number\n",
        "# start_patient = \"HRR339754\"\n",
        "# start_index = patient_folders.index(start_patient) if start_patient in patient_folders else 0\n",
        "\n",
        "# def process_patient(folder):\n",
        "#     print_with_timestamp(f\"Processing {folder}...\")\n",
        "\n",
        "#     # Define paths to the files\n",
        "#     matrix_path = os.path.join(base_dir, folder, folder + \"_output/counts_unfiltered/cells_x_genes.mtx\")\n",
        "#     genes_path = os.path.join(base_dir, folder, folder + \"_output/counts_unfiltered/cells_x_genes.genes.txt\")\n",
        "#     barcodes_path = os.path.join(base_dir, folder, folder + \"_output/counts_unfiltered/cells_x_genes.barcodes.txt\")\n",
        "\n",
        "#     # Check if the necessary files exist\n",
        "#     if os.path.exists(matrix_path) and os.path.exists(genes_path) and os.path.exists(barcodes_path):\n",
        "#         print_with_timestamp(\"Required files found, loading data...\")\n",
        "\n",
        "#         # Load data\n",
        "#         matrix = scipy.io.mmread(matrix_path)\n",
        "#         genes = pd.read_csv(genes_path, header=None)\n",
        "#         barcodes = pd.read_csv(barcodes_path, header=None)\n",
        "\n",
        "#         # H5AD File Creation\n",
        "#         matrix_csr = csr_matrix(matrix)\n",
        "#         adata = anndata.AnnData(X=matrix_csr)\n",
        "#         adata.var_names = genes[0]\n",
        "#         adata.obs_names = barcodes[0]\n",
        "#         adata.var_names_make_unique()\n",
        "#         adata.obs_names_make_unique()\n",
        "#         h5ad_path = os.path.join(base_dir, folder, folder + \"_H5AD.h5ad\")\n",
        "#         adata.write(h5ad_path)\n",
        "#         h5ad_size = get_file_size(h5ad_path)\n",
        "#         print_with_timestamp(f\"H5AD file created successfully. Size: {h5ad_size:.2f} MB\")\n",
        "\n",
        "#         # Loom File Creation\n",
        "#         matrix_csr_transposed = csr_matrix(matrix).T\n",
        "#         genes_array = np.array(genes[0])\n",
        "#         barcodes_array = np.array(barcodes[0])\n",
        "#         loom_path = os.path.join(base_dir, folder, folder + \".loom\")\n",
        "#         loompy.create(loom_path, matrix_csr_transposed, {\"Gene\": genes_array}, {\"CellID\": barcodes_array})\n",
        "#         loom_size = get_file_size(loom_path)\n",
        "#         print_with_timestamp(f\"Loom file created successfully. Size: {loom_size:.2f} MB\")\n",
        "\n",
        "#         # Clear memory\n",
        "#         del matrix, genes, barcodes, matrix_csr, adata, matrix_csr_transposed, genes_array, barcodes_array\n",
        "#         gc.collect()\n",
        "\n",
        "#         print_with_timestamp(f\"Finished processing {folder}. H5AD and Loom files created.\")\n",
        "#     else:\n",
        "#         print_with_timestamp(f\"Required files for {folder} are missing. Skipping.\")\n",
        "\n",
        "# # Using ThreadPoolExecutor with a controlled number of workers\n",
        "# with concurrent.futures.ThreadPoolExecutor(max_workers=8) as executor:\n",
        "#     futures = [executor.submit(process_patient, folder) for folder in patient_folders[start_index:]]\n",
        "#     concurrent.futures.wait(futures)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "02b64ad6-1ef3-4998-b92e-ca6599a7d1cb",
      "metadata": {
        "id": "02b64ad6-1ef3-4998-b92e-ca6599a7d1cb"
      },
      "source": [
        "# Files for all patients"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "3005de97-e765-404f-b43c-5b0fe74f9388",
      "metadata": {
        "tags": [],
        "id": "3005de97-e765-404f-b43c-5b0fe74f9388",
        "outputId": "87504bb6-1a5e-4986-a6ce-24f3f0278433"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "[2024-01-22 09:13:35] Loom file missing or empty for HRR339754\n",
            "[2024-01-22 09:13:36] H5AD file missing or empty for HRR340077\n",
            "[2024-01-22 09:13:36] Loom file missing or empty for HRR340077\n",
            "[2024-01-22 09:13:36] H5AD file missing or empty for HRR340078\n",
            "[2024-01-22 09:13:36] Loom file missing or empty for HRR340078\n",
            "[2024-01-22 09:13:36] H5AD file missing or empty for HRR340113\n",
            "[2024-01-22 09:13:36] Loom file missing or empty for HRR340113\n",
            "[2024-01-22 09:13:36] H5AD file missing or empty for HRR340265\n",
            "[2024-01-22 09:13:36] Loom file missing or empty for HRR340265\n",
            "[2024-01-22 09:13:36] H5AD file missing or empty for HRR340267\n",
            "[2024-01-22 09:13:36] Loom file missing or empty for HRR340267\n",
            "[2024-01-22 09:13:36] H5AD file missing or empty for HRR340269\n",
            "[2024-01-22 09:13:36] Loom file missing or empty for HRR340269\n",
            "[2024-01-22 09:13:36] H5AD file missing or empty for HRR340271\n",
            "[2024-01-22 09:13:36] Loom file missing or empty for HRR340271\n",
            "[2024-01-22 09:13:36] H5AD file missing or empty for HRR340273\n",
            "[2024-01-22 09:13:36] Loom file missing or empty for HRR340273\n",
            "[2024-01-22 09:13:36] H5AD file missing or empty for HRR340299\n",
            "[2024-01-22 09:13:36] Loom file missing or empty for HRR340299\n",
            "[2024-01-22 09:13:36] H5AD file missing or empty for HRR340301\n",
            "[2024-01-22 09:13:36] Loom file missing or empty for HRR340301\n",
            "[2024-01-22 09:13:36] H5AD file missing or empty for HRR340303\n",
            "[2024-01-22 09:13:36] Loom file missing or empty for HRR340303\n",
            "[2024-01-22 09:13:36] H5AD file missing or empty for HRR340305\n",
            "[2024-01-22 09:13:36] Loom file missing or empty for HRR340305\n",
            "[2024-01-22 09:13:36] H5AD file missing or empty for HRR340307\n",
            "[2024-01-22 09:13:36] Loom file missing or empty for HRR340307\n",
            "[2024-01-22 09:13:36] H5AD file missing or empty for HRR340309\n",
            "[2024-01-22 09:13:36] Loom file missing or empty for HRR340309\n",
            "[2024-01-22 09:13:36] H5AD file missing or empty for HRR340311\n",
            "[2024-01-22 09:13:36] Loom file missing or empty for HRR340311\n",
            "[2024-01-22 09:13:36] H5AD file missing or empty for HRR340313\n",
            "[2024-01-22 09:13:36] Loom file missing or empty for HRR340313\n",
            "[2024-01-22 09:13:36] H5AD file missing or empty for HRR340315\n",
            "[2024-01-22 09:13:36] Loom file missing or empty for HRR340315\n",
            "[2024-01-22 09:13:36] H5AD file missing or empty for HRR340317\n",
            "[2024-01-22 09:13:36] Loom file missing or empty for HRR340317\n",
            "[2024-01-22 09:13:36] H5AD file missing or empty for HRR340319\n",
            "[2024-01-22 09:13:36] Loom file missing or empty for HRR340319\n",
            "[2024-01-22 09:13:36] H5AD file missing or empty for HRR340321\n",
            "[2024-01-22 09:13:36] Loom file missing or empty for HRR340321\n",
            "[2024-01-22 09:13:36] H5AD file missing or empty for HRR340323\n",
            "[2024-01-22 09:13:36] Loom file missing or empty for HRR340323\n",
            "[2024-01-22 09:13:36] H5AD file missing or empty for HRR340325\n",
            "[2024-01-22 09:13:36] Loom file missing or empty for HRR340325\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340622\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340623\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340635\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340635\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340661\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340661\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340683\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340683\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340685\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340685\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340708\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340708\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340717\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340717\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340718\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340718\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340719\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340719\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340720\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340720\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340729\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340729\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340730\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340730\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340731\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340731\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340732\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340732\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340744\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340744\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340800\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340800\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340845\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340845\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340846\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340846\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340847\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340847\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340848\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340848\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340885\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340885\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340888\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340888\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340891\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340891\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340894\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340894\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340897\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340897\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340900\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340900\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340902\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340902\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340905\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340905\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340907\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340907\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340910\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340910\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340913\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340913\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340915\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340915\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340918\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340918\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340920\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340920\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340923\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340923\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340926\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340926\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340929\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340929\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340931\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340931\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340932\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340932\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340935\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340935\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340965\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340965\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340966\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340966\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340967\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340967\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340968\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340968\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340969\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340969\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340970\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340970\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340972\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340972\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340973\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340973\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340974\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340974\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340975\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340975\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340976\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340976\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340977\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340977\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340978\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340978\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR340979\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR340979\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341220\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341220\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341221\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341221\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341222\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341222\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341223\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341223\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341228\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341228\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341229\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341229\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341230\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341230\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341231\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341231\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341236\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341236\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341237\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341237\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341238\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341238\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341239\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341239\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341244\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341244\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341245\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341245\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341246\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341246\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341247\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341247\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341252\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341252\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341253\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341253\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341254\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341254\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341255\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341255\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341260\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341260\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341261\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341261\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341262\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341262\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341263\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341263\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341268\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341268\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341269\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341269\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341270\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341270\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341271\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341271\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341276\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341276\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341277\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341277\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341278\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341278\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341279\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341279\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341284\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341284\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341285\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341285\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341286\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341286\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341287\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341287\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341292\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341292\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341293\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341293\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341294\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341294\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341295\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341295\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341300\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341300\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341301\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341301\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341302\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341302\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341303\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341303\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341308\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341308\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341309\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341309\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341310\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341310\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341311\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341311\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341316\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341316\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341317\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341317\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341318\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341318\n",
            "[2024-01-22 09:13:37] H5AD file missing or empty for HRR341319\n",
            "[2024-01-22 09:13:37] Loom file missing or empty for HRR341319\n",
            "[2024-01-22 09:13:37] Some patient folders are missing valid H5AD or Loom files.\n"
          ]
        }
      ],
      "source": [
        "# # check if all loom and h5ad files are created\n",
        "\n",
        "# import os\n",
        "# import datetime\n",
        "\n",
        "# def print_with_timestamp(message):\n",
        "#     timestamp = datetime.datetime.now().strftime(\"%Y-%m-%d %H:%M:%S\")\n",
        "#     print(f\"[{timestamp}] {message}\")\n",
        "\n",
        "# def check_files_exist_and_non_empty(folder):\n",
        "#     h5ad_path = os.path.join(base_dir, folder, folder + \"_H5AD.h5ad\")\n",
        "#     loom_path = os.path.join(base_dir, folder, folder + \".loom\")\n",
        "\n",
        "#     h5ad_exists = os.path.exists(h5ad_path) and os.path.getsize(h5ad_path) > 0\n",
        "#     loom_exists = os.path.exists(loom_path) and os.path.getsize(loom_path) > 0\n",
        "\n",
        "#     if h5ad_exists and loom_exists:\n",
        "#         return True\n",
        "#     else:\n",
        "#         if not h5ad_exists:\n",
        "#             print_with_timestamp(f\"H5AD file missing or empty for {folder}\")\n",
        "#         if not loom_exists:\n",
        "#             print_with_timestamp(f\"Loom file missing or empty for {folder}\")\n",
        "#         return False\n",
        "\n",
        "# base_dir = \"Datafiles\"\n",
        "# patient_folders = [folder for folder in os.listdir(base_dir) if folder.startswith(\"HRR\")]\n",
        "# patient_folders.sort()\n",
        "\n",
        "# all_files_valid = True\n",
        "# for folder in patient_folders:\n",
        "#     if not check_files_exist_and_non_empty(folder):\n",
        "#         all_files_valid = False\n",
        "\n",
        "# if all_files_valid:\n",
        "#     print_with_timestamp(\"All patient folders have valid H5AD and Loom files.\")\n",
        "# else:\n",
        "#     print_with_timestamp(\"Some patient folders are missing valid H5AD or Loom files.\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "a012c387-26ec-4bdd-acb5-116d85669a14",
      "metadata": {
        "tags": [],
        "id": "a012c387-26ec-4bdd-acb5-116d85669a14"
      },
      "outputs": [],
      "source": [
        "# # Edited for loading only necessary data at the moment it is being processed rather than\n",
        "# # the whole database, reuse variables, and clear memory aggressively.\n",
        "# import os\n",
        "# import pandas as pd\n",
        "# import scipy.io\n",
        "# from scipy.sparse import csr_matrix\n",
        "# import anndata\n",
        "# import loompy\n",
        "# import numpy as np\n",
        "# import datetime\n",
        "# import gc\n",
        "# import concurrent.futures\n",
        "\n",
        "# def print_with_timestamp(message):\n",
        "#     timestamp = datetime.datetime.now().strftime(\"%Y-%m-%d %H:%M:%S\")\n",
        "#     print(f\"[{timestamp}] {message}\")\n",
        "\n",
        "# def get_file_size(file_path):\n",
        "#     size_bytes = os.path.getsize(file_path)\n",
        "#     size_megabytes = size_bytes / (1024 * 1024)\n",
        "#     return size_megabytes\n",
        "\n",
        "# base_dir = \"Datafiles\"\n",
        "# patient_folders = [folder for folder in os.listdir(base_dir) if folder.startswith(\"HRR\")]\n",
        "# patient_folders.sort()\n",
        "\n",
        "# start_patient = \"HRR341172\"\n",
        "# start_index = patient_folders.index(start_patient) if start_patient in patient_folders else 0\n",
        "\n",
        "# def process_patient(folder):\n",
        "#     print_with_timestamp(f\"Processing {folder}...\")\n",
        "\n",
        "#     matrix_path = os.path.join(base_dir, folder, folder + \"_output/counts_unfiltered/cells_x_genes.mtx\")\n",
        "#     genes_path = os.path.join(base_dir, folder, folder + \"_output/counts_unfiltered/cells_x_genes.genes.txt\")\n",
        "#     barcodes_path = os.path.join(base_dir, folder, folder + \"_output/counts_unfiltered/cells_x_genes.barcodes.txt\")\n",
        "\n",
        "#     if os.path.exists(matrix_path) and os.path.exists(genes_path) and os.path.exists(barcodes_path):\n",
        "#         print_with_timestamp(\"Required files found, loading data...\")\n",
        "\n",
        "#         matrix = scipy.io.mmread(matrix_path)\n",
        "#         genes = pd.read_csv(genes_path, header=None)\n",
        "#         barcodes = pd.read_csv(barcodes_path, header=None)\n",
        "\n",
        "#         # Using csr_matrix to keep the data sparse\n",
        "#         matrix_csr = csr_matrix(matrix)\n",
        "\n",
        "#         adata = anndata.AnnData(X=matrix_csr)\n",
        "#         adata.var_names = genes[0]\n",
        "#         adata.obs_names = barcodes[0]\n",
        "#         adata.var_names_make_unique()\n",
        "#         adata.obs_names_make_unique()\n",
        "\n",
        "#         h5ad_path = os.path.join(base_dir, folder, folder + \"_H5AD.h5ad\")\n",
        "#         adata.write(h5ad_path)\n",
        "#         print_with_timestamp(f\"H5AD file for {folder} created. Size: {get_file_size(h5ad_path):.2f} MB\")\n",
        "\n",
        "#         matrix_csr_transposed = matrix_csr.T\n",
        "#         loom_path = os.path.join(base_dir, folder, folder + \".loom\")\n",
        "#         loompy.create(loom_path, matrix_csr_transposed, {\"Gene\": genes[0].values}, {\"CellID\": barcodes[0].values})\n",
        "#         print_with_timestamp(f\"Loom file for {folder} created. Size: {get_file_size(loom_path):.2f} MB\")\n",
        "\n",
        "#         del matrix, genes, barcodes, matrix_csr, adata, matrix_csr_transposed\n",
        "#         gc.collect()\n",
        "\n",
        "#         print_with_timestamp(f\"Finished processing {folder}. H5AD and Loom files for {folder} created.\")\n",
        "#     else:\n",
        "#         print_with_timestamp(f\"Required files for {folder} are missing. Skipping.\")\n",
        "\n",
        "# def process_batch(folders):\n",
        "#     with concurrent.futures.ThreadPoolExecutor(max_workers=1) as executor:\n",
        "#         futures = [executor.submit(process_patient, folder) for folder in folders]\n",
        "#         concurrent.futures.wait(futures)\n",
        "#     gc.collect()\n",
        "\n",
        "# # Define batch size\n",
        "# batch_size = 2\n",
        "\n",
        "# # Process in batches\n",
        "# for i in range(start_index, len(patient_folders), batch_size):\n",
        "#     batch_folders = patient_folders[i:i+batch_size]\n",
        "#     print_with_timestamp(f\"Processing batch {i//batch_size + 1}\")\n",
        "#     process_batch(batch_folders)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "d0ee8000-d441-4cc5-bcbc-5730b4d8f2ac",
      "metadata": {
        "id": "d0ee8000-d441-4cc5-bcbc-5730b4d8f2ac"
      },
      "outputs": [],
      "source": [
        "import os\n",
        "import pandas as pd\n",
        "import scipy.io\n",
        "from scipy.sparse import csr_matrix\n",
        "import anndata\n",
        "import loompy\n",
        "import numpy as np\n",
        "import datetime\n",
        "import gc\n",
        "\n",
        "def print_with_timestamp(message):\n",
        "    timestamp = datetime.datetime.now().strftime(\"%Y-%m-%d %H:%M:%S\")\n",
        "    print(f\"[{timestamp}] {message}\")\n",
        "\n",
        "def get_file_size(file_path):\n",
        "    size_bytes = os.path.getsize(file_path)\n",
        "    size_megabytes = size_bytes / (1024 * 1024)\n",
        "    return size_megabytes\n",
        "\n",
        "base_dir = \"Datafiles\"\n",
        "\n",
        "# List of patient folders for which files are missing\n",
        "missing_patients = [\n",
        "    \"HRR339754\", \"HRR340077\", \"HRR340078\", \"HRR340113\", \"HRR340265\", \"HRR340267\",\n",
        "    \"HRR340269\", \"HRR340271\", \"HRR340273\", \"HRR340299\", \"HRR340301\", \"HRR340303\",\n",
        "    \"HRR340305\", \"HRR340307\", \"HRR340309\", \"HRR340311\", \"HRR340313\", \"HRR340315\",\n",
        "    \"HRR340317\", \"HRR340319\", \"HRR340321\", \"HRR340323\", \"HRR340325\", \"HRR340622\",\n",
        "    \"HRR340623\", \"HRR340635\", \"HRR340661\", \"HRR340683\", \"HRR340685\", \"HRR340708\",\n",
        "    \"HRR340717\", \"HRR340718\", \"HRR340719\", \"HRR340720\", \"HRR340729\", \"HRR340730\",\n",
        "    \"HRR340731\", \"HRR340732\", \"HRR340744\", \"HRR340800\", \"HRR340845\", \"HRR340846\",\n",
        "    \"HRR340847\", \"HRR340848\", \"HRR340885\", \"HRR340888\", \"HRR340891\", \"HRR340894\",\n",
        "    \"HRR340897\", \"HRR340900\", \"HRR340902\", \"HRR340905\", \"HRR340907\", \"HRR340910\",\n",
        "    \"HRR340913\", \"HRR340915\", \"HRR340918\", \"HRR340920\", \"HRR340923\", \"HRR340926\",\n",
        "    \"HRR340929\", \"HRR340931\", \"HRR340932\", \"HRR340935\", \"HRR340965\", \"HRR340966\",\n",
        "    \"HRR340967\", \"HRR340968\", \"HRR340969\", \"HRR340970\", \"HRR340972\", \"HRR340973\",\n",
        "    \"HRR340974\", \"HRR340975\", \"HRR340976\", \"HRR340977\", \"HRR340978\", \"HRR340979\",\n",
        "    \"HRR341220\", \"HRR341221\", \"HRR341222\", \"HRR341223\", \"HRR341228\", \"HRR341229\",\n",
        "    \"HRR341230\", \"HRR341231\", \"HRR341236\", \"HRR341237\", \"HRR341238\", \"HRR341239\",\n",
        "    \"HRR341244\", \"HRR341245\", \"HRR341246\", \"HRR341247\", \"HRR341252\", \"HRR341253\",\n",
        "    \"HRR341254\", \"HRR341255\", \"HRR341260\", \"HRR341261\", \"HRR341262\", \"HRR341263\",\n",
        "    \"HRR341268\", \"HRR341269\", \"HRR341270\", \"HRR341271\", \"HRR341276\", \"HRR341277\",\n",
        "    \"HRR341278\", \"HRR341279\", \"HRR341284\", \"HRR341285\", \"HRR341286\", \"HRR341287\",\n",
        "    \"HRR341292\", \"HRR341293\", \"HRR341294\", \"HRR341295\", \"HRR341300\", \"HRR341301\",\n",
        "    \"HRR341302\", \"HRR341303\", \"HRR341308\", \"HRR341309\", \"HRR341310\", \"HRR341311\",\n",
        "    \"HRR341316\", \"HRR341317\", \"HRR341318\", \"HRR341319\"\n",
        "]\n",
        "\n",
        "def process_patient(folder):\n",
        "    print_with_timestamp(f\"Processing {folder}...\")\n",
        "\n",
        "    matrix_path = os.path.join(base_dir, folder, folder + \"_output/counts_unfiltered/cells_x_genes.mtx\")\n",
        "    genes_path = os.path.join(base_dir, folder, folder + \"_output/counts_unfiltered/cells_x_genes.genes.txt\")\n",
        "    barcodes_path = os.path.join(base_dir, folder, folder + \"_output/counts_unfiltered/cells_x_genes.barcodes.txt\")\n",
        "\n",
        "    if os.path.exists(matrix_path) and os.path.exists(genes_path) and os.path.exists(barcodes_path):\n",
        "        print_with_timestamp(\"Required files found, loading data...\")\n",
        "\n",
        "        matrix = scipy.io.mmread(matrix_path)\n",
        "        genes = pd.read_csv(genes_path, header=None)\n",
        "        barcodes = pd.read_csv(barcodes_path, header=None)\n",
        "\n",
        "        matrix_csr = csr_matrix(matrix)\n",
        "\n",
        "        adata = anndata.AnnData(X=matrix_csr)\n",
        "        adata.var_names = genes[0]\n",
        "        adata.obs_names = barcodes[0]\n",
        "        adata.var_names_make_unique()\n",
        "        adata.obs_names_make_unique()\n",
        "\n",
        "        h5ad_path = os.path.join(base_dir, folder, folder + \"_H5AD.h5ad\")\n",
        "        adata.write(h5ad_path)\n",
        "        print_with_timestamp(f\"H5AD file for {folder} created. Size: {get_file_size(h5ad_path):.2f} MB\")\n",
        "\n",
        "        matrix_csr_transposed = matrix_csr.T\n",
        "        loom_path = os.path.join(base_dir, folder, folder + \".loom\")\n",
        "        loompy.create(loom_path, matrix_csr_transposed, {\"Gene\": genes[0].values}, {\"CellID\": barcodes[0].values})\n",
        "        print_with_timestamp(f\"Loom file for {folder} created. Size: {get_file_size(loom_path):.2f} MB\")\n",
        "\n",
        "        del matrix, genes, barcodes, matrix_csr, adata, matrix_csr_transposed\n",
        "        gc.collect()\n",
        "\n",
        "        print_with_timestamp(f\"Finished processing {folder}. H5AD and Loom files for {folder} created.\")\n",
        "    else:\n",
        "        print_with_timestamp(f\"Required files for {folder} are missing. Skipping.\")\n",
        "\n",
        "# Process each missing patient\n",
        "for folder in missing_patients:\n",
        "    process_patient(folder)"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.3"
    },
    "colab": {
      "provenance": [],
      "include_colab_link": true
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}